---
title: "Loading large-scale H5AD datasets"
subtitle: 'On-disk SingleCellExperiment'
author: "Developed by [Gabriel Hoffman](http://gabrielhoffman.github.io/)"
documentclass: article
vignette: >
  %\VignetteIndexEntry{Loading large-scale H5AD datasets}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
output:
  html_document:
    toc: true
    toc_float: true
---

<!---

cd /Users/gabrielhoffman/workspace/repos/variancePartition/vignettes

# rm -rf dreamlet_cache/

rmarkdown::render("h5ad_on_disk.Rmd")


--->



```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  warning=FALSE,
  message=FALSE,
  error = FALSE,
  eval = FALSE,
  tidy = FALSE,
  dev = c("png"),
  cache = TRUE)
```

As the scale of single cell RNA-seq datasets continues to increase, loading the entire dataset into memory if often impractical.  Dreamlet can take advantage of on-disk data access to dramatically reduce memory usage.

Storing 18K genes across 2M cells with double precision would require 288 Gb memory.  Even if the dataset is sparse so that 80\% of the entires are zero, loading the entire dataset requires 58 Gb of data.  Since `R` uses copy-by-value, data duplication quickly becomes a problem even for a high memory machine.   

Instead of loading the entire dataset into memory, we adopt an on-disk approach that takes advantage of the [H5AD](https://anndata.readthedocs.io/en/latest/fileformat-prose.html) file format built on the [HDF5](https://en.wikipedia.org/wiki/Hierarchical_Data_Format) format in order to dramatically reduce memory usage while still retaining high performance.  The [zellkonverter](https://github.com/theislab/zellkonverter) package uses a [DelayedArray](https://bioconductor.org/packages/DelayedArray/) backend to provide a seamless interface to an on-disk H5AD dataset through the interface of the [SingleCellExperiment](https://bioconductor.org/packages/SingleCellExperiment/) class.  SingleCellExperiment is [Bioconductor's](https://bioconductor.org) core data class for single cell data, and data from other formats (i.e. [Seurat](https://cran.r-project.org/web/packages/Seurat/index.html)) can easily be [converted](https://rdrr.io/cran/Seurat/man/as.SingleCellExperiment.html) to it.  Importantly,  `R`'s copy-by-value is no longer an issue because the data is not duplicated when it is passed to another function, since it remains on-disk.      

This enables any analysis designed for the SingleCellExperiment class to take advantage of seamless on-disk access to large-scale datasets.

## Using zellkonverter
`zellkonverter` >=1.3.0 is required in general  and >=1.3.3 is required for the last example below.  `zellkonverter` is an interface for python code.  Note that the python dependencies are installed the first time `readH5AD()` is run, rather than when the package is installed.  So the first run of `readH5AD()` can take a few minutes, but subsequent runs will be fast.   


### Standard usage 
```{r example}
library(zellkonverter)
library(SingleCellExperiment)

# Create SingleCellExperiment object that points to on-disk H5AD file
sce = readH5AD(h5ad_file, use_hdf5=TRUE)
```

### Merge multiple datasets
```{r example2}
# Read a series of H5AD files into a list
# then combine them into a single merged SingleCellExperiment
sce.lst = lapply( h5ad_files, function(file){
  readH5AD(file, use_hdf5=TRUE)
  })
sce = do.call(cbind, sce.list)
```

### Access alternative data
Some software, including [pegasus](https://pegasus.readthedocs.io/en/stable/), store normalized data in the default portion of the file and save raw counts in the `raw/` field.  Here, load a H5AD file generated by pegasus. 
```{r example3}
# read raw/ from H5AD file
# raw = TRUE tells readH5AD() to read alternative data
# Must use zellkonverter >=1.3.3
sce_in = readH5AD(h5ad_file, use_hdf5=TRUE, raw=TRUE)

# extract raw counts
# include Dimnames and colData
sce = altExp( sce_in, "raw", withDimnames=TRUE, withColData=TRUE)

# convert these alternative data to SingleCellExperiment
sce = as(sce, "SingleCellExperiment")

# Copy over reducedDims too
reducedDims(sce) <- reducedDims(sce_in)
```









