<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="dreamlet">
<title>Modeling continuous cell-level covariates • dreamlet</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Modeling continuous cell-level covariates">
<meta property="og:description" content="dreamlet">
<meta property="og:image" content="http://DiseaseNeurogenomics.github.io/dreamlet/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dreamlet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.64</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/dreamlet.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/cell_level_covariates.html">Modeling continuous cell-level covariates</a>
    <a class="dropdown-item" href="../articles/h5ad_on_disk.html">Handling large H5AD datasets</a>
    <a class="dropdown-item" href="../articles/mashr.html">mashr analysis after dreamlet</a>
    <a class="dropdown-item" href="../articles/non_linear_effects.html">Testing non-linear effects</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/DiseaseNeurogenomics/dreamlet/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Modeling continuous cell-level covariates</h1>
            <h3 data-toc-skip class="subtitle">Collapse using mean value
for pseudobulk data</h3>
                        <h4 data-toc-skip class="author">Developed by <a href="http://gabrielhoffman.github.io/" class="external-link">Gabriel Hoffman</a>
</h4>
            
            <h4 data-toc-skip class="date">Run on 2023-03-07
16:03:27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeurogenomics/dreamlet/blob/HEAD/vignettes/cell_level_covariates.Rmd" class="external-link"><code>vignettes/cell_level_covariates.Rmd</code></a></small>
      <div class="d-none name"><code>cell_level_covariates.Rmd</code></div>
    </div>

    
    
<!---

cd /Users/gabrielhoffman/workspace/repos/dreamlet/vignettes

rmarkdown::render("cell_level_covariates.Rmd")


--->
<style>
body {
text-align: justify}
</style>
<p>Since read counts are summed across cells in a pseudobulk approach,
modeling continuous cell-level covariates also requires a collapsing
step. Here we summarize the values of a variable from a set of cells
using the mean, and store the value for each cell type. Including these
variables in a regression formula uses the summarized values from the
corresponding cell type.</p>
<p>We demonstrate this feature on a lightly modified analysis of PBMCs
from 8 individuals stimulated with interferon-β (<a href="https://www.nature.com/articles/nbt.4042" class="external-link">Kang, et al, 2018,
Nature Biotech</a>).</p>
<div class="section level2">
<h2 id="standard-processing">Standard processing<a class="anchor" aria-label="anchor" href="#standard-processing"></a>
</h2>
<p>Here is the code from the main vignette:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://DiseaseNeurogenomics.github.io/dreamlet" class="external-link">dreamlet</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HelenaLC/muscat" class="external-link">muscat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Download data, specifying EH2259 for the Kang, et al study</span></span>
<span><span class="va">eh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH2259"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># only keep singlet cells with sufficient reads</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">multiplets</span> <span class="op">==</span> <span class="st">'singlet'</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># compute QC metrics</span></span>
<span><span class="va">qc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html" class="external-link">perCellQCMetrics</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove cells with few or many detected genes</span></span>
<span><span class="va">ol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span>metric <span class="op">=</span> <span class="va">qc</span><span class="op">$</span><span class="va">detected</span>, nmads <span class="op">=</span> <span class="fl">2</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="op">!</span><span class="va">ol</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># set variable indicating stimulated (stim) or control (ctrl)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">StimStatus</span> <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">stim</span></span></code></pre></div>
<p>In many datasets, continuous cell-level variables could be mapped
reads, gene count, mitochondrial rate, etc. There are no continuous
cell-level variables in this dataset, so we can simulate two from a
normal distribution:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">$</span><span class="va">value1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">value2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="pseudobulk">Pseudobulk<a class="anchor" aria-label="anchor" href="#pseudobulk"></a>
</h2>
<p>Now compute the pseudobulk using standard code:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">StimStatus</span>, <span class="va">sce</span><span class="op">$</span><span class="va">ind</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create pseudobulk </span></span>
<span><span class="va">pb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregateToPseudoBulk.html">aggregateToPseudoBulk</a></span><span class="op">(</span><span class="va">sce</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"counts"</span>,     </span>
<span>    cluster_id <span class="op">=</span> <span class="st">"cell"</span>,  </span>
<span>    sample_id <span class="op">=</span> <span class="st">"id"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>The means per variable, cell type, and sample are stored in the
pseudobulk <code>SingleCellExperiment</code> object:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">metadata</span><span class="op">(</span><span class="va">pb</span><span class="op">)</span><span class="op">$</span><span class="va">aggr_means</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 128 × 6</span></span></span>
<span><span class="co">## <span style="color: #949494;"># Groups:   cell [8]</span></span></span>
<span><span class="co">##    cell    id         ind cluster   value1   value2</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> B cells ctrl101    101    3.96 -<span style="color: #BB0000;">0.026</span><span style="color: #BB0000; text-decoration: underline;">2</span>   0.038<span style="text-decoration: underline;">1</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> B cells ctrl1015  <span style="text-decoration: underline;">1</span>015    4.00 -<span style="color: #BB0000;">0.028</span><span style="color: #BB0000; text-decoration: underline;">7</span>  -<span style="color: #BB0000;">0.056</span><span style="color: #BB0000; text-decoration: underline;">8</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> B cells ctrl1016  <span style="text-decoration: underline;">1</span>016    4    -<span style="color: #BB0000;">0.007</span><span style="color: #BB0000; text-decoration: underline;">49</span>  0.107  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> B cells ctrl1039  <span style="text-decoration: underline;">1</span>039    4.04 -<span style="color: #BB0000;">0.063</span><span style="color: #BB0000; text-decoration: underline;">2</span>  -<span style="color: #BB0000;">0.060</span><span style="color: #BB0000; text-decoration: underline;">2</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> B cells ctrl107    107    4     0.043<span style="text-decoration: underline;">1</span>   0.042<span style="text-decoration: underline;">7</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> B cells ctrl1244  <span style="text-decoration: underline;">1</span>244    4    -<span style="color: #BB0000;">0.089</span><span style="color: #BB0000; text-decoration: underline;">0</span>   0.055<span style="text-decoration: underline;">2</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> B cells ctrl1256  <span style="text-decoration: underline;">1</span>256    4.01 -<span style="color: #BB0000;">0.064</span><span style="color: #BB0000; text-decoration: underline;">8</span>  -<span style="color: #BB0000;">0.006</span><span style="color: #BB0000; text-decoration: underline;">55</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> B cells ctrl1488  <span style="text-decoration: underline;">1</span>488    4.02 -<span style="color: #BB0000;">0.043</span><span style="color: #BB0000; text-decoration: underline;">9</span>   0.005<span style="text-decoration: underline;">04</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> B cells stim101    101    4.09 -<span style="color: #BB0000;">0.067</span><span style="color: #BB0000; text-decoration: underline;">0</span>   0.042<span style="text-decoration: underline;">1</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> B cells stim1015  <span style="text-decoration: underline;">1</span>015    4.06  0.120   -<span style="color: #BB0000;">0.047</span><span style="color: #BB0000; text-decoration: underline;">6</span> </span></span>
<span><span class="co">## <span style="color: #949494;"># … with 118 more rows</span></span></span></code></pre>
</div>
<div class="section level2">
<h2 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<p>Including these variables in a regression formula uses the summarized
values from the corresponding cell type. This happens behind the scenes,
so the user doesn’t need to distinguish bewteen sample-level variables
stored in <code>colData(pb)</code> and cell-level variables stored in
<code>metadata(pb)$aggr_means</code>.</p>
<p>Variance partition and hypothesis testing proceeds as ususal:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">form</span> <span class="op">=</span> <span class="op">~</span> <span class="va">StimStatus</span> <span class="op">+</span> <span class="va">value1</span> <span class="op">+</span> <span class="va">value2</span></span>
<span></span>
<span><span class="co"># Normalize and apply voom/voomWithDreamWeights</span></span>
<span><span class="va">res.proc</span> <span class="op">=</span> <span class="fu"><a href="../reference/processAssays.html">processAssays</a></span><span class="op">(</span> <span class="va">pb</span>, <span class="va">form</span>, min.count<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run variance partitioning analysis</span></span>
<span><span class="va">vp.lst</span> <span class="op">=</span> <span class="fu"><a href="../reference/fitVarPart.html">fitVarPart</a></span><span class="op">(</span> <span class="va">res.proc</span>, <span class="va">form</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Summarize variance fractions genome-wide for each cell type</span></span>
<span><span class="fu">plotVarPart</span><span class="op">(</span><span class="va">vp.lst</span>, label.angle<span class="op">=</span><span class="fl">60</span><span class="op">)</span> </span></code></pre></div>
<p><img src="cell_level_covariates_files/figure-html/processAssays-1.png" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Differential expression analysis within each assay</span></span>
<span><span class="va">res.dl</span> <span class="op">=</span> <span class="fu"><a href="../reference/dreamlet.html">dreamlet</a></span><span class="op">(</span> <span class="va">res.proc</span>, <span class="va">form</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># dreamlet results include coefficients for value1 and value2</span></span>
<span><span class="va">res.dl</span></span></code></pre></div>
<pre><code><span><span class="co">## class: dreamletResult </span></span>
<span><span class="co">## assays(8): B cells CD14+ Monocytes ... Megakaryocytes NK cells</span></span>
<span><span class="co">## Genes:</span></span>
<span><span class="co">##  min: 182 </span></span>
<span><span class="co">##  max: 5262 </span></span>
<span><span class="co">## details(4): assay n_retain formula formDropsTerms</span></span>
<span><span class="co">## coefNames(4): (Intercept) StimStatusstim value1 value2</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><pre><code><span><span class="co">## R version 4.2.0 (2022-04-22)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin19.6.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Catalina 10.15.7</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib</span></span>
<span><span class="co">## LAPACK: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libLAPACK.dylib</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] splines   stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] lme4_1.1-31                 Matrix_1.5-3               </span></span>
<span><span class="co">##  [3] muscData_1.10.0             scater_1.24.0              </span></span>
<span><span class="co">##  [5] scuttle_1.6.3               SingleCellExperiment_1.18.1</span></span>
<span><span class="co">##  [7] SummarizedExperiment_1.26.1 Biobase_2.58.0             </span></span>
<span><span class="co">##  [9] GenomicRanges_1.48.0        GenomeInfoDb_1.32.4        </span></span>
<span><span class="co">## [11] IRanges_2.30.1              S4Vectors_0.34.0           </span></span>
<span><span class="co">## [13] MatrixGenerics_1.8.1        matrixStats_0.63.0         </span></span>
<span><span class="co">## [15] ExperimentHub_2.4.0         AnnotationHub_3.4.0        </span></span>
<span><span class="co">## [17] BiocFileCache_2.4.0         dbplyr_2.2.1               </span></span>
<span><span class="co">## [19] BiocGenerics_0.44.0         muscat_1.11.2              </span></span>
<span><span class="co">## [21] dreamlet_0.0.62             variancePartition_1.28.5   </span></span>
<span><span class="co">## [23] BiocParallel_1.32.0         limma_3.54.0               </span></span>
<span><span class="co">## [25] ggplot2_3.4.1              </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] rappdirs_0.3.3                scattermore_0.8              </span></span>
<span><span class="co">##   [3] mashr_0.2.57                  ragg_1.2.4                   </span></span>
<span><span class="co">##   [5] tidyr_1.2.1                   clusterGeneration_1.3.7      </span></span>
<span><span class="co">##   [7] bit64_4.0.5                   knitr_1.41                   </span></span>
<span><span class="co">##   [9] irlba_2.3.5.1                 DelayedArray_0.22.0          </span></span>
<span><span class="co">##  [11] data.table_1.14.6             KEGGREST_1.36.3              </span></span>
<span><span class="co">##  [13] RCurl_1.98-1.9                doParallel_1.0.17            </span></span>
<span><span class="co">##  [15] generics_0.1.3                ScaledMatrix_1.4.1           </span></span>
<span><span class="co">##  [17] RhpcBLASctl_0.21-247.1        RSQLite_2.2.19               </span></span>
<span><span class="co">##  [19] future_1.29.0                 bit_4.0.5                    </span></span>
<span><span class="co">##  [21] httpuv_1.6.6                  assertthat_0.2.1             </span></span>
<span><span class="co">##  [23] viridis_0.6.2                 xfun_0.35                    </span></span>
<span><span class="co">##  [25] hms_1.1.2                     jquerylib_0.1.4              </span></span>
<span><span class="co">##  [27] promises_1.2.0.1              babelgene_22.9               </span></span>
<span><span class="co">##  [29] evaluate_0.18                 fansi_1.0.4                  </span></span>
<span><span class="co">##  [31] progress_1.2.2                caTools_1.18.2               </span></span>
<span><span class="co">##  [33] Rgraphviz_2.40.0              DBI_1.1.3                    </span></span>
<span><span class="co">##  [35] geneplotter_1.74.0            purrr_0.3.5                  </span></span>
<span><span class="co">##  [37] ellipsis_0.3.2                dplyr_1.0.10                 </span></span>
<span><span class="co">##  [39] backports_1.4.1               annotate_1.74.0              </span></span>
<span><span class="co">##  [41] aod_1.3.2                     sparseMatrixStats_1.8.0      </span></span>
<span><span class="co">##  [43] vctrs_0.5.2                   abind_1.4-5                  </span></span>
<span><span class="co">##  [45] cachem_1.0.6                  withr_2.5.0                  </span></span>
<span><span class="co">##  [47] sctransform_0.3.5             prettyunits_1.1.1            </span></span>
<span><span class="co">##  [49] softImpute_1.4-1              cluster_2.1.4                </span></span>
<span><span class="co">##  [51] crayon_1.5.2                  genefilter_1.78.0            </span></span>
<span><span class="co">##  [53] labeling_0.4.2                edgeR_3.38.4                 </span></span>
<span><span class="co">##  [55] pkgconfig_2.0.3               nlme_3.1-160                 </span></span>
<span><span class="co">##  [57] vipor_0.4.5                   blme_1.0-5                   </span></span>
<span><span class="co">##  [59] rlang_1.0.6                   globals_0.16.2               </span></span>
<span><span class="co">##  [61] lifecycle_1.0.3               filelock_1.0.2               </span></span>
<span><span class="co">##  [63] rsvd_1.0.5                    invgamma_1.1                 </span></span>
<span><span class="co">##  [65] rprojroot_2.0.3               graph_1.74.0                 </span></span>
<span><span class="co">##  [67] ashr_2.2-54                   Rhdf5lib_1.18.2              </span></span>
<span><span class="co">##  [69] boot_1.3-28.1                 beeswarm_0.4.0               </span></span>
<span><span class="co">##  [71] GlobalOptions_0.1.2           png_0.1-8                    </span></span>
<span><span class="co">##  [73] viridisLite_0.4.1             rjson_0.2.21                 </span></span>
<span><span class="co">##  [75] bitops_1.0-7                  KernSmooth_2.23-20           </span></span>
<span><span class="co">##  [77] rhdf5filters_1.8.0            EnrichmentBrowser_2.26.0     </span></span>
<span><span class="co">##  [79] Biostrings_2.64.1             blob_1.2.3                   </span></span>
<span><span class="co">##  [81] DelayedMatrixStats_1.18.2     shape_1.4.6                  </span></span>
<span><span class="co">##  [83] mixsqp_0.3-48                 stringr_1.5.0                </span></span>
<span><span class="co">##  [85] SQUAREM_2021.1                parallelly_1.32.1            </span></span>
<span><span class="co">##  [87] remaCor_0.0.11                rmeta_3.0                    </span></span>
<span><span class="co">##  [89] beachmat_2.12.0               scales_1.2.1                 </span></span>
<span><span class="co">##  [91] memoise_2.0.1                 GSEABase_1.58.0              </span></span>
<span><span class="co">##  [93] magrittr_2.0.3                plyr_1.8.8                   </span></span>
<span><span class="co">##  [95] gplots_3.1.3                  zlibbioc_1.42.0              </span></span>
<span><span class="co">##  [97] compiler_4.2.0                RColorBrewer_1.1-3           </span></span>
<span><span class="co">##  [99] clue_0.3-63                   KEGGgraph_1.56.0             </span></span>
<span><span class="co">## [101] DESeq2_1.36.0                 cli_3.6.0                    </span></span>
<span><span class="co">## [103] XVector_0.36.0                lmerTest_3.1-3               </span></span>
<span><span class="co">## [105] listenv_0.8.0                 TMB_1.9.1                    </span></span>
<span><span class="co">## [107] MASS_7.3-58.1                 tidyselect_1.2.0             </span></span>
<span><span class="co">## [109] stringi_1.7.8                 textshaping_0.3.6            </span></span>
<span><span class="co">## [111] highr_0.9                     zenith_0.99.13               </span></span>
<span><span class="co">## [113] yaml_2.3.6                    BiocSingular_1.12.0          </span></span>
<span><span class="co">## [115] locfit_1.5-9.6                ggrepel_0.9.2                </span></span>
<span><span class="co">## [117] grid_4.2.0                    sass_0.4.4                   </span></span>
<span><span class="co">## [119] tools_4.2.0                   future.apply_1.10.0          </span></span>
<span><span class="co">## [121] parallel_4.2.0                circlize_0.4.15              </span></span>
<span><span class="co">## [123] foreach_1.5.2                 gridExtra_2.3                </span></span>
<span><span class="co">## [125] farver_2.1.1                  RcppZiggurat_0.1.6           </span></span>
<span><span class="co">## [127] digest_0.6.30                 BiocManager_1.30.19          </span></span>
<span><span class="co">## [129] shiny_1.7.3                   Rcpp_1.0.9                   </span></span>
<span><span class="co">## [131] broom_1.0.1                   later_1.3.0                  </span></span>
<span><span class="co">## [133] BiocVersion_3.15.2            httr_1.4.4                   </span></span>
<span><span class="co">## [135] AnnotationDbi_1.58.0          ComplexHeatmap_2.12.1        </span></span>
<span><span class="co">## [137] Rdpack_2.4                    colorspace_2.1-0             </span></span>
<span><span class="co">## [139] XML_3.99-0.13                 fs_1.5.2                     </span></span>
<span><span class="co">## [141] truncnorm_1.0-8               pkgdown_2.0.6                </span></span>
<span><span class="co">## [143] systemfonts_1.0.4             xtable_1.8-4                 </span></span>
<span><span class="co">## [145] jsonlite_1.8.4                nloptr_2.0.3                 </span></span>
<span><span class="co">## [147] Rfast_2.0.6                   R6_2.5.1                     </span></span>
<span><span class="co">## [149] RUnit_0.4.32                  mime_0.12                    </span></span>
<span><span class="co">## [151] pillar_1.8.1                  htmltools_0.5.3              </span></span>
<span><span class="co">## [153] glue_1.6.2                    fastmap_1.1.0                </span></span>
<span><span class="co">## [155] minqa_1.2.5                   BiocNeighbors_1.14.0         </span></span>
<span><span class="co">## [157] interactiveDisplayBase_1.34.0 codetools_0.2-18             </span></span>
<span><span class="co">## [159] mvtnorm_1.1-3                 utf8_1.2.3                   </span></span>
<span><span class="co">## [161] lattice_0.20-45               bslib_0.4.1                  </span></span>
<span><span class="co">## [163] tibble_3.1.8                  numDeriv_2016.8-1.1          </span></span>
<span><span class="co">## [165] pbkrtest_0.5.1                curl_4.3.3                   </span></span>
<span><span class="co">## [167] ggbeeswarm_0.6.0              gtools_3.9.4                 </span></span>
<span><span class="co">## [169] survival_3.4-0                glmmTMB_1.1.5                </span></span>
<span><span class="co">## [171] rmarkdown_2.18                desc_1.4.2                   </span></span>
<span><span class="co">## [173] munsell_0.5.0                 GetoptLong_1.0.5             </span></span>
<span><span class="co">## [175] rhdf5_2.40.0                  GenomeInfoDbData_1.2.8       </span></span>
<span><span class="co">## [177] iterators_1.0.14              HDF5Array_1.24.2             </span></span>
<span><span class="co">## [179] reshape2_1.4.4                gtable_0.3.1                 </span></span>
<span><span class="co">## [181] msigdbr_7.5.1                 rbibutils_2.2.10</span></span></code></pre>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
