<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="dreamlet">
<title>Dreamlet analysis of single cell RNA-seq • dreamlet</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Dreamlet analysis of single cell RNA-seq">
<meta property="og:description" content="dreamlet">
<meta property="og:image" content="http://DiseaseNeurogenomics.github.io/dreamlet/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dreamlet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.18</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/dreamlet.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/cell_covs.html">Modeling continuous cell-level covariates</a>
    <a class="dropdown-item" href="../articles/h5ad_on_disk.html">Handling large H5AD datasets</a>
    <a class="dropdown-item" href="../articles/mashr.html">mashr analysis after dreamlet</a>
    <a class="dropdown-item" href="../articles/non_lin_eff.html">Testing non-linear effects</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/DiseaseNeurogenomics/dreamlet/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Dreamlet analysis of single cell RNA-seq</h1>
            <h3 data-toc-skip class="subtitle">Linear (mixed) model
analysis of pseudobulk data</h3>
                        <h4 data-toc-skip class="author">Developed by <a href="http://gabrielhoffman.github.io/" class="external-link">Gabriel Hoffman</a>
</h4>
            
            <h4 data-toc-skip class="date">Run on 2023-06-28
12:31:02.003934</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeurogenomics/dreamlet/blob/HEAD/vignettes/dreamlet.Rmd" class="external-link"><code>vignettes/dreamlet.Rmd</code></a></small>
      <div class="d-none name"><code>dreamlet.Rmd</code></div>
    </div>

    
    
<!---

cd /Users/gabrielhoffman/workspace/repos/dreamlet/vignettes

# rm -rf dreamlet_cache/

rmarkdown::render("dreamlet.Rmd")


 devtools::reload("/Users/gabrielhoffman/workspace/repos/dreamlet")

devtools::reload("/Users/gabrielhoffman/workspace/repos/zenith")



--->
<style>
body {
text-align: justify}
</style>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>As the scale of single cell/nucleus RNA-seq has increased, so has the
complexity of study designs. Analysis of datasets with simple study
designs can be performed using linear model as in the <a href="https://bioconductor.org/packages/muscat" class="external-link">muscat package</a>. Yet
analysis of datsets with complex study designs such as repeated measures
or many technical batches can benefit from linear mixed model analysis
to model to correlation structure between samples. We previously
developed <a href="https://doi.org/10.1093/bioinformatics/btaa687" class="external-link">dream</a> to apply
linear mixed models to bulk RNA-seq data using a <a href="https://bioconductor.org/packages/release/bioc/html/limma.html" class="external-link">limma</a>-style
workflow. Dreamlet extends the previous work of dream and <a href="https://www.nature.com/articles/s41467-020-19894-4" class="external-link">muscat</a> to
apply linear mixed models to pseudobulk data. Dreamlet also supports
linear models and facilitates application of 1) <a href="https://bioconductor.org/packages/variancePartition" class="external-link">variancePartition</a>
to quantify the contribution of multiple variables to expression
variation, and 2) <a href="https://github.com/GabrielHoffman/zenith" class="external-link">zenith</a> to perform
gene set analysis on the differential expression signatures.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install this package, start R (version “4.3”) and enter:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select release #1 or #2</span></span>
<span></span>
<span><span class="co"># 1) Bioconductor release</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"dreamlet"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 2) Latest stable release</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"DiseaseNeurogenomics/dreamlet"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="process-single-cell-count-data">Process single cell count data<a class="anchor" aria-label="anchor" href="#process-single-cell-count-data"></a>
</h2>
<p>Here we perform analysis of PBMCs from 8 individuals stimulated with
interferon-β (<a href="https://www.nature.com/articles/nbt.4042" class="external-link">Kang,
et al, 2018, Nature Biotech</a>). This is a small dataset that does not
have repeated measures or high dimensional batch effects, so the
sophisticated features of dreamlet are not strictly necessary. But this
gives us an opportunity to walk through a standard dreamlet
workflow.</p>
<div class="section level3">
<h3 id="preprocess-data">Preprocess data<a class="anchor" aria-label="anchor" href="#preprocess-data"></a>
</h3>
<p>Here, single cell RNA-seq data is downloaded from <a href="https://bioconductor.org/packages/ExperimentHub/" class="external-link">ExperimentHub</a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://DiseaseNeurogenomics.github.io/dreamlet" class="external-link">dreamlet</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HelenaLC/muscat" class="external-link">muscat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://DiseaseNeuroGenomics.github.io/zenith" class="external-link">zenith</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Download data, specifying EH2259 for the Kang, et al study</span></span>
<span><span class="va">eh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH2259"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># only keep singlet cells with sufficient reads</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">multiplets</span> <span class="op">==</span> <span class="st">'singlet'</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># compute QC metrics</span></span>
<span><span class="va">qc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html" class="external-link">perCellQCMetrics</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove cells with few or many detected genes</span></span>
<span><span class="va">ol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span>metric <span class="op">=</span> <span class="va">qc</span><span class="op">$</span><span class="va">detected</span>, nmads <span class="op">=</span> <span class="fl">2</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="op">!</span><span class="va">ol</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># set variable indicating stimulated (stim) or control (ctrl)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">StimStatus</span> <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">stim</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="aggregate-to-pseudobulk">Aggregate to pseudobulk<a class="anchor" aria-label="anchor" href="#aggregate-to-pseudobulk"></a>
</h3>
<p>Dreamlet, like muscat, performs analysis at the pseudobulk-level by
summing <u>raw counts</u> across cells for a given sample and cell type.
<code>aggregateToPseudoBulk</code> is substantially faster for large
on-disk datasets than <a href="https://rdrr.io/bioc/muscat/man/aggregateData.html" class="external-link"><code>muscat::aggregateData</code></a>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Since 'ind' is the individual and 'StimStatus' is the stimulus status,</span></span>
<span><span class="co"># create unique identifier for each sample</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">StimStatus</span>, <span class="va">sce</span><span class="op">$</span><span class="va">ind</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create pseudobulk data by specifying cluster_id and sample_id</span></span>
<span><span class="co"># Count data for each cell type is then stored in the `assay` field</span></span>
<span><span class="co"># assay: entry in assayNames(sce) storing raw counts</span></span>
<span><span class="co"># cluster_id: variable in colData(sce) indicating cell clusters</span></span>
<span><span class="co"># sample_id: variable in colData(sce) indicating sample id for aggregating cells</span></span>
<span><span class="va">pb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregateToPseudoBulk.html">aggregateToPseudoBulk</a></span><span class="op">(</span><span class="va">sce</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"counts"</span>,     </span>
<span>    cluster_id <span class="op">=</span> <span class="st">"cell"</span>,  </span>
<span>    sample_id <span class="op">=</span> <span class="st">"id"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># one 'assay' per cell type</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">pb</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "B cells"           "CD14+ Monocytes"   "CD4 T cells"      </span></span>
<span><span class="co">## [4] "CD8 T cells"       "Dendritic cells"   "FCGR3A+ Monocytes"</span></span>
<span><span class="co">## [7] "Megakaryocytes"    "NK cells"</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="voom-for-pseudobulk">Voom for pseudobulk<a class="anchor" aria-label="anchor" href="#voom-for-pseudobulk"></a>
</h2>
<p>Apply <a href="https://rdrr.io/bioc/limma/man/voom.html" class="external-link">voom</a>-style
normalization for pseudobulk counts within each cell cluster using <a href="https://gabrielhoffman.github.io/variancePartition/reference/voomWithDreamWeights.html" class="external-link">voomWithDreamWeights</a>
to handle random effects (if specified).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Normalize and apply voom/voomWithDreamWeights</span></span>
<span><span class="va">res.proc</span> <span class="op">=</span> <span class="fu"><a href="../reference/processAssays.html">processAssays</a></span><span class="op">(</span> <span class="va">pb</span>, <span class="op">~</span> <span class="va">StimStatus</span>, min.count<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># the resulting object of class dreamletProcessedData stores </span></span>
<span><span class="co"># normalized data and other information</span></span>
<span><span class="va">res.proc</span></span></code></pre></div>
<pre><code><span><span class="co">## class: dreamletProcessedData </span></span>
<span><span class="co">## assays(8): B cells CD14+ Monocytes ... Megakaryocytes NK cells</span></span>
<span><span class="co">## colData(4): ind stim multiplets StimStatus</span></span>
<span><span class="co">## metadata(3): cell id cluster</span></span>
<span><span class="co">## Samples:</span></span>
<span><span class="co">##  min: 5 </span></span>
<span><span class="co">##  max: 16</span></span>
<span><span class="co">## Genes:</span></span>
<span><span class="co">##  min: 182 </span></span>
<span><span class="co">##  max: 5262 </span></span>
<span><span class="co">## details(3): assay n_retained formula</span></span></code></pre>
<p><code><a href="../reference/processAssays.html">processAssays()</a></code> retains samples with at least
<code>min.cells</code> in a given cell type. While dropping a few
samples usually is not a problem, in some cases dropping sames can mean
that a variable included in the regression formula no longer has any
variation. For example, dropping all stimulated samples from analysis of
a given cell type would be mean the variable <code>StimStatus</code> has
no variation and is perfectly colinear with the intercept term. This
colinearity issue is detected internally and variables with these
problem are dropped from the regression formula for that particular cell
type. The number of samples retained and the resulting formula used in
each cell type can be accessed as follows. In this analysis, samples are
dropped from 3 cell types but the original formula remains valid in each
case.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># view details of dropping samples</span></span>
<span><span class="fu"><a href="../reference/details-methods.html">details</a></span><span class="op">(</span><span class="va">res.proc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 8 rows and 3 columns</span></span>
<span><span class="co">##               assay n_retained     formula</span></span>
<span><span class="co">##         &lt;character&gt;  &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">## 1           B cells         16 ~StimStatus</span></span>
<span><span class="co">## 2   CD14+ Monocytes         16 ~StimStatus</span></span>
<span><span class="co">## 3       CD4 T cells         16 ~StimStatus</span></span>
<span><span class="co">## 4       CD8 T cells         16 ~StimStatus</span></span>
<span><span class="co">## 5   Dendritic cells          5 ~StimStatus</span></span>
<span><span class="co">## 6 FCGR3A+ Monocytes         16 ~StimStatus</span></span>
<span><span class="co">## 7    Megakaryocytes          9 ~StimStatus</span></span>
<span><span class="co">## 8          NK cells         16 ~StimStatus</span></span></code></pre>
<p>Here the mean-variance trend from voom is shown for each cell type.
Cell types with sufficient number of cells and reads show a clear
mean-variance trend. While in rare cell types like megakaryocytes, fewer
genes have sufficient reads and the trend is less apparent.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show voom plot for each cell clusters  </span></span>
<span><span class="fu"><a href="../reference/plotVoom-methods.html">plotVoom</a></span><span class="op">(</span> <span class="va">res.proc</span><span class="op">)</span></span></code></pre></div>
<p><img src="dreamlet_files/figure-html/voom.plots-1.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Show plots for subset of cell clusters</span></span>
<span><span class="co"># plotVoom( res.proc[1:3] )</span></span>
<span></span>
<span><span class="co"># Show plots for one cell cluster</span></span>
<span><span class="co"># plotVoom( res.proc[["B cells"]])</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="variance-partitioning">Variance partitioning<a class="anchor" aria-label="anchor" href="#variance-partitioning"></a>
</h2>
<p>The <a href="http://bioconductor.org/packages/variancePartition/" class="external-link">variancePartition</a>
package uses linear and linear mixed models to quanify the contribution
of multiple sources of expression variation at the gene-level. For each
gene it fits a linear (mixed) model and evalutes the fraction of
expression variation explained by each variable.</p>
<p>Variance fractions can be visualized at the gene-level for each cell
type using a bar plot, or genome-wide using a violin plot.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run variance partitioning analysis</span></span>
<span><span class="va">vp.lst</span> <span class="op">=</span> <span class="fu"><a href="../reference/fitVarPart.html">fitVarPart</a></span><span class="op">(</span> <span class="va">res.proc</span>, <span class="op">~</span> <span class="va">StimStatus</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Show variance fractions at the gene-level for each cell type</span></span>
<span><span class="va">genes</span> <span class="op">=</span> <span class="va">vp.lst</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span>
<span><span class="fu"><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/plotPercentBars-method.html" class="external-link">plotPercentBars</a></span><span class="op">(</span><span class="va">vp.lst</span><span class="op">[</span><span class="va">vp.lst</span><span class="op">$</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">genes</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="dreamlet_files/figure-html/vp.barplt-1.png" width="700"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Summarize variance fractions genome-wide for each cell type</span></span>
<span><span class="fu"><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/plotVarPart-method.html" class="external-link">plotVarPart</a></span><span class="op">(</span><span class="va">vp.lst</span>, label.angle<span class="op">=</span><span class="fl">60</span><span class="op">)</span> </span></code></pre></div>
<p><img src="dreamlet_files/figure-html/vp.violin-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="differential-expression">Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h2>
<p>Since the normalized expression data and metadata are stored within
<code>res.proc</code>, only the regression formula remains to be
specified. Here we only included the stimulus status, but analyses of
larger datasets can include covariates and random effects. With formula
<code>~ StimStatus</code>, an intercept is fit and coefficient
<code>StimStatusstim</code> log fold change between simulated and
controls.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Differential expression analysis within each assay,</span></span>
<span><span class="co"># evaluated on the voom normalized data </span></span>
<span><span class="va">res.dl</span> <span class="op">=</span> <span class="fu"><a href="../reference/dreamlet.html">dreamlet</a></span><span class="op">(</span> <span class="va">res.proc</span>, <span class="op">~</span> <span class="va">StimStatus</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># names of estimated coefficients</span></span>
<span><span class="fu"><a href="../reference/coefNames-methods.html">coefNames</a></span><span class="op">(</span><span class="va">res.dl</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## [1] "(Intercept)"    "StimStatusstim"</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the resulting object of class dreamletResult</span></span>
<span><span class="co"># stores results and other information</span></span>
<span><span class="va">res.dl</span></span></code></pre></div>
<pre><code><span><span class="co">## class: dreamletResult </span></span>
<span><span class="co">## assays(8): B cells CD14+ Monocytes ... Megakaryocytes NK cells</span></span>
<span><span class="co">## Genes:</span></span>
<span><span class="co">##  min: 182 </span></span>
<span><span class="co">##  max: 5262 </span></span>
<span><span class="co">## details(6): assay n_retain ... n_errors error_initial</span></span>
<span><span class="co">## coefNames(2): (Intercept) StimStatusstim</span></span></code></pre>
<div class="section level3">
<h3 id="volcano-plots">Volcano plots<a class="anchor" aria-label="anchor" href="#volcano-plots"></a>
</h3>
<p>The volcano plot can indicate the strength of the differential
expression signal with each cell type. Red points indicate FDR &lt;
0.05.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotVolcano-methods.html">plotVolcano</a></span><span class="op">(</span> <span class="va">res.dl</span>, coef <span class="op">=</span> <span class="st">'StimStatusstim'</span> <span class="op">)</span></span></code></pre></div>
<p><img src="dreamlet_files/figure-html/plotVolcano-1.png" width="768"></p>
</div>
<div class="section level3">
<h3 id="gene-level-heatmap">Gene-level heatmap<a class="anchor" aria-label="anchor" href="#gene-level-heatmap"></a>
</h3>
<p>For each cell type and specified gene, show z-statistic from
<code>dreamlet</code> analysis. Grey indicates that insufficient reads
were observed to include the gene in the analysis.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ISG20"</span>, <span class="st">"ISG15"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotGeneHeatmap-methods.html">plotGeneHeatmap</a></span><span class="op">(</span> <span class="va">res.dl</span>, coef<span class="op">=</span><span class="st">"StimStatusstim"</span>, genes<span class="op">=</span><span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<p><img src="dreamlet_files/figure-html/plotGeneHeatmap-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="extract-results">Extract results<a class="anchor" aria-label="anchor" href="#extract-results"></a>
</h3>
<p>Each entry in <code>res.dl</code> stores a model fit by <a href="https://rdrr.io/bioc/variancePartition/man/dream-method.html" class="external-link"><code>dream()</code></a>,
and results can be extracted using <code>topTable()</code> as in
<code>limma</code> by specifying the coefficient of interest. The
results shows the gene name, log fold change, average expression,
t-statistic, p-value, FDR (i.e. adj.P.Val).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># results from full analysis</span></span>
<span><span class="fu">topTable</span><span class="op">(</span><span class="va">res.dl</span>, coef <span class="op">=</span> <span class="st">'StimStatusstim'</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 10 rows and 9 columns</span></span>
<span><span class="co">##              assay          ID     logFC   AveExpr         t     P.Value</span></span>
<span><span class="co">##        &lt;character&gt; &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">## 1          B cells       ISG15   5.49576  10.20272   30.6252 7.46807e-23</span></span>
<span><span class="co">## 2          B cells       ISG20   3.25431  11.38360   29.1273 2.86433e-22</span></span>
<span><span class="co">## 3      CD4 T cells       ISG20   3.08493  10.40388   37.7758 5.09258e-22</span></span>
<span><span class="co">## 4      CD4 T cells        MT2A   4.32241   7.80640   37.0507 7.85918e-22</span></span>
<span><span class="co">## 5  CD14+ Monocytes       IL1RN   7.05190   7.79103   39.3498 1.71111e-21</span></span>
<span><span class="co">## 6      CD4 T cells        IFI6   5.78844   8.52892   34.0945 5.04100e-21</span></span>
<span><span class="co">## 7      CD4 T cells       HERC5   4.29230   6.74189   33.0579 1.00368e-20</span></span>
<span><span class="co">## 8         NK cells       ISG15   4.61275  11.01479   27.3330 1.42681e-20</span></span>
<span><span class="co">## 9      CD4 T cells     TNFSF10   4.49042   7.49081   32.1706 1.84022e-20</span></span>
<span><span class="co">## 10        NK cells       ISG20   3.70085  10.97479   26.9881 1.95877e-20</span></span>
<span><span class="co">##      adj.P.Val         B     z.std</span></span>
<span><span class="co">##      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1  1.09437e-18   42.2686   30.6252</span></span>
<span><span class="co">## 2  2.09869e-18   40.9738   29.1273</span></span>
<span><span class="co">## 3  2.48756e-18   40.5111   37.7758</span></span>
<span><span class="co">## 4  2.87921e-18   39.9926   37.0507</span></span>
<span><span class="co">## 5  5.01493e-18   38.2656   39.3498</span></span>
<span><span class="co">## 6  1.23118e-17   38.2015   34.0945</span></span>
<span><span class="co">## 7  2.10114e-17   37.4037   33.0579</span></span>
<span><span class="co">## 8  2.61356e-17   37.0440   27.3330</span></span>
<span><span class="co">## 9  2.87039e-17   36.9222   32.1706</span></span>
<span><span class="co">## 10 2.87039e-17   36.7473   26.9881</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># only B cells</span></span>
<span><span class="fu">topTable</span><span class="op">(</span><span class="va">res.dl</span><span class="op">[[</span><span class="st">"B cells"</span><span class="op">]</span><span class="op">]</span>, coef <span class="op">=</span> <span class="st">'StimStatusstim'</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           logFC   AveExpr        t      P.Value    adj.P.Val        B</span></span>
<span><span class="co">## ISG15  5.495757 10.202722 30.62515 7.468067e-23 1.464488e-19 42.26855</span></span>
<span><span class="co">## ISG20  3.254305 11.383603 29.12732 2.864329e-22 2.808475e-19 40.97377</span></span>
<span><span class="co">## LY6E   4.205535  8.706648 24.31831 3.482197e-20 2.276196e-17 36.15791</span></span>
<span><span class="co">## PLSCR1 4.085769  8.271191 23.47482 8.836105e-20 3.556815e-17 35.20946</span></span>
<span><span class="co">## EPSTI1 3.666105  8.107930 23.45165 9.068880e-20 3.556815e-17 35.18356</span></span>
<span><span class="co">## SAT1   2.141210  9.657178 21.92490 5.308264e-19 1.631423e-16 33.45779</span></span>
<span><span class="co">## IRF7   3.665702  8.678591 21.84738 5.823541e-19 1.631423e-16 33.38287</span></span>
<span><span class="co">## TRIM22 2.934663  7.201113 21.68805 7.051546e-19 1.728510e-16 33.11351</span></span>
<span><span class="co">## UBE2L6 2.697049  9.241028 21.53019 8.534002e-19 1.859464e-16 32.99410</span></span>
<span><span class="co">## SOCS1  2.433571  8.588454 21.16992 1.325324e-18 2.598960e-16 32.56550</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="forest-plot">Forest plot<a class="anchor" aria-label="anchor" href="#forest-plot"></a>
</h3>
<p>A forest plot shows the log fold change and standard error of a given
gene across all cell types. The color indicates the FDR.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotForest-methods.html">plotForest</a></span><span class="op">(</span> <span class="va">res.dl</span>, coef <span class="op">=</span> <span class="st">'StimStatusstim'</span>, gene <span class="op">=</span> <span class="st">'ISG20'</span><span class="op">)</span></span></code></pre></div>
<p><img src="dreamlet_files/figure-html/forrest-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="advanced-used-of-contrasts">Advanced used of contrasts<a class="anchor" aria-label="anchor" href="#advanced-used-of-contrasts"></a>
</h3>
<p>A hypothesis test of the <em>difference</em> between two or more
coefficients can be performed using contrasts. The contrast matrix is
evaluated for each cell type in the backend, so only the contrast string
must be supplied to <code><a href="../reference/dreamlet.html">dreamlet()</a></code>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a contrasts called 'Diff' that is the difference between expression</span></span>
<span><span class="co"># in the stimulated and controls.</span></span>
<span><span class="co"># More than one can be specified</span></span>
<span><span class="va">contrasts</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Diff <span class="op">=</span> <span class="st">'StimStatusstim - StimStatusctrl'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Evalaute the regression model without an intercept term.</span></span>
<span><span class="co"># Instead estimate the mean expression in stimulated, controls and then</span></span>
<span><span class="co"># set Diff to the difference between the two</span></span>
<span><span class="va">res.dl2</span> <span class="op">=</span> <span class="fu"><a href="../reference/dreamlet.html">dreamlet</a></span><span class="op">(</span> <span class="va">res.proc</span>, <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">StimStatus</span>, contrasts<span class="op">=</span><span class="va">contrasts</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># see estimated coefficients</span></span>
<span><span class="fu"><a href="../reference/coefNames-methods.html">coefNames</a></span><span class="op">(</span><span class="va">res.dl2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Diff"           "StimStatusctrl" "StimStatusstim"</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Volcano plot of Diff</span></span>
<span><span class="fu"><a href="../reference/plotVolcano-methods.html">plotVolcano</a></span><span class="op">(</span> <span class="va">res.dl2</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, coef <span class="op">=</span> <span class="st">'Diff'</span> <span class="op">)</span></span></code></pre></div>
<p><img src="dreamlet_files/figure-html/dreamlet.contrasts-1.png" width="700"></p>
<p>This new <code>Diff</code> variable can then be used downstream for
analysis asking for a coefficient. But note that since there is no
intercept term in this model, the meaning of <code>StimStatusstim</code>
changes here. When the formula is <code>0 + StimStatus</code> then
<code>StimStatusstim</code> is the mean expression in stimulated
samples.</p>
<p>For further information about using contrasts see <a href="https://gabrielhoffman.github.io/variancePartition/reference/makeContrastsDream.html" class="external-link">makeContrastsDream()</a>
and <a href="https://gabrielhoffman.github.io/variancePartition/articles/dream.html#advanced-hypothesis-testing-1" class="external-link">vignette</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="gene-set-analysis">Gene set analysis<a class="anchor" aria-label="anchor" href="#gene-set-analysis"></a>
</h2>
<p>While standard enrichment methods like Fishers exact test, requires
specifying a FDR cutoff to identify differentially expressed genes.
However, dichotomizing differential expression results is often too
simple and ignores the quantitative variation captured by the
differential expression test statistics. Here we use
<code>zenith</code>, a wrapper for <a href="https://rdrr.io/bioc/limma/man/camera.html" class="external-link"><code>limma::camera</code></a>,
to perform gene set analysis using the full spectrum of differential
expression test statistics. <code>zenith/camera</code> is a <a href="https://www.nature.com/articles/nrg.2016.29" class="external-link">competetive test</a>
that compares the mean test statistic for genes in a given gene set, to
genes not in that set while accounting for correlation between
genes.</p>
<p>Here, <code>zenith_gsa</code> takes a <code>dreamletResult</code>
object, the coefficient of interest, and gene sets as a
<code>GeneSetCollection</code> object from <a href="https://bioconductor.org/packages/GSEABase/" class="external-link">GSEABase</a>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load Gene Ontology database </span></span>
<span><span class="co"># use gene 'SYMBOL', or 'ENSEMBL' id</span></span>
<span><span class="co"># use get_MSigDB() to load MSigDB</span></span>
<span><span class="co"># Use Cellular Component (i.e. CC) to reduce run time here</span></span>
<span><span class="va">go.gs</span> <span class="op">=</span> <span class="fu"><a href="http://gabrielhoffman.github.io/zenith/reference/get_GeneOntology.html" class="external-link">get_GeneOntology</a></span><span class="op">(</span><span class="st">"CC"</span>, to<span class="op">=</span><span class="st">"SYMBOL"</span><span class="op">)</span></span>
<span>   </span>
<span><span class="co"># Run zenith gene set analysis on result of dreamlet</span></span>
<span><span class="va">res_zenith</span> <span class="op">=</span> <span class="fu"><a href="http://gabrielhoffman.github.io/zenith/reference/zenith_gsa-methods.html" class="external-link">zenith_gsa</a></span><span class="op">(</span><span class="va">res.dl</span>, coef <span class="op">=</span> <span class="st">'StimStatusstim'</span>, <span class="va">go.gs</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># examine results for each ell type and gene set</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res_zenith</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     assay           coef</span></span>
<span><span class="co">## 1 B cells StimStatusstim</span></span>
<span><span class="co">## 2 B cells StimStatusstim</span></span>
<span><span class="co">## 3 B cells StimStatusstim</span></span>
<span><span class="co">## 4 B cells StimStatusstim</span></span>
<span><span class="co">## 5 B cells StimStatusstim</span></span>
<span><span class="co">## 6 B cells StimStatusstim</span></span>
<span><span class="co">##                                                        Geneset NGenes</span></span>
<span><span class="co">## 1                                GO0022626: cytosolic ribosome     74</span></span>
<span><span class="co">## 2                 GO0022625: cytosolic large ribosomal subunit     45</span></span>
<span><span class="co">## 3                             GO0031256: leading edge membrane     20</span></span>
<span><span class="co">## 4 GO0090575: RNA polymerase II transcription regulator complex     37</span></span>
<span><span class="co">## 5                          GO0000151: ubiquitin ligase complex     34</span></span>
<span><span class="co">## 6                           GO0005839: proteasome core complex     13</span></span>
<span><span class="co">##   Correlation     delta        se      p.less   p.greater      PValue Direction</span></span>
<span><span class="co">## 1        0.01 -1.418199 0.4215391 0.002314999 0.997685001 0.004629999      Down</span></span>
<span><span class="co">## 2        0.01 -1.435250 0.4925328 0.005661944 0.994338056 0.011323888      Down</span></span>
<span><span class="co">## 3        0.01 -1.850421 0.6696010 0.007618276 0.992381724 0.015236552      Down</span></span>
<span><span class="co">## 4        0.01  1.455379 0.5275552 0.992310405 0.007689595 0.015379191        Up</span></span>
<span><span class="co">## 5        0.01  1.425649 0.5441665 0.989909601 0.010090399 0.020180797        Up</span></span>
<span><span class="co">## 6        0.01  2.007864 0.8050928 0.987117087 0.012882913 0.025765826        Up</span></span>
<span><span class="co">##         FDR</span></span>
<span><span class="co">## 1 0.2561073</span></span>
<span><span class="co">## 2 0.4296796</span></span>
<span><span class="co">## 3 0.4457768</span></span>
<span><span class="co">## 4 0.4457768</span></span>
<span><span class="co">## 5 0.4587962</span></span>
<span><span class="co">## 6 0.5008720</span></span></code></pre>
<div class="section level3">
<h3 id="heatmap-of-top-genesets">Heatmap of top genesets<a class="anchor" aria-label="anchor" href="#heatmap-of-top-genesets"></a>
</h3>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># for each cell type select 5 genesets with largest t-statistic</span></span>
<span><span class="co"># and 1 geneset with the lowest</span></span>
<span><span class="co"># Grey boxes indicate the gene set could not be evaluted because</span></span>
<span><span class="co">#    to few genes were represented</span></span>
<span><span class="fu"><a href="http://gabrielhoffman.github.io/zenith/reference/plotZenithResults.html" class="external-link">plotZenithResults</a></span><span class="op">(</span><span class="va">res_zenith</span>, <span class="fl">5</span>, <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="dreamlet_files/figure-html/heatmap-1.png" width="768"></p>
<div class="section level4">
<h4 id="all-gene-sets-with-fdr-30">All gene sets with FDR &lt; 30%<a class="anchor" aria-label="anchor" href="#all-gene-sets-with-fdr-30"></a>
</h4>
<p>Here, show all genes with FDR &lt; 5% in any cell type</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get genesets with FDR &lt; 30%</span></span>
<span><span class="co"># Few significant genesets because uses Cellular Component (i.e. CC) </span></span>
<span><span class="va">gs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">res_zenith</span><span class="op">$</span><span class="va">Geneset</span><span class="op">[</span><span class="va">res_zenith</span><span class="op">$</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.3</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># keep only results of these genesets</span></span>
<span><span class="va">df</span> <span class="op">=</span> <span class="va">res_zenith</span><span class="op">[</span><span class="va">res_zenith</span><span class="op">$</span><span class="va">Geneset</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">gs</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># plot results, but with no limit based on the highest/lowest t-statistic</span></span>
<span><span class="fu"><a href="http://gabrielhoffman.github.io/zenith/reference/plotZenithResults.html" class="external-link">plotZenithResults</a></span><span class="op">(</span><span class="va">df</span>, <span class="cn">Inf</span>, <span class="cn">Inf</span><span class="op">)</span></span></code></pre></div>
<p><img src="dreamlet_files/figure-html/heatmap2-1.png" width="768"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="comparing-expression-in-clusters">Comparing expression in clusters<a class="anchor" aria-label="anchor" href="#comparing-expression-in-clusters"></a>
</h2>
<p>Identifying genes that are differentially expressed between cell
clusters incorporates a paired analysis design, since each individual is
observed for each cell cluster.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># test differential expression between B cells and the rest of the cell clusters</span></span>
<span><span class="va">ct.pairs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4 T cells"</span>, <span class="st">"rest"</span><span class="op">)</span></span>
<span> </span>
<span><span class="va">fit</span> <span class="op">=</span> <span class="fu"><a href="../reference/dreamletCompareClusters.html">dreamletCompareClusters</a></span><span class="op">(</span> <span class="va">pb</span>, <span class="va">ct.pairs</span>, method<span class="op">=</span><span class="st">"fixed"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The coefficient 'compare' is the value logFC between test and baseline:</span></span>
<span><span class="co"># compare = cellClustertest - cellClusterbaseline </span></span>
<span><span class="va">df_Bcell</span> <span class="op">=</span> <span class="fu">topTable</span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="st">"compare"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">df_Bcell</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##              logFC   AveExpr         t      P.Value    adj.P.Val        B</span></span>
<span><span class="co">## TYROBP   -7.409815  7.917618 -74.81920 6.617529e-29 3.110238e-25 55.09105</span></span>
<span><span class="co">## FTL      -4.920218 13.169468 -53.69652 1.277073e-25 3.001122e-22 48.79651</span></span>
<span><span class="co">## C15orf48 -7.177537  8.398874 -48.99879 1.024387e-24 1.604873e-21 46.53806</span></span>
<span><span class="co">## FCER1G   -6.459344  8.016642 -45.02615 6.986218e-24 6.956542e-21 44.69191</span></span>
<span><span class="co">## CST3     -6.728115  6.935446 -44.61372 8.607869e-24 6.956542e-21 44.41761</span></span>
<span><span class="co">## CD63     -5.128597  8.760309 -44.55239 8.880692e-24 6.956542e-21 44.53728</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="gene-cluster-specificity">Gene-cluster specificity<a class="anchor" aria-label="anchor" href="#gene-cluster-specificity"></a>
</h2>
<p>Evaluate the specificity of each gene for each cluster, retaining
only highly expressed genes:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df_cts</span> <span class="op">=</span> <span class="fu"><a href="../reference/cellTypeSpecificity.html">cellTypeSpecificity</a></span><span class="op">(</span> <span class="va">pb</span> <span class="op">)</span></span>
<span> </span>
<span><span class="co"># retain only genes with total CPM summed across cell type &gt; 100</span></span>
<span><span class="va">df_cts</span> <span class="op">=</span> <span class="va">df_cts</span><span class="op">[</span><span class="va">df_cts</span><span class="op">$</span><span class="va">totalCPM</span> <span class="op">&gt;</span> <span class="fl">100</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># Violin plot of specificity score for each cell type</span></span>
<span><span class="fu"><a href="../reference/plotViolin-methods.html">plotViolin</a></span><span class="op">(</span><span class="va">df_cts</span><span class="op">)</span> </span></code></pre></div>
<p><img src="dreamlet_files/figure-html/cellTypeSpecificity-1.png" width="700"></p>
<p>Highlight expression fraction for most specific gene from each cell
type:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">df_cts</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">df_cts</span>, <span class="fl">2</span>, <span class="va">which.max</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/plotPercentBars-method.html" class="external-link">plotPercentBars</a></span><span class="op">(</span> <span class="va">df_cts</span>, genes <span class="op">=</span> <span class="va">genes</span> <span class="op">)</span></span></code></pre></div>
<p><img src="dreamlet_files/figure-html/plotPercentBars-1.png" width="700"></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dreamlet</span><span class="fu">::</span><span class="fu"><a href="../reference/plotHeatmap-methods.html">plotHeatmap</a></span><span class="op">(</span> <span class="va">df_cts</span>, genes <span class="op">=</span> <span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<p><img src="dreamlet_files/figure-html/plotPercentBars-2.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin22.4.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Ventura 13.3.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Users/gabrielhoffman/prog/R-4.3.0/lib/libRblas.dylib </span></span>
<span><span class="co">## LAPACK: /usr/local/Cellar/r/4.2.3/lib/R/lib/libRlapack.dylib;  LAPACK version 3.10.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/New_York</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] GO.db_3.17.0                org.Hs.eg.db_3.17.0        </span></span>
<span><span class="co">##  [3] AnnotationDbi_1.62.1        muscData_1.14.0            </span></span>
<span><span class="co">##  [5] scater_1.28.0               scuttle_1.10.1             </span></span>
<span><span class="co">##  [7] SingleCellExperiment_1.22.0 SummarizedExperiment_1.30.1</span></span>
<span><span class="co">##  [9] Biobase_2.60.0              GenomicRanges_1.52.0       </span></span>
<span><span class="co">## [11] GenomeInfoDb_1.36.0         IRanges_2.34.0             </span></span>
<span><span class="co">## [13] S4Vectors_0.38.1            MatrixGenerics_1.12.0      </span></span>
<span><span class="co">## [15] matrixStats_0.63.0          zenith_1.2.0               </span></span>
<span><span class="co">## [17] ExperimentHub_2.8.0         AnnotationHub_3.8.0        </span></span>
<span><span class="co">## [19] BiocFileCache_2.8.0         dbplyr_2.3.2               </span></span>
<span><span class="co">## [21] BiocGenerics_0.46.0         muscat_1.14.0              </span></span>
<span><span class="co">## [23] dreamlet_0.99.13            variancePartition_2.0.5    </span></span>
<span><span class="co">## [25] BiocParallel_1.34.2         limma_3.56.1               </span></span>
<span><span class="co">## [27] ggplot2_3.4.2               BiocStyle_2.28.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fs_1.6.2                      bitops_1.0-7                 </span></span>
<span><span class="co">##   [3] httr_1.4.6                    RColorBrewer_1.1-3           </span></span>
<span><span class="co">##   [5] doParallel_1.0.17             Rgraphviz_2.44.0             </span></span>
<span><span class="co">##   [7] numDeriv_2016.8-1.1           tools_4.3.0                  </span></span>
<span><span class="co">##   [9] sctransform_0.3.5             backports_1.4.1              </span></span>
<span><span class="co">##  [11] utf8_1.2.3                    R6_2.5.1                     </span></span>
<span><span class="co">##  [13] GetoptLong_1.0.5              withr_2.5.0                  </span></span>
<span><span class="co">##  [15] prettyunits_1.1.1             gridExtra_2.3                </span></span>
<span><span class="co">##  [17] cli_3.6.1                     textshaping_0.3.6            </span></span>
<span><span class="co">##  [19] labeling_0.4.2                sass_0.4.6                   </span></span>
<span><span class="co">##  [21] KEGGgraph_1.60.0              SQUAREM_2021.1               </span></span>
<span><span class="co">##  [23] mvtnorm_1.1-3                 blme_1.0-5                   </span></span>
<span><span class="co">##  [25] pkgdown_2.0.7                 mixsqp_0.3-48                </span></span>
<span><span class="co">##  [27] systemfonts_1.0.4             parallelly_1.35.0            </span></span>
<span><span class="co">##  [29] invgamma_1.1                  RSQLite_2.3.1                </span></span>
<span><span class="co">##  [31] generics_0.1.3                shape_1.4.6                  </span></span>
<span><span class="co">##  [33] gtools_3.9.4                  dplyr_1.1.2                  </span></span>
<span><span class="co">##  [35] Matrix_1.5-4.1                ggbeeswarm_0.7.2             </span></span>
<span><span class="co">##  [37] fansi_1.0.4                   abind_1.4-5                  </span></span>
<span><span class="co">##  [39] lifecycle_1.0.3               yaml_2.3.7                   </span></span>
<span><span class="co">##  [41] edgeR_3.42.2                  clusterGeneration_1.3.7      </span></span>
<span><span class="co">##  [43] gplots_3.1.3                  grid_4.3.0                   </span></span>
<span><span class="co">##  [45] blob_1.2.4                    promises_1.2.0.1             </span></span>
<span><span class="co">##  [47] crayon_1.5.2                  lattice_0.21-8               </span></span>
<span><span class="co">##  [49] beachmat_2.16.0               msigdbr_7.5.1                </span></span>
<span><span class="co">##  [51] annotate_1.78.0               KEGGREST_1.40.0              </span></span>
<span><span class="co">##  [53] pillar_1.9.0                  knitr_1.43                   </span></span>
<span><span class="co">##  [55] ComplexHeatmap_2.16.0         rjson_0.2.21                 </span></span>
<span><span class="co">##  [57] boot_1.3-28.1                 corpcor_1.6.10               </span></span>
<span><span class="co">##  [59] future.apply_1.11.0           codetools_0.2-19             </span></span>
<span><span class="co">##  [61] glue_1.6.2                    data.table_1.14.8            </span></span>
<span><span class="co">##  [63] vctrs_0.6.2                   png_0.1-8                    </span></span>
<span><span class="co">##  [65] Rdpack_2.4                    gtable_0.3.3                 </span></span>
<span><span class="co">##  [67] assertthat_0.2.1              cachem_1.0.8                 </span></span>
<span><span class="co">##  [69] xfun_0.39                     mime_0.12                    </span></span>
<span><span class="co">##  [71] rbibutils_2.2.13              S4Arrays_1.0.4               </span></span>
<span><span class="co">##  [73] Rfast_2.0.7                   iterators_1.0.14             </span></span>
<span><span class="co">##  [75] ellipsis_0.3.2                interactiveDisplayBase_1.38.0</span></span>
<span><span class="co">##  [77] nlme_3.1-162                  pbkrtest_0.5.2               </span></span>
<span><span class="co">##  [79] bit64_4.0.5                   progress_1.2.2               </span></span>
<span><span class="co">##  [81] filelock_1.0.2                rprojroot_2.0.3              </span></span>
<span><span class="co">##  [83] bslib_0.4.2                   TMB_1.9.4                    </span></span>
<span><span class="co">##  [85] irlba_2.3.5.1                 vipor_0.4.5                  </span></span>
<span><span class="co">##  [87] KernSmooth_2.23-21            colorspace_2.1-0             </span></span>
<span><span class="co">##  [89] rmeta_3.0                     DBI_1.1.3                    </span></span>
<span><span class="co">##  [91] DESeq2_1.40.1                 tidyselect_1.2.0             </span></span>
<span><span class="co">##  [93] bit_4.0.5                     compiler_4.3.0               </span></span>
<span><span class="co">##  [95] curl_5.0.0                    graph_1.78.0                 </span></span>
<span><span class="co">##  [97] BiocNeighbors_1.18.0          desc_1.4.2                   </span></span>
<span><span class="co">##  [99] DelayedArray_0.26.3           bookdown_0.34                </span></span>
<span><span class="co">## [101] scales_1.2.1                  caTools_1.18.2               </span></span>
<span><span class="co">## [103] remaCor_0.0.11                rappdirs_0.3.3               </span></span>
<span><span class="co">## [105] stringr_1.5.0                 digest_0.6.31                </span></span>
<span><span class="co">## [107] minqa_1.2.5                   rmarkdown_2.21               </span></span>
<span><span class="co">## [109] aod_1.3.2                     XVector_0.40.0               </span></span>
<span><span class="co">## [111] RhpcBLASctl_0.23-42           htmltools_0.5.5              </span></span>
<span><span class="co">## [113] pkgconfig_2.0.3               lme4_1.1-33                  </span></span>
<span><span class="co">## [115] sparseMatrixStats_1.12.0      highr_0.10                   </span></span>
<span><span class="co">## [117] mashr_0.2.69                  fastmap_1.1.1                </span></span>
<span><span class="co">## [119] rlang_1.1.1                   GlobalOptions_0.1.2          </span></span>
<span><span class="co">## [121] shiny_1.7.4                   DelayedMatrixStats_1.22.0    </span></span>
<span><span class="co">## [123] farver_2.1.1                  jquerylib_0.1.4              </span></span>
<span><span class="co">## [125] jsonlite_1.8.4                BiocSingular_1.16.0          </span></span>
<span><span class="co">## [127] RCurl_1.98-1.12               magrittr_2.0.3               </span></span>
<span><span class="co">## [129] GenomeInfoDbData_1.2.10       munsell_0.5.0                </span></span>
<span><span class="co">## [131] Rcpp_1.0.10                   babelgene_22.9               </span></span>
<span><span class="co">## [133] viridis_0.6.3                 EnrichmentBrowser_2.30.1     </span></span>
<span><span class="co">## [135] RcppZiggurat_0.1.6            stringi_1.7.12               </span></span>
<span><span class="co">## [137] zlibbioc_1.46.0               MASS_7.3-60                  </span></span>
<span><span class="co">## [139] plyr_1.8.8                    parallel_4.3.0               </span></span>
<span><span class="co">## [141] listenv_0.9.0                 ggrepel_0.9.3                </span></span>
<span><span class="co">## [143] Biostrings_2.68.1             splines_4.3.0                </span></span>
<span><span class="co">## [145] hms_1.1.3                     circlize_0.4.15              </span></span>
<span><span class="co">## [147] locfit_1.5-9.7                RUnit_0.4.32                 </span></span>
<span><span class="co">## [149] reshape2_1.4.4                ScaledMatrix_1.8.1           </span></span>
<span><span class="co">## [151] BiocVersion_3.17.1            XML_3.99-0.14                </span></span>
<span><span class="co">## [153] evaluate_0.21                 BiocManager_1.30.20          </span></span>
<span><span class="co">## [155] httpuv_1.6.11                 nloptr_2.0.3                 </span></span>
<span><span class="co">## [157] foreach_1.5.2                 tidyr_1.3.0                  </span></span>
<span><span class="co">## [159] purrr_1.0.1                   future_1.32.0                </span></span>
<span><span class="co">## [161] clue_0.3-64                   scattermore_1.1              </span></span>
<span><span class="co">## [163] ashr_2.2-54                   rsvd_1.0.5                   </span></span>
<span><span class="co">## [165] broom_1.0.4                   xtable_1.8-4                 </span></span>
<span><span class="co">## [167] later_1.3.1                   viridisLite_0.4.2            </span></span>
<span><span class="co">## [169] ragg_1.2.5                    truncnorm_1.0-9              </span></span>
<span><span class="co">## [171] tibble_3.2.1                  lmerTest_3.1-3               </span></span>
<span><span class="co">## [173] glmmTMB_1.1.7                 memoise_2.0.1                </span></span>
<span><span class="co">## [175] beeswarm_0.4.0                cluster_2.1.4                </span></span>
<span><span class="co">## [177] globals_0.16.2                GSEABase_1.62.0</span></span></code></pre>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
