<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Dreamlet analysis of single cell RNA-seq • dreamlet</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Dreamlet analysis of single cell RNA-seq">
<meta property="og:description" content="dreamlet">
<meta property="og:image" content="http://gabrielhoffman.github.io/dreamlet/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dreamlet</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.34</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/dreamlet.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/h5ad_on_disk.html">Loading large-scale H5AD datasets</a>
    </li>
    <li>
      <a href="../articles/mashr.html">mashr analysis following dreamlet</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/GabrielHoffman/dreamlet/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Dreamlet analysis of single cell RNA-seq</h1>
            <h3 data-toc-skip class="subtitle">Linear (mixed) model analysis of pseudobulk data</h3>
                        <h4 data-toc-skip class="author">Developed by <a href="http://gabrielhoffman.github.io/" class="external-link">Gabriel Hoffman</a>
</h4>
            
            <h4 data-toc-skip class="date">Run on 2022-03-02 15:26:22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/GabrielHoffman/dreamlet/blob/HEAD/vignettes/dreamlet.Rmd" class="external-link"><code>vignettes/dreamlet.Rmd</code></a></small>
      <div class="hidden name"><code>dreamlet.Rmd</code></div>

    </div>

    
    
<!---

cd /Users/gabrielhoffman/workspace/repos/variancePartition/vignettes

# rm -rf dreamlet_cache/

rmarkdown::render("dreamlet.Rmd")


 devtools::reload("/Users/gabrielhoffman/workspace/repos/dreamlet")

devtools::reload("/Users/gabrielhoffman/workspace/repos/zenith")



--->
<p>As the scale of single cell/nucleus RNA-seq has increased, so has the complexity of study designs. Analysis of datasets with simple study designs can be performed using linear model as in the <a href="https://bioconductor.org/packages/muscat" class="external-link">muscat package</a>. Yet analysis of datsets with complex study designs such as repeated measures or many technical batches can benefit from linear mixed model analysis to model to correlation structure between samples. We previously developed <a href="https://doi.org/10.1093/bioinformatics/btaa687" class="external-link">dream</a> to apply linear mixed models to bulk RNA-seq data using a <a href="https://bioconductor.org/packages/release/bioc/html/limma.html" class="external-link">limma</a>-style workflow. Dreamlet extends the previous work of dream and <a href="https://www.nature.com/articles/s41467-020-19894-4" class="external-link">muscat</a> to apply linear mixed models to pseudobulk data. Dreamlet also supports linear models and facilitates application of 1) <a href="https://bioconductor.org/packages/variancePartition" class="external-link">variancePartition</a> to quantify the contribution of multiple variables to expression variation, and 2) <a href="https://github.com/GabrielHoffman/zenith" class="external-link">zenith</a> to perform gene set analysis on the differential expression signatures.</p>
<div class="section level2">
<h2 id="process-single-cell-count-data">Process single cell count data<a class="anchor" aria-label="anchor" href="#process-single-cell-count-data"></a>
</h2>
<p>Here we perform analysis of PBMCs from 8 individuals stimulated with interferon-β <a href="https://www.nature.com/articles/nbt.4042" class="external-link">Kang, et al, 2018, Nature Biotech</a>. This is a small dataset that does not have repeated measures or high dimensional batch effects, so the sophisticated features of dreamlet are not strictly necessary. But this gives us a oppurtunity to walk through a standard dreamlet workflow.</p>
<div class="section level3">
<h3 id="preprocess-data">Preprocess data<a class="anchor" aria-label="anchor" href="#preprocess-data"></a>
</h3>
<p>Here, single cell RNA-seq data is downloaded from <a href="https://bioconductor.org/packages/ExperimentHub/" class="external-link">ExperimentHub</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/GabrielHoffman/dreamlet" class="external-link">dreamlet</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HelenaLC/muscat" class="external-link">muscat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">zenith</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://haozhu233.github.io/kableExtra/" class="external-link">kableExtra</a></span><span class="op">)</span>

<span class="co"># Download data, specifying EH2259 for the Kang, et al study</span>
<span class="va">eh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH2259"</span><span class="op">]</span><span class="op">]</span>

<span class="co"># only keep singlet cells with sufficient reads</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">multiplets</span> <span class="op">==</span> <span class="st">'singlet'</span><span class="op">]</span>

<span class="co"># compute QC metrics</span>
<span class="va">qc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html" class="external-link">perCellQCMetrics</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>

<span class="co"># remove cells with few or many detected genes</span>
<span class="va">ol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span>metric <span class="op">=</span> <span class="va">qc</span><span class="op">$</span><span class="va">detected</span>, nmads <span class="op">=</span> <span class="fl">2</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="op">!</span><span class="va">ol</span><span class="op">]</span>

<span class="co"># compute normalized data</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">10</span>, <span class="op">]</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/librarySizeFactors.html" class="external-link">computeLibraryFactors</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>

<span class="co"># set variable indicating stimulated (stim) or control (ctrl)</span>
<span class="va">sce</span><span class="op">$</span><span class="va">StimStatus</span> <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">stim</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="aggregate-to-pseudobulk">Aggregate to pseudobulk<a class="anchor" aria-label="anchor" href="#aggregate-to-pseudobulk"></a>
</h3>
<p>Dreamlet, like muscat, performs analysis at the pseudobulk-level by summing reads across cells for a given sample and cell type. <code>aggregateToPseudoBulk</code> is substantially faster for large on-disk datasets than <a href="https://rdrr.io/bioc/muscat/man/aggregateData.html" class="external-link"><code>muscat::aggregateData</code></a>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Since 'ind' is the individual and 'StimStatus' is the stimulus status,</span>
<span class="co"># create unique identifier for each sample</span>
<span class="va">sce</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">StimStatus</span>, <span class="va">sce</span><span class="op">$</span><span class="va">ind</span><span class="op">)</span>

<span class="co"># Create pseudobulk data by specifying cluster_id and sample_id</span>
<span class="co"># Count data for each cell type is then stored in the `assay` field</span>
<span class="co"># assay: entry in assayNames(sce) storing raw counts</span>
<span class="co"># cluster_id: variable in colData(sce) indicating cell clusters</span>
<span class="co"># sample_id: variable in colData(sce) indicating sample id for aggregating cells</span>
<span class="va">pb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregateToPseudoBulk.html">aggregateToPseudoBulk</a></span><span class="op">(</span><span class="va">sce</span>,
    assay <span class="op">=</span> <span class="st">"counts"</span>,     
    cluster_id <span class="op">=</span> <span class="st">"cell"</span>,  
    sample_id <span class="op">=</span> <span class="st">"id"</span>,
    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># one 'assay' per cell type</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">pb</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "B cells"           "CD14+ Monocytes"   "CD4 T cells"      </span>
<span class="co">## [4] "CD8 T cells"       "Dendritic cells"   "FCGR3A+ Monocytes"</span>
<span class="co">## [7] "Megakaryocytes"    "NK cells"</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="voom-for-pseudobulk">Voom for pseudobulk<a class="anchor" aria-label="anchor" href="#voom-for-pseudobulk"></a>
</h2>
<p>Apply <a href="https://rdrr.io/bioc/limma/man/voom.html" class="external-link">voom</a>-style normalization for pseudobulk counts within each cell cluster using <a href="https://gabrielhoffman.github.io/variancePartition/reference/voomWithDreamWeights.html" class="external-link">voomWithDreamWeights</a> to handle random effects (if specified).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Normalize and apply voom/voomWithDreamWeights</span>
<span class="va">res.proc</span> <span class="op">=</span> <span class="fu"><a href="../reference/processAssays.html">processAssays</a></span><span class="op">(</span> <span class="va">pb</span>, <span class="op">~</span> <span class="va">StimStatus</span>, min.count<span class="op">=</span><span class="fl">5</span><span class="op">)</span>

<span class="co"># the resulting object of class dreamletProcessedData stores </span>
<span class="co"># normalized data and other information</span>
<span class="va">res.proc</span></code></pre></div>
<pre><code><span class="co">## class: dreamletProcessedData </span>
<span class="co">## assays(8): B cells CD14+ Monocytes ... Megakaryocytes NK cells</span>
<span class="co">## colData(4): ind stim multiplets StimStatus</span>
<span class="co">## metadata(0):</span>
<span class="co">## Samples:</span>
<span class="co">##  min: 5 </span>
<span class="co">##  max: 16</span>
<span class="co">## Genes:</span>
<span class="co">##  min: 180 </span>
<span class="co">##  max: 4455 </span>
<span class="co">## details(3): assay n_retained formula</span></code></pre>
<p><code><a href="../reference/processAssays.html">processAssays()</a></code> retains samples with at least <code>min.cells</code> in a given cell type. While dropping a few samples usually is not a problem, in some cases dropping sames can mean that a variable included in the regression formula no longer has any variation. For example, dropping all stimulated samples from analysis of a given cell type would be mean the variable <code>StimStatus</code> has no variation and is perfectly colinear with the intercept term. This colinearity issue is detected internally and variables with these problem are dropped from the regression formula for that particular cell type. The number of samples retained and the resulting formula used in each cell type can be accessed as follows. In this analysis, samples are dropped from 3 cell types but the original formula remains valid in each case.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># view details of dropping samples</span>
<span class="fu"><a href="../reference/details-methods.html">details</a></span><span class="op">(</span><span class="va">res.proc</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 8 rows and 3 columns</span>
<span class="co">##               assay n_retained     formula</span>
<span class="co">##         &lt;character&gt;  &lt;integer&gt; &lt;character&gt;</span>
<span class="co">## 1           B cells         16 ~StimStatus</span>
<span class="co">## 2   CD14+ Monocytes         16 ~StimStatus</span>
<span class="co">## 3       CD4 T cells         16 ~StimStatus</span>
<span class="co">## 4       CD8 T cells         16 ~StimStatus</span>
<span class="co">## 5   Dendritic cells          5 ~StimStatus</span>
<span class="co">## 6 FCGR3A+ Monocytes         16 ~StimStatus</span>
<span class="co">## 7    Megakaryocytes          9 ~StimStatus</span>
<span class="co">## 8          NK cells         16 ~StimStatus</span></code></pre>
<p>Here the mean-variance trend from voom is shown for each cell type. Cell types with sufficient number of cells and reads show a clear mean-variance trend. While in rare cell types like megakaryocytes, fewer genes have sufficient reads and the trend is less apparent.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># show voom plot for each cell clusters  </span>
<span class="fu"><a href="../reference/plotVoom-methods.html">plotVoom</a></span><span class="op">(</span> <span class="va">res.proc</span><span class="op">)</span></code></pre></div>
<p><img src="dreamlet_files/figure-html/voom.plots-1.png" width="700"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Show plots for subset of cell clusters</span>
<span class="co"># plotVoom( res.proc[1:3] )</span>

<span class="co"># Show plots for one cell cluster</span>
<span class="co"># plotVoom( res.proc[["B cells"]])</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="variance-partitioning">Variance partitioning<a class="anchor" aria-label="anchor" href="#variance-partitioning"></a>
</h2>
<p>The <a href="http://bioconductor.org/packages/variancePartition/" class="external-link">variancePartition</a> package uses linear and linear mixed models to quanify the contribution of multiple sources of expression variation at the gene-level. For each gene it fits a linear (mixed) model and evalutes the fraction of expression variation explained by each variable.</p>
<p>Variance fractions can be visualized at the gene-level for each cell type using a bar plot, or genome-wide using a violin plot.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># run variance partitioning analysis</span>
<span class="va">vp.lst</span> <span class="op">=</span> <span class="fu"><a href="../reference/fitVarPart.html">fitVarPart</a></span><span class="op">(</span> <span class="va">res.proc</span>, <span class="op">~</span> <span class="va">StimStatus</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Show variance fractions at the gene-level for each cell type</span>
<span class="va">genes</span> <span class="op">=</span> <span class="va">vp.lst</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>
<span class="fu">plotPercentBars</span><span class="op">(</span><span class="va">vp.lst</span><span class="op">[</span><span class="va">vp.lst</span><span class="op">$</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">genes</span>,<span class="op">]</span><span class="op">)</span></code></pre></div>
<p><img src="dreamlet_files/figure-html/vp.barplt-1.png" width="700"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Summarize variance fractions genome-wide for each cell type</span>
<span class="fu">plotVarPart</span><span class="op">(</span><span class="va">vp.lst</span>, label.angle<span class="op">=</span><span class="fl">60</span><span class="op">)</span></code></pre></div>
<p><img src="dreamlet_files/figure-html/vp.violin-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="differential-expression">Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h2>
<p>Since the normalized expression data and metadata are stored within <code>res.proc</code>, only the regression formula remains to be specified. Here we only included the stimulus status, but analyses of larger datasets can include covariates and random effects. With formula <code>~ StimStatus</code>, an intercept is fit and coefficient <code>StimStatusstim</code> log fold change between simulated and controls.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Differential expression analysis within each assay,</span>
<span class="co"># evaluated on the voom normalized data </span>
<span class="va">res.dl</span> <span class="op">=</span> <span class="fu"><a href="../reference/dreamlet.html">dreamlet</a></span><span class="op">(</span> <span class="va">res.proc</span>, <span class="op">~</span> <span class="va">StimStatus</span><span class="op">)</span>

<span class="co"># names of estimated coefficients</span>
<span class="fu"><a href="../reference/coefNames-methods.html">coefNames</a></span><span class="op">(</span><span class="va">res.dl</span><span class="op">)</span> </code></pre></div>
<pre><code><span class="co">## [1] "(Intercept)"    "StimStatusstim"</span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># the resulting object of class dreamletResult</span>
<span class="co"># stores results and other information</span>
<span class="va">res.dl</span></code></pre></div>
<pre><code><span class="co">## class: dreamletResult </span>
<span class="co">## assays(8): B cells CD14+ Monocytes ... Megakaryocytes NK cells</span>
<span class="co">## Genes:</span>
<span class="co">##  min: 180 </span>
<span class="co">##  max: 4455 </span>
<span class="co">## details(4): assay n_retain formula formDropsTerms</span>
<span class="co">## coefNames(2): (Intercept) StimStatusstim</span></code></pre>
<div class="section level3">
<h3 id="volcano-plots">Volcano plots<a class="anchor" aria-label="anchor" href="#volcano-plots"></a>
</h3>
<p>The volcano plot can indicate the strength of the differential expression signal with each cell type. Red points indicate FDR &lt; 0.05.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotVolcano-methods.html">plotVolcano</a></span><span class="op">(</span> <span class="va">res.dl</span>, coef <span class="op">=</span> <span class="st">'StimStatusstim'</span> <span class="op">)</span></code></pre></div>
<p><img src="dreamlet_files/figure-html/plotVolcano-1.png" width="768"></p>
</div>
<div class="section level3">
<h3 id="extract-results">Extract results<a class="anchor" aria-label="anchor" href="#extract-results"></a>
</h3>
<p>Each entry in <code>res.dl</code> stores a model fit by <a href="https://rdrr.io/bioc/variancePartition/man/dream-method.html" class="external-link"><code>dream()</code></a>, and results can be extracted using <code>topTable()</code> as in <code>limma</code> by specifying the coefficient of interest. The results shows the gene name, log fold change, average expression, t-statistic, p-value, FDR (i.e. adj.P.Val).</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">topTable</span><span class="op">(</span><span class="va">res.dl</span><span class="op">[[</span><span class="st">"B cells"</span><span class="op">]</span><span class="op">]</span>, coef <span class="op">=</span> <span class="st">'StimStatusstim'</span> <span class="op">)</span></code></pre></div>
<pre><code><span class="co">##           logFC   AveExpr        t      P.Value    adj.P.Val        B</span>
<span class="co">## ISG15  5.526203 10.240731 29.94809 3.942031e-21 5.704118e-18 38.37168</span>
<span class="co">## ISG20  3.317362 11.421612 28.30217 1.561277e-20 1.129584e-17 37.05904</span>
<span class="co">## LY6E   4.290878  8.744657 21.01998 2.000133e-17 7.584859e-15 29.93200</span>
<span class="co">## UBE2L6 2.719712  9.279037 20.81040 2.537188e-17 7.584859e-15 29.69832</span>
<span class="co">## IRF7   3.726571  8.716599 20.78195 2.620891e-17 7.584859e-15 29.66924</span>
<span class="co">## EPSTI1 3.696210  8.145939 20.57331 3.329155e-17 8.028813e-15 29.41155</span>
<span class="co">## PLSCR1 4.066102  8.309200 19.70214 9.257824e-17 1.913724e-14 28.40669</span>
<span class="co">## DYNLT1 3.101261  7.912137 18.12641 6.543225e-16 1.080168e-13 26.47151</span>
<span class="co">## IFITM2 2.963947  8.769575 18.10590 6.718392e-16 1.080168e-13 26.43400</span>
<span class="co">## SAT1   2.173411  9.695187 17.40672 1.679515e-15 2.430258e-13 25.43741</span></code></pre>
</div>
<div class="section level3">
<h3 id="forest-plot">Forest plot<a class="anchor" aria-label="anchor" href="#forest-plot"></a>
</h3>
<p>A forest plot shows the log fold change and standard error of a given gene across all cell types. The color indicates the FDR.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotForest-methods.html">plotForest</a></span><span class="op">(</span> <span class="va">res.dl</span>, coef <span class="op">=</span> <span class="st">'StimStatusstim'</span>, gene <span class="op">=</span> <span class="st">'ISG20'</span><span class="op">)</span></code></pre></div>
<p><img src="dreamlet_files/figure-html/forrest-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="advanced-used-of-contrasts">Advanced used of contrasts<a class="anchor" aria-label="anchor" href="#advanced-used-of-contrasts"></a>
</h3>
<p>A hypothesis test of the <em>difference</em> between two or more coefficients can be performed using contrasts. The contrast matrix is evaluated for each cell type in the backend, so only the contrast string must be supplied to <code><a href="../reference/dreamlet.html">dreamlet()</a></code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create a contrasts called 'Diff' that is the difference between expression</span>
<span class="co"># in the stimulated and controls.</span>
<span class="co"># More than one can be specified</span>
<span class="va">contrasts</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Diff <span class="op">=</span> <span class="st">'StimStatusstim - StimStatusctrl'</span><span class="op">)</span>

<span class="co"># Evalaute the regression model without an intercept term.</span>
<span class="co"># Instead estimate the mean expression in stimulated, controls and then</span>
<span class="co"># set Diff to the difference between the two</span>
<span class="va">res.dl2</span> <span class="op">=</span> <span class="fu"><a href="../reference/dreamlet.html">dreamlet</a></span><span class="op">(</span> <span class="va">res.proc</span>, <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">StimStatus</span>, contrasts<span class="op">=</span><span class="va">contrasts</span><span class="op">)</span>

<span class="co"># see estimated coefficients</span>
<span class="fu"><a href="../reference/coefNames-methods.html">coefNames</a></span><span class="op">(</span><span class="va">res.dl2</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "Diff"           "StimStatusctrl" "StimStatusstim"</span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Volcano plot of Diff</span>
<span class="fu"><a href="../reference/plotVolcano-methods.html">plotVolcano</a></span><span class="op">(</span> <span class="va">res.dl2</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, coef <span class="op">=</span> <span class="st">'Diff'</span> <span class="op">)</span></code></pre></div>
<p><img src="dreamlet_files/figure-html/dreamlet.contrasts-1.png" width="700"></p>
<p>This new <code>Diff</code> variable can then be used downstream for analysis asking for a coefficient. But note that since there is no intercept term in this model, the meaning of <code>StimStatusstim</code> changes here. When the formula is <code>0 + StimStatus</code> then <code>StimStatusstim</code> is the mean expression in stimulated samples.</p>
<p>For further information about using contrasts see <a href="https://gabrielhoffman.github.io/variancePartition/reference/makeContrastsDream.html" class="external-link">makeContrastsDream()</a> and <a href="https://gabrielhoffman.github.io/variancePartition/articles/dream.html#advanced-hypothesis-testing-1" class="external-link">vignette</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="gene-set-analysis">Gene set analysis<a class="anchor" aria-label="anchor" href="#gene-set-analysis"></a>
</h2>
<p>While standard enrichment methods like Fishers exact test, requires specifying a FDR cutoff to identify differentially expressed genes. However, dichotomizing differential expression results is often too simple and ignores the quantitative variation captured by the differential expression test statistics. Here we use <code>zenith</code>, a wrapper for <a href="https://rdrr.io/bioc/limma/man/camera.html" class="external-link"><code>limma::camera</code></a>, to perform gene set analysis using the full spectrum of differential expression test statistics. <code>zenith/camera</code> is a <a href="https://www.nature.com/articles/nrg.2016.29" class="external-link">competetive test</a> that compares the mean test statistic for genes in a given gene set, to genes not in that set while accounting for correlation between genes.</p>
<p>Here, <code>zenith_gsa</code> takes a <code>dreamletResult</code> object, the coefficient of interest, and gene sets as a <code>GeneSetCollection</code> object from <a href="https://bioconductor.org/packages/GSEABase/" class="external-link">GSEABase</a>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load Gene Ontology database </span>
<span class="co"># use gene 'SYMBOL', or 'ENSEMBL' id</span>
<span class="co"># use get_MSigDB() to load MSigDB</span>
<span class="va">go.gs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/zenith/man/get_GeneOntology.html" class="external-link">get_GeneOntology</a></span><span class="op">(</span>to<span class="op">=</span><span class="st">"SYMBOL"</span><span class="op">)</span>
   
<span class="co"># Run zenith gene set analysis on result of dreamlet</span>
<span class="va">res_zenith</span> <span class="op">=</span> <span class="fu"><a href="../reference/zenith_gsa-methods.html">zenith_gsa</a></span><span class="op">(</span><span class="va">res.dl</span>, coef <span class="op">=</span> <span class="st">'StimStatusstim'</span>, <span class="va">go.gs</span><span class="op">)</span>

<span class="co"># examine results for each ell type and gene set</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res_zenith</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##     assay           coef                                                Geneset</span>
<span class="co">## 1 B cells StimStatusstim               GO0034340: response to type I interferon</span>
<span class="co">## 2 B cells StimStatusstim         GO0060337: type I interferon signaling pathway</span>
<span class="co">## 3 B cells StimStatusstim      GO0071357: cellular response to type I interferon</span>
<span class="co">## 4 B cells StimStatusstim        GO0048525: negative regulation of viral process</span>
<span class="co">## 5 B cells StimStatusstim                   GO0051607: defense response to virus</span>
<span class="co">## 6 B cells StimStatusstim GO0060333: interferon-gamma-mediated signaling pathway</span>
<span class="co">##   NGenes Correlation    delta        se    p.less    p.greater       PValue</span>
<span class="co">## 1     17        0.01 4.586323 0.6352284 0.9999978 2.212697e-06 4.425394e-06</span>
<span class="co">## 2     17        0.01 4.586323 0.6352284 0.9999978 2.212697e-06 4.425394e-06</span>
<span class="co">## 3     17        0.01 4.586323 0.6352284 0.9999978 2.212697e-06 4.425394e-06</span>
<span class="co">## 4     13        0.01 4.747930 0.7156646 0.9999944 5.622816e-06 1.124563e-05</span>
<span class="co">## 5     24        0.01 3.479750 0.5536610 0.9999900 1.002634e-05 2.005268e-05</span>
<span class="co">## 6     19        0.01 3.333906 0.6115057 0.9999574 4.263035e-05 8.526071e-05</span>
<span class="co">##   Direction         FDR</span>
<span class="co">## 1        Up 0.002898844</span>
<span class="co">## 2        Up 0.002898844</span>
<span class="co">## 3        Up 0.002898844</span>
<span class="co">## 4        Up 0.004890780</span>
<span class="co">## 5        Up 0.007662352</span>
<span class="co">## 6        Up 0.020576251</span></code></pre>
<div class="section level3">
<h3 id="heatmap-of-top-genesets">Heatmap of top genesets<a class="anchor" aria-label="anchor" href="#heatmap-of-top-genesets"></a>
</h3>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># for each cell type select 5 genesets with largest t-statistic</span>
<span class="co"># and 1 geneset with the lowest</span>
<span class="co"># Grey boxes indicate the gene set could not be evaluted because</span>
<span class="co">#    to few genes were represented</span>
<span class="fu"><a href="../reference/plotZenithResults.html">plotZenithResults</a></span><span class="op">(</span><span class="va">res_zenith</span>, <span class="fl">5</span>, <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p><img src="dreamlet_files/figure-html/heatmap-1.png" width="480"></p>
<div class="section level4">
<h4 id="all-gene-sets-with-fdr-5">All gene sets with FDR &lt; 5%<a class="anchor" aria-label="anchor" href="#all-gene-sets-with-fdr-5"></a>
</h4>
<p>Here, show all genes with FDR &lt; 5% in any cell type</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get genesets with FDR &lt; 5%</span>
<span class="va">gs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">res_zenith</span><span class="op">$</span><span class="va">Geneset</span><span class="op">[</span><span class="va">res_zenith</span><span class="op">$</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">]</span><span class="op">)</span>

<span class="co"># keep only results of these genesets</span>
<span class="va">df</span> <span class="op">=</span> <span class="va">res_zenith</span><span class="op">[</span><span class="va">res_zenith</span><span class="op">$</span><span class="va">Geneset</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">gs</span>,<span class="op">]</span>

<span class="co"># plot results, but with no limit based on the highest/lowest t-statistic</span>
<span class="fu"><a href="../reference/plotZenithResults.html">plotZenithResults</a></span><span class="op">(</span><span class="va">df</span>, <span class="cn">Inf</span>, <span class="cn">Inf</span><span class="op">)</span></code></pre></div>
<p><img src="dreamlet_files/figure-html/heatmap2-1.png" width="480"></p>
<!---
# Variation in cell counts
The number of cells observed in of each type varies across samples.  Visualizing these difference and evaluating weither these differences are explained by other measured variables can give insight into the biology of this variation in cell type composition.    
Here, compare fraction of each cell type bewteen simulated and unstimulated samples


```r
plotCellComposition(pb)     
```

<img src="dreamlet_files/figure-html/plotCellComposition-1.png" width="480" />

Use [crumblr](https://gabrielhoffman.github.io/crumblr/reference/crumblr.html) to transform the cell counts for each sample into log count ratios and precision weights for analysis with dream.

```r
library(crumblr)

# get transformed counts and precision weights with crumblr
cobj = crumblr( cellCounts(pb) )

# perform standard dream analysis
fit = dream(cobj, ~ StimStatus, as.data.frame(colData(pb)))
fit = eBayes(fit)

topTable(fit, coef='StimStatusstim', sort.by="none")
```

```
##                         logFC    AveExpr          t   P.Value adj.P.Val
## B cells           -0.09882673  0.5516882 -0.6989167 0.4943409 0.7706781
## CD14+ Monocytes   -0.12323960  1.2698117 -0.9641347 0.3489000 0.7706781
## CD4 T cells       -0.08092309  2.0201947 -0.3882013 0.7028311 0.7706781
## CD8 T cells       -0.14896362  0.0857175 -0.3589998 0.7241489 0.7706781
## Dendritic cells    0.32314289 -2.1849234  1.1280014 0.2754583 0.7706781
## FCGR3A+ Monocytes  0.07153536 -0.2567492  0.3637101 0.7206938 0.7706781
## Megakaryocytes     0.06001827 -1.8655172  0.2963275 0.7706781 0.7706781
## NK cells           0.08379008  0.3797777  0.5846539 0.5666914 0.7706781
##                           B
## B cells           -4.742373
## CD14+ Monocytes   -4.683602
## CD4 T cells       -4.923287
## CD8 T cells       -4.764307
## Dendritic cells   -4.588102
## FCGR3A+ Monocytes -4.734407
## Megakaryocytes    -4.634451
## NK cells          -4.754474
```

Variance partioning analysis quantifies the contribution of multiple sources of variation in cell type composition differences. 

```r
vp = fitExtractVarPartModel(cobj, ~ StimStatus, as.data.frame(colData(pb)))

plotPercentBars(vp)
```

<img src="dreamlet_files/figure-html/ctc.crumblr.cp-1.png" width="700" />
--->
</div>
</div>
</div>
<div class="section level2">
<h2 id="comparing-expression-in-clusters">Comparing expression in clusters<a class="anchor" aria-label="anchor" href="#comparing-expression-in-clusters"></a>
</h2>
<p>Identifying genes that are differentially expressed between cell clusters incorporates a paired analysis design, since each individual is observed for each cell cluster.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># test differential expression between B cells and the rest of the cell clusters</span>
<span class="va">ct.pairs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B cells"</span>, <span class="st">"rest"</span><span class="op">)</span>
 
<span class="va">fit</span> <span class="op">=</span> <span class="fu"><a href="../reference/dreamletCompareClusters.html">dreamletCompareClusters</a></span><span class="op">(</span> <span class="va">pb</span>, <span class="va">ct.pairs</span>, method<span class="op">=</span><span class="st">"fixed"</span><span class="op">)</span>

<span class="co"># The coefficient 'compare' is the value logFC between test and baseline:</span>
<span class="co"># compare = cellClustertest - cellClusterbaseline </span>
<span class="va">df_Bcell</span> <span class="op">=</span> <span class="fu">topTable</span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="st">"compare"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">df_Bcell</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##              logFC   AveExpr         t      P.Value    adj.P.Val        B</span>
<span class="co">## C15orf48 -6.635081  8.200179 -52.73567 7.166069e-26 1.949171e-22 48.48023</span>
<span class="co">## TIMP1    -6.664147  9.326366 -48.94376 4.060244e-25 5.521932e-22 47.17369</span>
<span class="co">## ANXA1    -5.596470  8.477805 -36.83076 2.936001e-22 2.661975e-19 40.90691</span>
<span class="co">## CCL2     -6.532150  8.564950 -36.29678 4.111548e-22 2.795853e-19 40.60157</span>
<span class="co">## FTL      -3.434776 13.363302 -33.18651 3.232134e-21 1.758281e-18 38.66849</span>
<span class="co">## S100A10  -4.040431  8.873840 -27.35835 2.680319e-19 1.215078e-16 34.29294</span></code></pre>
</div>
<div class="section level2">
<h2 id="gene-cluster-specificity">Gene-cluster specificity<a class="anchor" aria-label="anchor" href="#gene-cluster-specificity"></a>
</h2>
<p>Evaluate the specificity of each gene for each cluster:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df_cts</span> <span class="op">=</span> <span class="fu"><a href="../reference/cellTypeSpecificity.html">cellTypeSpecificity</a></span><span class="op">(</span> <span class="va">pb</span> <span class="op">)</span>

<span class="co"># Violin plot of specificity score for each cell type</span>
<span class="fu"><a href="../reference/plotViolin-methods.html">plotViolin</a></span><span class="op">(</span><span class="va">df_cts</span><span class="op">)</span></code></pre></div>
<p><img src="dreamlet_files/figure-html/cellTypeSpecificity-1.png" width="700"></p>
<p>Highlight expression fraction for most specific gene from each cell type:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">df_cts</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">df_cts</span>, <span class="fl">2</span>, <span class="va">which.max</span><span class="op">)</span><span class="op">]</span>
<span class="fu">plotPercentBars</span><span class="op">(</span> <span class="va">df_cts</span>, genes <span class="op">=</span> <span class="va">genes</span> <span class="op">)</span></code></pre></div>
<p><img src="dreamlet_files/figure-html/plotPercentBars-1.png" width="700"></p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">dreamlet</span><span class="fu">::</span><span class="fu"><a href="../reference/plotHeatmap-methods.html">plotHeatmap</a></span><span class="op">(</span> <span class="va">df_cts</span>, genes <span class="op">=</span> <span class="va">genes</span><span class="op">)</span></code></pre></div>
<p><img src="dreamlet_files/figure-html/plotPercentBars-2.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'dedcc260e9d3743bc634e728855cdbaf',
    indexName: 'dreamlet',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
