<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Loading large-scale H5AD datasets • dreamlet</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Loading large-scale H5AD datasets">
<meta property="og:description" content="dreamlet">
<meta property="og:image" content="http://gabrielhoffman.github.io/dreamlet/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dreamlet</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.16</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/dreamlet.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/h5ad_on_disk.html">Loading large-scale H5AD datasets</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/GabrielHoffman/dreamlet/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="h5ad_on_disk_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Loading large-scale H5AD datasets</h1>
            <h3 class="subtitle">On-disk SingleCellExperiment</h3>
                        <h4 class="author">Developed by <a href="http://gabrielhoffman.github.io/">Gabriel Hoffman</a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/GabrielHoffman/dreamlet/blob/master/vignettes/h5ad_on_disk.Rmd"><code>vignettes/h5ad_on_disk.Rmd</code></a></small>
      <div class="hidden name"><code>h5ad_on_disk.Rmd</code></div>

    </div>

    
    
<!---

cd /Users/gabrielhoffman/workspace/repos/variancePartition/vignettes

# rm -rf dreamlet_cache/

rmarkdown::render("h5ad_on_disk.Rmd")


--->
<p>As the scale of single cell RNA-seq datasets continues to increase, loading the entire dataset into memory if often impractical. Dreamlet can take advantage of on-disk data access to dramatically reduce memory usage.</p>
<p>Storing 18K genes across 2M cells with double precision would require 288 Gb memory. Even if the dataset is sparse so that 80% of the entires are zero, loading the entire dataset requires 58 Gb of data. Since <code>R</code> uses copy-by-value, data duplication quickly becomes a problem even for a high memory machine.</p>
<p>Instead of loading the entire dataset into memory, we adopt an on-disk approach that takes advantage of the <a href="https://anndata.readthedocs.io/en/latest/fileformat-prose.html">H5AD</a> file format built on the <a href="https://en.wikipedia.org/wiki/Hierarchical_Data_Format">HDF5</a> format in order to dramatically reduce memory usage while still retaining high performance. The <a href="https://github.com/theislab/zellkonverter">zellkonverter</a> package uses a <a href="https://bioconductor.org/packages/DelayedArray/">DelayedArray</a> backend to provide a seamless interface to an on-disk H5AD dataset through the interface of the <a href="https://bioconductor.org/packages/SingleCellExperiment/">SingleCellExperiment</a> class. SingleCellExperiment is <a href="https://bioconductor.org">Bioconductor’s</a> core data class for single cell data, and data from other formats (i.e. <a href="https://cran.r-project.org/web/packages/Seurat/index.html">Seurat</a>) can easily be <a href="https://rdrr.io/cran/Seurat/man/as.SingleCellExperiment.html">converted</a> to it. Importantly, <code>R</code>’s copy-by-value is no longer an issue because the data is not duplicated when it is passed to another function, since it remains on-disk.</p>
<p>This enables any analysis designed for the SingleCellExperiment class to take advantage of seamless on-disk access to large-scale datasets.</p>
<div id="using-zellkonverter" class="section level2">
<h2 class="hasAnchor">
<a href="#using-zellkonverter" class="anchor"></a>Using zellkonverter</h2>
<p><code>zellkonverter</code> &gt;=1.3.0 is required in general and &gt;=1.3.3 is required for the last example below. <code>zellkonverter</code> is an interface for python code. Note that the python dependencies are installed the first time <code><a href="https://rdrr.io/pkg/zellkonverter/man/readH5AD.html">readH5AD()</a></code> is run, rather than when the package is installed. So the first run of <code><a href="https://rdrr.io/pkg/zellkonverter/man/readH5AD.html">readH5AD()</a></code> can take a few minutes, but subsequent runs will be fast.</p>
<div id="standard-usage" class="section level3">
<h3 class="hasAnchor">
<a href="#standard-usage" class="anchor"></a>Standard usage</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/theislab/zellkonverter">zellkonverter</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span>

<span class="co"># Create SingleCellExperiment object that points to on-disk H5AD file</span>
<span class="va">sce</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/zellkonverter/man/readH5AD.html">readH5AD</a></span><span class="op">(</span><span class="va">h5ad_file</span>, use_hdf5<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div id="merge-multiple-datasets" class="section level3">
<h3 class="hasAnchor">
<a href="#merge-multiple-datasets" class="anchor"></a>Merge multiple datasets</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Read a series of H5AD files into a list</span>
<span class="co"># then combine them into a single merged SingleCellExperiment</span>
<span class="va">sce.lst</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span> <span class="va">h5ad_files</span>, <span class="kw">function</span><span class="op">(</span><span class="va">file</span><span class="op">)</span><span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/pkg/zellkonverter/man/readH5AD.html">readH5AD</a></span><span class="op">(</span><span class="va">file</span>, use_hdf5<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
  <span class="op">}</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">cbind</span>, <span class="va">sce.list</span><span class="op">)</span></code></pre></div>
</div>
<div id="access-alternative-data" class="section level3">
<h3 class="hasAnchor">
<a href="#access-alternative-data" class="anchor"></a>Access alternative data</h3>
<p>Some software, including <a href="https://pegasus.readthedocs.io/en/stable/">pegasus</a>, store normalized data in the default portion of the file and save raw counts in the <code>raw/</code> field. Here, load a H5AD file generated by pegasus.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># read raw/ from H5AD file</span>
<span class="co"># raw = TRUE tells readH5AD() to read alternative data</span>
<span class="co"># Must use zellkonverter &gt;=1.3.3</span>
<span class="va">sce_in</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/zellkonverter/man/readH5AD.html">readH5AD</a></span><span class="op">(</span><span class="va">h5ad_file</span>, use_hdf5<span class="op">=</span><span class="cn">TRUE</span>, raw<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># extract raw counts</span>
<span class="co"># include Dimnames and colData</span>
<span class="va">sce</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html">altExp</a></span><span class="op">(</span> <span class="va">sce_in</span>, <span class="st">"raw"</span>, withDimnames<span class="op">=</span><span class="cn">TRUE</span>, withColData<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># convert these alternative data to SingleCellExperiment</span>
<span class="va">sce</span> <span class="op">=</span> <span class="fu">as</span><span class="op">(</span><span class="va">sce</span>, <span class="st">"SingleCellExperiment"</span><span class="op">)</span>

<span class="co"># Copy over reducedDims too</span>
<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDims</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html">reducedDims</a></span><span class="op">(</span><span class="va">sce_in</span><span class="op">)</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="http://gabrielhoffman.github.io">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
