<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Loading large-scale H5AD datasets â€¢ dreamlet</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Loading large-scale H5AD datasets">
<meta property="og:description" content="dreamlet">
<meta property="og:image" content="http://gabrielhoffman.github.io/dreamlet/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dreamlet</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.31</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/dreamlet.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/h5ad_on_disk.html">Loading large-scale H5AD datasets</a>
    </li>
    <li>
      <a href="../articles/mashr.html">mashr analysis following dreamlet</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/GabrielHoffman/dreamlet/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Loading large-scale H5AD datasets</h1>
            <h3 data-toc-skip class="subtitle">On-disk SingleCellExperiment</h3>
                        <h4 data-toc-skip class="author">Developed by <a href="http://gabrielhoffman.github.io/" class="external-link">Gabriel Hoffman</a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/GabrielHoffman/dreamlet/blob/HEAD/vignettes/h5ad_on_disk.Rmd" class="external-link"><code>vignettes/h5ad_on_disk.Rmd</code></a></small>
      <div class="hidden name"><code>h5ad_on_disk.Rmd</code></div>

    </div>

    
    
<!---

cd /Users/gabrielhoffman/workspace/repos/variancePartition/vignettes

# rm -rf dreamlet_cache/

rmarkdown::render("h5ad_on_disk.Rmd")


--->
<p>As the scale of single cell RNA-seq datasets continues to increase, loading the entire dataset into memory if often impractical. Dreamlet can take advantage of efficient data storage through the <a href="https://bioconductor.org" class="external-link">Bioconductors</a> <a href="https://bioconductor.org/packages/SingleCellExperiment/" class="external-link">SingleCellExperiment</a> interface to dramatically reduce memory usage. There are two general methods for storing large single-cell datasets. <a href="https://bioconductor.org/packages/SingleCellExperiment/" class="external-link">SingleCellExperiment</a>, and therefore <code>dreamlet</code>, are compatible with both:</p>
<ol style="list-style-type: decimal">
<li><p><strong>On-disk storage</strong>: The data remains on the physical disk and is only read into memory when requested. In R there is little data duplication in memory since operations are delayed until the result is requested using <code>DelayedMatrix</code>. Memory usage is minimized at the cost of performance since disk access can be slow.</p></li>
<li><p><strong>In-memory storage</strong>: Single cell read counts are generally sprase with, say, 20% of read counts being non-zero. The counts can be stored as a <code>sparseMatrix</code> that stores only non-zero entries. Performance is higher since data is in memory, but excessive memory usage can be a problem.</p></li>
</ol>
<!---
on-disk data access to dramatically reduce memory usage.



Storing 18K genes across 2M cells with double precision would require 288 Gb memory.  Even if the dataset is sparse so that 80\% of the entires are zero, loading the entire dataset requires 58 Gb of data.  Since `R` uses copy-by-value, data duplication quickly becomes a problem even for a high memory machine.   

Instead of loading the entire dataset into memory, we adopt an on-disk approach that takes advantage of the [H5AD](https://anndata.readthedocs.io/en/latest/fileformat-prose.html) file format built on the [HDF5](https://en.wikipedia.org/wiki/Hierarchical_Data_Format) format in order to dramatically reduce memory usage while still retaining high performance.  The [zellkonverter](https://github.com/theislab/zellkonverter) package uses a [DelayedArray](https://bioconductor.org/packages/DelayedArray/) backend to provide a seamless interface to an on-disk H5AD dataset through the interface of the [SingleCellExperiment](https://bioconductor.org/packages/SingleCellExperiment/) class.  SingleCellExperiment is [Bioconductor's](https://bioconductor.org) core data class for single cell data, and data from other formats (i.e. [Seurat](https://cran.r-project.org/web/packages/Seurat/index.html)) can easily be [converted](https://rdrr.io/cran/Seurat/man/as.SingleCellExperiment.html) to it.  Importantly,  `R`'s copy-by-value is no longer an issue because the data is not duplicated when it is passed to another function, since it remains on-disk.      

This enables any analysis designed for the [SingleCellExperiment](https://bioconductor.org/packages/SingleCellExperiment/) class to take advantage of seamless on-disk access to large-scale datasets.

---><div class="section level2">
<h2 id="on-disk-storage-using-zellkonverter">On-disk storage using zellkonverter<a class="anchor" aria-label="anchor" href="#on-disk-storage-using-zellkonverter"></a>
</h2>
<p><a href="https://github.com/theislab/zellkonverter" class="external-link">zellkonverter</a> takes advantage of the <a href="https://anndata.readthedocs.io/en/latest/fileformat-prose.html" class="external-link">H5AD</a> file format built on the <a href="https://en.wikipedia.org/wiki/Hierarchical_Data_Format" class="external-link">HDF5</a> format in order to dramatically reduce memory usage while still retaining performance. The <a href="https://github.com/theislab/zellkonverter" class="external-link">zellkonverter</a> package uses a <a href="https://bioconductor.org/packages/DelayedArray/" class="external-link">DelayedArray</a> backend to provide a seamless interface to an on-disk H5AD dataset through the interface of the <a href="https://bioconductor.org/packages/SingleCellExperiment/" class="external-link">SingleCellExperiment</a> class.</p>
<p><code>zellkonverter</code> &gt;=1.3.0 is required in general and &gt;=1.3.3 is required for the last example below. <code>zellkonverter</code> is an interface for python code. Note that the python dependencies are installed the first time <code><a href="https://rdrr.io/pkg/zellkonverter/man/readH5AD.html" class="external-link">readH5AD()</a></code> is run, rather than when the package is installed. So the first run of <code><a href="https://rdrr.io/pkg/zellkonverter/man/readH5AD.html" class="external-link">readH5AD()</a></code> can take a few minutes, but subsequent runs will be fast.</p>
<div class="section level3">
<h3 id="standard-usage">Standard usage<a class="anchor" aria-label="anchor" href="#standard-usage"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/theislab/zellkonverter" class="external-link">zellkonverter</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span>

<span class="co"># Create SingleCellExperiment object that points to on-disk H5AD file</span>
<span class="va">sce</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/zellkonverter/man/readH5AD.html" class="external-link">readH5AD</a></span><span class="op">(</span><span class="va">h5ad_file</span>, use_hdf5<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="merge-multiple-datasets">Merge multiple datasets<a class="anchor" aria-label="anchor" href="#merge-multiple-datasets"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Read a series of H5AD files into a list</span>
<span class="co"># then combine them into a single merged SingleCellExperiment</span>
<span class="va">sce.lst</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span> <span class="va">h5ad_files</span>, <span class="kw">function</span><span class="op">(</span><span class="va">file</span><span class="op">)</span><span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/pkg/zellkonverter/man/readH5AD.html" class="external-link">readH5AD</a></span><span class="op">(</span><span class="va">file</span>, use_hdf5<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
  <span class="op">}</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">cbind</span>, <span class="va">sce.list</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="access-alternative-data">Access alternative data<a class="anchor" aria-label="anchor" href="#access-alternative-data"></a>
</h3>
<p>Some software, including <a href="https://pegasus.readthedocs.io/en/stable/" class="external-link">pegasus</a>, store normalized data in the default portion of the file and save raw counts in the <code>raw/</code> field. Here, load a H5AD file generated by pegasus.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># read raw/ from H5AD file</span>
<span class="co"># raw = TRUE tells readH5AD() to read alternative data</span>
<span class="co"># Must use zellkonverter &gt;=1.3.3</span>
<span class="va">sce_in</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/zellkonverter/man/readH5AD.html" class="external-link">readH5AD</a></span><span class="op">(</span><span class="va">h5ad_file</span>, use_hdf5<span class="op">=</span><span class="cn">TRUE</span>, raw<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># extract raw counts</span>
<span class="co"># include Dimnames and colData</span>
<span class="va">sce</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span> <span class="va">sce_in</span>, <span class="st">"raw"</span>, withDimnames<span class="op">=</span><span class="cn">TRUE</span>, withColData<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># convert these alternative data to SingleCellExperiment</span>
<span class="va">sce</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"SingleCellExperiment"</span><span class="op">)</span>

<span class="co"># Copy over reducedDims too</span>
<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDims</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDims</a></span><span class="op">(</span><span class="va">sce_in</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="using-seurat">Using Seurat<a class="anchor" aria-label="anchor" href="#using-seurat"></a>
</h2>
<p><a href="https://satijalab.org/seurat" class="external-link">Seurat</a> can also convert and import H5AD files, and then convert to <a href="https://bioconductor.org/packages/SingleCellExperiment/" class="external-link">SingleCellExperiment</a>. The resulting <code>SingleCellExperiment</code> object stores data as a <code>sparseMatrix</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mojaveazure.github.io/seurat-disk/" class="external-link">SeuratDisk</a></span><span class="op">)</span>

<span class="co"># Convert h5ad file to h5seurat format</span>
<span class="fu"><a href="https://mojaveazure.github.io/seurat-disk/reference/Convert.html" class="external-link">Convert</a></span><span class="op">(</span><span class="va">h5ad_file</span>, dest <span class="op">=</span> <span class="st">"h5seurat"</span><span class="op">)</span>

<span class="co"># load Seurat file</span>
<span class="va">obj</span> <span class="op">=</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-disk/reference/LoadH5Seurat.html" class="external-link">LoadH5Seurat</a></span><span class="op">(</span><span class="va">h5seurat_file</span><span class="op">)</span>

<span class="co"># Convert Seurat object to SingleCellExperiment</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">as.SingleCellExperiment</span><span class="op">(</span><span class="va">obj</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="comparing-interfaces">Comparing interfaces<a class="anchor" aria-label="anchor" href="#comparing-interfaces"></a>
</h2>
<p>The <code>SingleCellExperiment</code> interface to <code>zellkonverter</code> and <code>Seurat</code> hides the backend differences from the typical R user. The usage of <code>dreamlet</code> is the same in both cases. For small to medium datasets, the performance differences should be minimal. However, for large datasets there can be a substantial difference in performance. Use of <code>zellkonverter</code> and <code>DelayedMatrix</code> minimize memory usage at the cost of computationl performance. Use of <code>Seurat</code> and <code>sparseMatrix</code> can be faster if there is sufficent memory.</p>
<!---
Note that a [SingleCellExperiment](https://bioconductor.org/packages/SingleCellExperiment/) created by [zellkonverter](https://github.com/theislab/zellkonverter) stores the counts as a `DelayedMatrix` so the data remains on disk and are only accessed as needed.  This reduces memory usage substantailly, and `aggregateToPseudoBulk()` uses methods optimized for this data structure.

On the other hand, a [SingleCellExperiment](https://bioconductor.org/packages/SingleCellExperiment/) created using [Seurat](https://satijalab.org/seurat) and then `as.SingleCellExperiment()` stores count data in memory as a `sparseMatrix`.  This uses less memory than a dense matrix, but substantailly more than a `DelayedMatrix`.  While `aggregateToPseudoBulk()` has been developed to speed up analysis of `sparseMatrix` data, it can be considerablely slower then using the `DelayedMatrix` above.  
--->
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'dedcc260e9d3743bc634e728855cdbaf',
    indexName: 'dreamlet',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
