<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Testing non-linear effects • dreamlet</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Testing non-linear effects">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dreamlet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/dreamlet.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/mashr.html">mashr analysis after dreamlet</a></li>
    <li><a class="dropdown-item" href="../articles/cell_covs.html">Modeling continuous cell-level covariates</a></li>
    <li><a class="dropdown-item" href="../articles/non_lin_eff.html">Testing non-linear effects</a></li>
    <li><a class="dropdown-item" href="../articles/h5ad_on_disk.html">Handling large H5AD datasets</a></li>
    <li><a class="dropdown-item" href="../articles/errors.html">Error handling</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseNeurogenomics/dreamlet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Testing non-linear effects</h1>
            <h3 data-toc-skip class="subtitle">Categories and continuous
variables</h3>
                        <h4 data-toc-skip class="author">Developed by <a href="http://gabrielhoffman.github.io/" class="external-link">Gabriel Hoffman</a>
</h4>
            
            <h4 data-toc-skip class="date">Run on 2024-10-22
16:36:18</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeurogenomics/dreamlet/blob/HEAD/vignettes/non_lin_eff.Rmd" class="external-link"><code>vignettes/non_lin_eff.Rmd</code></a></small>
      <div class="d-none name"><code>non_lin_eff.Rmd</code></div>
    </div>

    
    
<!---

cd /Users/gabrielhoffman/workspace/repos/dreamlet/vignettes

rmarkdown::render("non_lin_eff.Rmd")


--->
<style>
body {
text-align: justify}
</style>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Typical analysis using regression models assumes a linear affect of
the covariate on the response. Here we consider testing non-linear
effects in the case of 1) continuous and 2) ordered categorical
variables.</p>
<p>We demonstrate this feature on a lightly modified analysis of PBMCs
from 8 individuals stimulated with interferon-β (<a href="https://www.nature.com/articles/nbt.4042" class="external-link">Kang, et al, 2018,
Nature Biotech</a>).</p>
</div>
<div class="section level2">
<h2 id="standard-processing">Standard processing<a class="anchor" aria-label="anchor" href="#standard-processing"></a>
</h2>
<p>Here is the code from the main vignette:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://DiseaseNeurogenomics.github.io/dreamlet" class="external-link">dreamlet</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HelenaLC/muscat" class="external-link">muscat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Download data, specifying EH2259 for the Kang, et al study</span></span>
<span><span class="va">eh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH2259"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># only keep singlet cells with sufficient reads</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">multiplets</span> <span class="op">==</span> <span class="st">"singlet"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># compute QC metrics</span></span>
<span><span class="va">qc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html" class="external-link">perCellQCMetrics</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove cells with few or many detected genes</span></span>
<span><span class="va">ol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span>metric <span class="op">=</span> <span class="va">qc</span><span class="op">$</span><span class="va">detected</span>, nmads <span class="op">=</span> <span class="fl">2</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="op">!</span><span class="va">ol</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># set variable indicating stimulated (stim) or control (ctrl)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">StimStatus</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">stim</span></span>
<span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">StimStatus</span>, <span class="va">sce</span><span class="op">$</span><span class="va">ind</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create pseudobulk</span></span>
<span><span class="va">pb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregateToPseudoBulk.html">aggregateToPseudoBulk</a></span><span class="op">(</span><span class="va">sce</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  cluster_id <span class="op">=</span> <span class="st">"cell"</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"id"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="continuous-variable">Continuous variable<a class="anchor" aria-label="anchor" href="#continuous-variable"></a>
</h2>
<p>Consider the continuous variable <code>Age</code>. Typical analysis
only considers linear effects using a single regression coefficient, but
we also want to consider the non-linear effects of age. We can peform a
<a href="https://bmcmedresmethodol.biomedcentral.com/articles/10.1186/s12874-019-0666-3" class="external-link">basis
expansion using splines</a> instead use 3 coefficients to model the age
effect.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Simulate age between 18 and 65</span></span>
<span><span class="va">pb</span><span class="op">$</span><span class="va">Age</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pb</span><span class="op">)</span>, <span class="fl">18</span>, <span class="fl">65</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># formula included non-linear effects of Age</span></span>
<span><span class="co"># by using a natural spline of degree 3</span></span>
<span><span class="co"># This corresponds to using 3 coefficients instead of 1</span></span>
<span><span class="va">form</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="fu">splines</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/splines/ns.html" class="external-link">ns</a></span><span class="op">(</span><span class="va">Age</span>, <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Normalize and apply voom/voomWithDreamWeights</span></span>
<span><span class="va">res.proc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/processAssays.html">processAssays</a></span><span class="op">(</span><span class="va">pb</span>, <span class="va">form</span>, min.count <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Differential expression analysis within each assay</span></span>
<span><span class="va">res.dl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dreamlet.html">dreamlet</a></span><span class="op">(</span><span class="va">res.proc</span>, <span class="va">form</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The spline has degree 3, so there are 3 coefficients</span></span>
<span><span class="co"># estimated for Age effects</span></span>
<span><span class="fu"><a href="../reference/coefNames-methods.html">coefNames</a></span><span class="op">(</span><span class="va">res.dl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "(Intercept)"          "splines::ns(Age, 3)1" "splines::ns(Age, 3)2"</span></span>
<span><span class="co">## [4] "splines::ns(Age, 3)3"</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Jointly test effects of the 3 spline components</span></span>
<span><span class="co"># The test of the 3 coefficients is performed with an F-statistic</span></span>
<span><span class="fu">topTable</span><span class="op">(</span><span class="va">res.dl</span>, coef <span class="op">=</span> <span class="fu"><a href="../reference/coefNames-methods.html">coefNames</a></span><span class="op">(</span><span class="va">res.dl</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, number <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 3 rows and 9 columns</span></span>
<span><span class="co">##         assay                     ID splines..ns.Age..3.1 splines..ns.Age..3.2</span></span>
<span><span class="co">##   &lt;character&gt;            &lt;character&gt;            &lt;numeric&gt;            &lt;numeric&gt;</span></span>
<span><span class="co">## 1 CD4 T cells                  GTF3A             0.751933              2.24124</span></span>
<span><span class="co">## 2 CD4 T cells                   RGS2            -0.591826             -3.72216</span></span>
<span><span class="co">## 3 CD4 T cells HLA-DRB1_ENSG0000019..            -0.924649             -1.94493</span></span>
<span><span class="co">##   splines..ns.Age..3.3   AveExpr         F     P.Value adj.P.Val</span></span>
<span><span class="co">##              &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1            -0.418741   8.63391   16.4736 2.82474e-05  0.178952</span></span>
<span><span class="co">## 2             0.385533   6.97805   14.9336 5.15218e-05  0.178952</span></span>
<span><span class="co">## 3             1.251838   5.24781   14.7922 5.45632e-05  0.178952</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="ordered-categorical">Ordered categorical<a class="anchor" aria-label="anchor" href="#ordered-categorical"></a>
</h2>
<p>We can also test non-linear effects in the case of categorical
variables with a natural ordering to the categories. Consider time
course data with 4 time points. Each time point is a category and has a
natural ordering from first to last.</p>
<p>We have multiple options to model the time course.</p>
<ul>
<li>
<p><strong>Continuous:</strong> Modeling time point as a continuous
variable uses a single regression coefficient to model the linear
effects of the time course. This is simple, models the order of the time
points, but ignores non-linear effects</p>
<p>Model using <code>as.numeric(TimePoint)</code></p>
</li>
<li>
<p><strong>Categorical:</strong> Including time point as a typical
categorical variable uses estimated the mean response value for each
category. So it estimates 4 coefficients. While this can be useful for
comparing two categories, it ignores the order of the time points.</p>
<p>Model using <code>factor(TimePoint)</code></p>
</li>
<li>
<p><strong>Ordered categorical:</strong> Here, the trend across
ordered time points is modled using orthogonal polynomials. The trend is
decomposed into independent linear, quadratic, etc., effects that can be
tested either jointly or by themselves.</p>
<p>Model using:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"time_1"</span>, <span class="st">"time_2"</span>, <span class="st">"time_3"</span>, <span class="st">"time_4"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">ordered</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">TimePoint</span><span class="op">)</span>, <span class="va">ord</span><span class="op">)</span></span></code></pre></div>
</li>
</ul>
<p>Here we simulated 4 time points, and perform differential expression
analysis.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Consider data generated across 4 time points</span></span>
<span><span class="co"># While there are no time points in the real data</span></span>
<span><span class="co"># we can add some for demonstration purposes</span></span>
<span><span class="va">pb</span><span class="op">$</span><span class="va">TimePoint</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">ordered</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"time_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># examine the ordering</span></span>
<span><span class="va">pb</span><span class="op">$</span><span class="va">TimePoint</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] time_1 time_2 time_3 time_4 time_1 time_2 time_3 time_4 time_1 time_2</span></span>
<span><span class="co">## [11] time_3 time_4 time_1 time_2 time_3 time_4</span></span>
<span><span class="co">## Levels: time_1 &lt; time_2 &lt; time_3 &lt; time_4</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use formula including time point</span></span>
<span><span class="va">form</span> <span class="op">&lt;-</span> <span class="op">~</span><span class="va">TimePoint</span></span>
<span></span>
<span><span class="co"># Normalize and apply voom/voomWithDreamWeights</span></span>
<span><span class="va">res.proc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/processAssays.html">processAssays</a></span><span class="op">(</span><span class="va">pb</span>, <span class="va">form</span>, min.count <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Differential expression analysis within each assay</span></span>
<span><span class="va">res.dl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dreamlet.html">dreamlet</a></span><span class="op">(</span><span class="va">res.proc</span>, <span class="va">form</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Examine the coefficient estimated</span></span>
<span><span class="co"># for TimePoint it estimates</span></span>
<span><span class="co"># linear (i.e. L)</span></span>
<span><span class="co"># quadratic (i.e. Q)</span></span>
<span><span class="co"># and cubic (i.e. C) effects</span></span>
<span><span class="fu"><a href="../reference/coefNames-methods.html">coefNames</a></span><span class="op">(</span><span class="va">res.dl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "(Intercept)" "TimePoint.L" "TimePoint.Q" "TimePoint.C"</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test only linear effect</span></span>
<span><span class="fu">topTable</span><span class="op">(</span><span class="va">res.dl</span>, coef <span class="op">=</span> <span class="st">"TimePoint.L"</span>, number <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 3 rows and 9 columns</span></span>
<span><span class="co">##         assay          ID     logFC   AveExpr         t     P.Value adj.P.Val</span></span>
<span><span class="co">##   &lt;character&gt; &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1 CD4 T cells        DCXR -0.671645   6.52867  -5.42880 4.78666e-05  0.393058</span></span>
<span><span class="co">## 2 CD4 T cells        GGA2 -0.890112   4.98987  -5.26372 6.69370e-05  0.393058</span></span>
<span><span class="co">## 3 CD8 T cells        FTH1 -0.759875  14.55380  -4.99199 8.13672e-05  0.393058</span></span>
<span><span class="co">##           B     z.std</span></span>
<span><span class="co">##   &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1  1.686147  -5.42880</span></span>
<span><span class="co">## 2  0.936906  -5.26372</span></span>
<span><span class="co">## 3  1.627753  -4.99199</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test linear, quadratic and cubic effcts</span></span>
<span><span class="va">coefs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TimePoint.L"</span>, <span class="st">"TimePoint.Q"</span>, <span class="st">"TimePoint.C"</span><span class="op">)</span></span>
<span><span class="fu">topTable</span><span class="op">(</span><span class="va">res.dl</span>, coef <span class="op">=</span> <span class="va">coefs</span>, number <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 3 rows and 9 columns</span></span>
<span><span class="co">##         assay                   ID TimePoint.L TimePoint.Q TimePoint.C</span></span>
<span><span class="co">##   &lt;character&gt;          &lt;character&gt;   &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">## 1 CD8 T cells                 CD52    0.903032   -1.528456   -0.984065</span></span>
<span><span class="co">## 2 CD8 T cells CCL5_ENSG00000161570    0.626201   -0.986502   -0.645468</span></span>
<span><span class="co">## 3 CD8 T cells                  CD2    0.887578   -1.152415   -0.195616</span></span>
<span><span class="co">##     AveExpr         F     P.Value adj.P.Val</span></span>
<span><span class="co">##   &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1   8.69458   16.0946 1.90868e-05 0.0853298</span></span>
<span><span class="co">## 2  11.88811   16.0433 1.94982e-05 0.0853298</span></span>
<span><span class="co">## 3   9.67181   15.7793 2.17777e-05 0.0853298</span></span></code></pre>
<div class="section level3">
<h3 id="sample-filtering">Sample filtering<a class="anchor" aria-label="anchor" href="#sample-filtering"></a>
</h3>
<p>Due to variation in cell and read count for each sample,
<code><a href="../reference/processAssays.html">processAssays()</a></code> filters out some sample. This filtering is
summarized here:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/details-methods.html">details</a></span><span class="op">(</span><span class="va">res.dl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##               assay n_retain    formula formDropsTerms n_genes n_errors</span></span>
<span><span class="co">## 1           B cells       16 ~TimePoint          FALSE    1961        0</span></span>
<span><span class="co">## 2   CD14+ Monocytes       16 ~TimePoint          FALSE    3087        0</span></span>
<span><span class="co">## 3       CD4 T cells       16 ~TimePoint          FALSE    5262        0</span></span>
<span><span class="co">## 4       CD8 T cells       16 ~TimePoint          FALSE    1030        0</span></span>
<span><span class="co">## 5   Dendritic cells       13 ~TimePoint          FALSE     164        0</span></span>
<span><span class="co">## 6 FCGR3A+ Monocytes       16 ~TimePoint          FALSE    1160        0</span></span>
<span><span class="co">## 7    Megakaryocytes       13 ~TimePoint          FALSE     172        0</span></span>
<span><span class="co">## 8          NK cells       16 ~TimePoint          FALSE    1656        0</span></span>
<span><span class="co">##   error_initial</span></span>
<span><span class="co">## 1         FALSE</span></span>
<span><span class="co">## 2         FALSE</span></span>
<span><span class="co">## 3         FALSE</span></span>
<span><span class="co">## 4         FALSE</span></span>
<span><span class="co">## 5         FALSE</span></span>
<span><span class="co">## 6         FALSE</span></span>
<span><span class="co">## 7         FALSE</span></span>
<span><span class="co">## 8         FALSE</span></span></code></pre>
<p>Whle all 16 samples are detained in B cells, only 9 are retained for
megakaryocytes. This can result in a time point being dropped, and so
the polynomial expansion for some cell types can have a lower degree.
The combined results will then have <code>NA</code> values for these
coefficients. For example, for <code>TIMP1</code> in
<code>Megakaryocytes</code> above there is not enought data to fit the
cubic term, so <code>TimePoint.C</code> is <code>NA</code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><pre><code><span><span class="co">## R version 4.4.0 (2024-04-24)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin23.5.0</span></span>
<span><span class="co">## Running under: macOS Sonoma 14.6.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib </span></span>
<span><span class="co">## LAPACK: /Users/gabrielhoffman/prog/R-4.4.0/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/New_York</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] muscData_1.18.0             scater_1.32.1              </span></span>
<span><span class="co">##  [3] scuttle_1.14.0              ExperimentHub_2.12.0       </span></span>
<span><span class="co">##  [5] AnnotationHub_3.12.0        BiocFileCache_2.12.0       </span></span>
<span><span class="co">##  [7] dbplyr_2.5.0                muscat_1.18.0              </span></span>
<span><span class="co">##  [9] dreamlet_1.3.3              SingleCellExperiment_1.26.0</span></span>
<span><span class="co">## [11] SummarizedExperiment_1.34.0 Biobase_2.64.0             </span></span>
<span><span class="co">## [13] GenomicRanges_1.56.1        GenomeInfoDb_1.40.1        </span></span>
<span><span class="co">## [15] IRanges_2.38.1              S4Vectors_0.42.1           </span></span>
<span><span class="co">## [17] BiocGenerics_0.50.0         MatrixGenerics_1.16.0      </span></span>
<span><span class="co">## [19] matrixStats_1.4.1           variancePartition_1.35.5   </span></span>
<span><span class="co">## [21] BiocParallel_1.38.0         limma_3.60.4               </span></span>
<span><span class="co">## [23] ggplot2_3.5.1               BiocStyle_2.32.1           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fs_1.6.4                  bitops_1.0-8             </span></span>
<span><span class="co">##   [3] httr_1.4.7                RColorBrewer_1.1-3       </span></span>
<span><span class="co">##   [5] doParallel_1.0.17         Rgraphviz_2.48.0         </span></span>
<span><span class="co">##   [7] numDeriv_2016.8-1.1       tools_4.4.0              </span></span>
<span><span class="co">##   [9] sctransform_0.4.1         backports_1.5.0          </span></span>
<span><span class="co">##  [11] utf8_1.2.4                R6_2.5.1                 </span></span>
<span><span class="co">##  [13] metafor_4.6-0             mgcv_1.9-1               </span></span>
<span><span class="co">##  [15] GetoptLong_1.0.5          withr_3.0.1              </span></span>
<span><span class="co">##  [17] prettyunits_1.2.0         gridExtra_2.3            </span></span>
<span><span class="co">##  [19] cli_3.6.3                 textshaping_0.4.0        </span></span>
<span><span class="co">##  [21] sass_0.4.9                KEGGgraph_1.64.0         </span></span>
<span><span class="co">##  [23] SQUAREM_2021.1            mvtnorm_1.3-1            </span></span>
<span><span class="co">##  [25] blme_1.0-6                pkgdown_2.1.1            </span></span>
<span><span class="co">##  [27] mixsqp_0.3-54             systemfonts_1.1.0        </span></span>
<span><span class="co">##  [29] zenith_1.6.0              parallelly_1.38.0        </span></span>
<span><span class="co">##  [31] invgamma_1.1              RSQLite_2.3.7            </span></span>
<span><span class="co">##  [33] generics_0.1.3            shape_1.4.6.1            </span></span>
<span><span class="co">##  [35] gtools_3.9.5              dplyr_1.1.4              </span></span>
<span><span class="co">##  [37] Matrix_1.7-0              metadat_1.2-0            </span></span>
<span><span class="co">##  [39] ggbeeswarm_0.7.2          fansi_1.0.6              </span></span>
<span><span class="co">##  [41] abind_1.4-8               lifecycle_1.0.4          </span></span>
<span><span class="co">##  [43] yaml_2.3.10               edgeR_4.2.1              </span></span>
<span><span class="co">##  [45] mathjaxr_1.6-0            gplots_3.1.3.1           </span></span>
<span><span class="co">##  [47] SparseArray_1.4.8         grid_4.4.0               </span></span>
<span><span class="co">##  [49] blob_1.2.4                crayon_1.5.3             </span></span>
<span><span class="co">##  [51] lattice_0.22-6            beachmat_2.20.0          </span></span>
<span><span class="co">##  [53] msigdbr_7.5.1             annotate_1.82.0          </span></span>
<span><span class="co">##  [55] KEGGREST_1.44.1           pillar_1.9.0             </span></span>
<span><span class="co">##  [57] knitr_1.48                ComplexHeatmap_2.20.0    </span></span>
<span><span class="co">##  [59] rjson_0.2.23              boot_1.3-31              </span></span>
<span><span class="co">##  [61] corpcor_1.6.10            future.apply_1.11.2      </span></span>
<span><span class="co">##  [63] codetools_0.2-20          glue_1.8.0               </span></span>
<span><span class="co">##  [65] data.table_1.16.0         vctrs_0.6.5              </span></span>
<span><span class="co">##  [67] png_0.1-8                 Rdpack_2.6.1             </span></span>
<span><span class="co">##  [69] gtable_0.3.5              assertthat_0.2.1         </span></span>
<span><span class="co">##  [71] cachem_1.1.0              xfun_0.47                </span></span>
<span><span class="co">##  [73] mime_0.12                 rbibutils_2.2.16         </span></span>
<span><span class="co">##  [75] S4Arrays_1.4.1            Rfast_2.1.0              </span></span>
<span><span class="co">##  [77] iterators_1.0.14          statmod_1.5.0            </span></span>
<span><span class="co">##  [79] nlme_3.1-166              pbkrtest_0.5.3           </span></span>
<span><span class="co">##  [81] bit64_4.5.2               filelock_1.0.3           </span></span>
<span><span class="co">##  [83] progress_1.2.3            EnvStats_3.0.0           </span></span>
<span><span class="co">##  [85] bslib_0.8.0               TMB_1.9.15               </span></span>
<span><span class="co">##  [87] irlba_2.3.5.1             vipor_0.4.7              </span></span>
<span><span class="co">##  [89] KernSmooth_2.23-24        colorspace_2.1-1         </span></span>
<span><span class="co">##  [91] rmeta_3.0                 DBI_1.2.3                </span></span>
<span><span class="co">##  [93] DESeq2_1.44.0             tidyselect_1.2.1         </span></span>
<span><span class="co">##  [95] curl_5.2.3                bit_4.5.0                </span></span>
<span><span class="co">##  [97] compiler_4.4.0            graph_1.82.0             </span></span>
<span><span class="co">##  [99] BiocNeighbors_1.22.0      desc_1.4.3               </span></span>
<span><span class="co">## [101] DelayedArray_0.30.1       bookdown_0.40            </span></span>
<span><span class="co">## [103] scales_1.3.0              caTools_1.18.3           </span></span>
<span><span class="co">## [105] remaCor_0.0.18            rappdirs_0.3.3           </span></span>
<span><span class="co">## [107] stringr_1.5.1             digest_0.6.37            </span></span>
<span><span class="co">## [109] minqa_1.2.8               rmarkdown_2.28           </span></span>
<span><span class="co">## [111] aod_1.3.3                 XVector_0.44.0           </span></span>
<span><span class="co">## [113] RhpcBLASctl_0.23-42       htmltools_0.5.8.1        </span></span>
<span><span class="co">## [115] pkgconfig_2.0.3           lme4_1.1-35.5            </span></span>
<span><span class="co">## [117] sparseMatrixStats_1.16.0  mashr_0.2.79             </span></span>
<span><span class="co">## [119] fastmap_1.2.0             rlang_1.1.4              </span></span>
<span><span class="co">## [121] GlobalOptions_0.1.2       htmlwidgets_1.6.4        </span></span>
<span><span class="co">## [123] UCSC.utils_1.0.0          DelayedMatrixStats_1.26.0</span></span>
<span><span class="co">## [125] jquerylib_0.1.4           jsonlite_1.8.9           </span></span>
<span><span class="co">## [127] BiocSingular_1.20.0       RCurl_1.98-1.16          </span></span>
<span><span class="co">## [129] magrittr_2.0.3            GenomeInfoDbData_1.2.12  </span></span>
<span><span class="co">## [131] munsell_0.5.1             Rcpp_1.0.13              </span></span>
<span><span class="co">## [133] babelgene_22.9            viridis_0.6.5            </span></span>
<span><span class="co">## [135] EnrichmentBrowser_2.34.1  RcppZiggurat_0.1.6       </span></span>
<span><span class="co">## [137] stringi_1.8.4             zlibbioc_1.50.0          </span></span>
<span><span class="co">## [139] MASS_7.3-61               plyr_1.8.9               </span></span>
<span><span class="co">## [141] listenv_0.9.1             parallel_4.4.0           </span></span>
<span><span class="co">## [143] ggrepel_0.9.6             Biostrings_2.72.1        </span></span>
<span><span class="co">## [145] splines_4.4.0             hms_1.1.3                </span></span>
<span><span class="co">## [147] circlize_0.4.16           locfit_1.5-9.10          </span></span>
<span><span class="co">## [149] reshape2_1.4.4            ScaledMatrix_1.12.0      </span></span>
<span><span class="co">## [151] BiocVersion_3.19.1        XML_3.99-0.17            </span></span>
<span><span class="co">## [153] evaluate_1.0.0            RcppParallel_5.1.9       </span></span>
<span><span class="co">## [155] BiocManager_1.30.25       nloptr_2.1.1             </span></span>
<span><span class="co">## [157] foreach_1.5.2             tidyr_1.3.1              </span></span>
<span><span class="co">## [159] purrr_1.0.2               future_1.34.0            </span></span>
<span><span class="co">## [161] clue_0.3-65               scattermore_1.2          </span></span>
<span><span class="co">## [163] ashr_2.2-63               rsvd_1.0.5               </span></span>
<span><span class="co">## [165] broom_1.0.6               xtable_1.8-4             </span></span>
<span><span class="co">## [167] fANCOVA_0.6-1             viridisLite_0.4.2        </span></span>
<span><span class="co">## [169] ragg_1.3.3                truncnorm_1.0-9          </span></span>
<span><span class="co">## [171] tibble_3.2.1              lmerTest_3.1-3           </span></span>
<span><span class="co">## [173] glmmTMB_1.1.9             memoise_2.0.1            </span></span>
<span><span class="co">## [175] beeswarm_0.4.0            AnnotationDbi_1.66.0     </span></span>
<span><span class="co">## [177] cluster_2.1.6             globals_0.16.3           </span></span>
<span><span class="co">## [179] GSEABase_1.66.0</span></span></code></pre>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
