<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="dreamlet">
<title>Testing non-linear effects • dreamlet</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Testing non-linear effects">
<meta property="og:description" content="dreamlet">
<meta property="og:image" content="http://DiseaseNeurogenomics.github.io/dreamlet/logo.svg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dreamlet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.13</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/dreamlet.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/cell_covs.html">Modeling continuous cell-level covariates</a>
    <a class="dropdown-item" href="../articles/h5ad_on_disk.html">Handling large H5AD datasets</a>
    <a class="dropdown-item" href="../articles/mashr.html">mashr analysis after dreamlet</a>
    <a class="dropdown-item" href="../articles/non_lin_eff.html">Testing non-linear effects</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/DiseaseNeurogenomics/dreamlet/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Testing non-linear effects</h1>
            <h3 data-toc-skip class="subtitle">Categories and continuous
variables</h3>
                        <h4 data-toc-skip class="author">Developed by <a href="http://gabrielhoffman.github.io/" class="external-link">Gabriel Hoffman</a>
</h4>
            
            <h4 data-toc-skip class="date">Run on 2023-05-31
19:42:51.94604</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeurogenomics/dreamlet/blob/HEAD/vignettes/non_lin_eff.Rmd" class="external-link"><code>vignettes/non_lin_eff.Rmd</code></a></small>
      <div class="d-none name"><code>non_lin_eff.Rmd</code></div>
    </div>

    
    
<!---

cd /Users/gabrielhoffman/workspace/repos/dreamlet/vignettes

rmarkdown::render("non_lin_eff.Rmd")


--->
<style>
body {
text-align: justify}
</style>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Typical analysis using regression models assumes a linear affect of
the covariate on the response. Here we consider testing non-linear
effects in the case of 1) continuous and 2) ordered categorical
variables.</p>
<p>We demonstrate this feature on a lightly modified analysis of PBMCs
from 8 individuals stimulated with interferon-β (<a href="https://www.nature.com/articles/nbt.4042" class="external-link">Kang, et al, 2018,
Nature Biotech</a>).</p>
</div>
<div class="section level2">
<h2 id="standard-processing">Standard processing<a class="anchor" aria-label="anchor" href="#standard-processing"></a>
</h2>
<p>Here is the code from the main vignette:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://DiseaseNeurogenomics.github.io/dreamlet" class="external-link">dreamlet</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HelenaLC/muscat" class="external-link">muscat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Download data, specifying EH2259 for the Kang, et al study</span></span>
<span><span class="va">eh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH2259"</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># only keep singlet cells with sufficient reads</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">multiplets</span> <span class="op">==</span> <span class="st">'singlet'</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># compute QC metrics</span></span>
<span><span class="va">qc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html" class="external-link">perCellQCMetrics</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove cells with few or many detected genes</span></span>
<span><span class="va">ol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span>metric <span class="op">=</span> <span class="va">qc</span><span class="op">$</span><span class="va">detected</span>, nmads <span class="op">=</span> <span class="fl">2</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="op">!</span><span class="va">ol</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># set variable indicating stimulated (stim) or control (ctrl)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">StimStatus</span> <span class="op">=</span> <span class="va">sce</span><span class="op">$</span><span class="va">stim</span></span>
<span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">StimStatus</span>, <span class="va">sce</span><span class="op">$</span><span class="va">ind</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create pseudobulk </span></span>
<span><span class="va">pb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregateToPseudoBulk.html">aggregateToPseudoBulk</a></span><span class="op">(</span><span class="va">sce</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"counts"</span>,     </span>
<span>    cluster_id <span class="op">=</span> <span class="st">"cell"</span>,  </span>
<span>    sample_id <span class="op">=</span> <span class="st">"id"</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="continuous-variable">Continuous variable<a class="anchor" aria-label="anchor" href="#continuous-variable"></a>
</h2>
<p>Consider the continuous variable <code>Age</code>. Typical analysis
only considers linear effects using a single regression coefficient, but
we also want to consider the non-linear effects of age. We can peform a
<a href="https://bmcmedresmethodol.biomedcentral.com/articles/10.1186/s12874-019-0666-3" class="external-link">basis
expansion using splines</a> instead use 3 coefficients to model the age
effect.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Simulate age between 18 and 65</span></span>
<span><span class="va">pb</span><span class="op">$</span><span class="va">Age</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pb</span><span class="op">)</span>, <span class="fl">18</span>, <span class="fl">65</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># formula included non-linear effects of Age</span></span>
<span><span class="co"># by using a natural spline of degree 3</span></span>
<span><span class="co"># This corresponds to using 3 coefficients instead of 1</span></span>
<span><span class="va">form</span> <span class="op">=</span> <span class="op">~</span> <span class="fu">splines</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/splines/ns.html" class="external-link">ns</a></span><span class="op">(</span><span class="va">Age</span>, <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Normalize and apply voom/voomWithDreamWeights</span></span>
<span><span class="va">res.proc</span> <span class="op">=</span> <span class="fu"><a href="../reference/processAssays.html">processAssays</a></span><span class="op">(</span> <span class="va">pb</span>, <span class="va">form</span>, min.count<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Differential expression analysis within each assay</span></span>
<span><span class="va">res.dl</span> <span class="op">=</span> <span class="fu"><a href="../reference/dreamlet.html">dreamlet</a></span><span class="op">(</span> <span class="va">res.proc</span>, <span class="va">form</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The spline has degree 3, so there are 3 coefficients</span></span>
<span><span class="co"># estimated for Age effects</span></span>
<span><span class="fu"><a href="../reference/coefNames-methods.html">coefNames</a></span><span class="op">(</span><span class="va">res.dl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "(Intercept)"          "splines::ns(Age, 3)1" "splines::ns(Age, 3)2"</span></span>
<span><span class="co">## [4] "splines::ns(Age, 3)3"</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Jointly test effects of the 3 spline components</span></span>
<span><span class="co"># The test of the 3 coefficients is performed with an F-statistic</span></span>
<span><span class="fu">topTable</span><span class="op">(</span><span class="va">res.dl</span>, coef<span class="op">=</span><span class="fu"><a href="../reference/coefNames-methods.html">coefNames</a></span><span class="op">(</span><span class="va">res.dl</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, number<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 3 rows and 9 columns</span></span>
<span><span class="co">##         assay          ID splines..ns.Age..3.1 splines..ns.Age..3.2</span></span>
<span><span class="co">##   &lt;character&gt; &lt;character&gt;            &lt;numeric&gt;            &lt;numeric&gt;</span></span>
<span><span class="co">## 1    NK cells        HCST             0.779518              6.47076</span></span>
<span><span class="co">## 2    NK cells       NUTF2             1.383123             -2.09056</span></span>
<span><span class="co">## 3    NK cells    TRAF3IP3            -0.577738              6.15883</span></span>
<span><span class="co">##   splines..ns.Age..3.3   AveExpr         F     P.Value adj.P.Val</span></span>
<span><span class="co">##              &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1              1.31306   7.28001   12.6756 0.000118118  0.555049</span></span>
<span><span class="co">## 2             -2.47219   6.40995   11.7734 0.000181324  0.555049</span></span>
<span><span class="co">## 3              1.35218   6.52589   11.3674 0.000221430  0.555049</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="ordered-categorical">Ordered categorical<a class="anchor" aria-label="anchor" href="#ordered-categorical"></a>
</h2>
<p>We can also test non-linear effects in the case of categorical
variables with a natural ordering to the categories. Consider time
course data with 4 time points. Each time point is a category and has a
natural ordering from first to last.</p>
<p>We have multiple options to model the time course.</p>
<ul>
<li>
<p><strong>Continuous:</strong> Modeling time point as a continuous
variable uses a single regression coefficient to model the linear
effects of the time course. This is simple, models the order of the time
points, but ignores non-linear effects</p>
<p>Model using <code>as.numeric(TimePoint)</code></p>
</li>
<li>
<p><strong>Categorical:</strong> Including time point as a typical
categorical variable uses estimated the mean response value for each
category. So it estimates 4 coefficients. While this can be useful for
comparing two categories, it ignores the order of the time points.</p>
<p>Model using <code>factor(TimePoint)</code></p>
</li>
<li>
<p><strong>Ordered categorical:</strong> Here, the trend across
ordered time points is modled using orthogonal polynomials. The trend is
decomposed into independent linear, quadratic, etc., effects that can be
tested either jointly or by themselves.</p>
<p>Model using:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>ord <span class="ot">=</span> <span class="fu">c</span>(<span class="st">'time_1'</span>,<span class="st">'time_2'</span>,<span class="st">'time_3'</span>m,<span class="st">'time_4'</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ordered</span>(<span class="fu">factor</span>(TimePoint), ord)</span></code></pre></div>
</li>
</ul>
<p>Here we simulated 4 time points, and perform differential expression
analysis.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Consider data generated across 4 time points</span></span>
<span><span class="co"># While there are no time points in the real data</span></span>
<span><span class="co"># we can add some for demonstration purposes </span></span>
<span><span class="va">pb</span><span class="op">$</span><span class="va">TimePoint</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">ordered</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"time_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># examine the ordering</span></span>
<span><span class="va">pb</span><span class="op">$</span><span class="va">TimePoint</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] time_1 time_2 time_3 time_4 time_1 time_2 time_3 time_4 time_1 time_2</span></span>
<span><span class="co">## [11] time_3 time_4 time_1 time_2 time_3 time_4</span></span>
<span><span class="co">## Levels: time_1 &lt; time_2 &lt; time_3 &lt; time_4</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use formula including time point</span></span>
<span><span class="va">form</span> <span class="op">=</span> <span class="op">~</span> <span class="va">TimePoint</span></span>
<span></span>
<span><span class="co"># Normalize and apply voom/voomWithDreamWeights</span></span>
<span><span class="va">res.proc</span> <span class="op">=</span> <span class="fu"><a href="../reference/processAssays.html">processAssays</a></span><span class="op">(</span> <span class="va">pb</span>, <span class="va">form</span>, min.count<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Differential expression analysis within each assay</span></span>
<span><span class="va">res.dl</span> <span class="op">=</span> <span class="fu"><a href="../reference/dreamlet.html">dreamlet</a></span><span class="op">(</span> <span class="va">res.proc</span>, <span class="va">form</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Examine the coefficient estimated</span></span>
<span><span class="co"># for TimePoint it estimates </span></span>
<span><span class="co"># linear (i.e. L) </span></span>
<span><span class="co"># quadratic (i.e. Q)</span></span>
<span><span class="co"># and cubic (i.e. C) effects</span></span>
<span><span class="fu"><a href="../reference/coefNames-methods.html">coefNames</a></span><span class="op">(</span><span class="va">res.dl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "(Intercept)" "TimePoint.L" "TimePoint.Q" "TimePoint.C"</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test only linear effect</span></span>
<span><span class="fu">topTable</span><span class="op">(</span><span class="va">res.dl</span>, coef<span class="op">=</span><span class="st">"TimePoint.L"</span>, number<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 3 rows and 9 columns</span></span>
<span><span class="co">##             assay          ID     logFC   AveExpr         t     P.Value</span></span>
<span><span class="co">##       &lt;character&gt; &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">## 1     CD4 T cells        DCXR -0.678397   6.52867  -5.65493 3.08306e-05</span></span>
<span><span class="co">## 2     CD4 T cells        GGA2 -0.890912   4.98987  -5.27996 6.55506e-05</span></span>
<span><span class="co">## 3 CD14+ Monocytes        CAPG -0.874746   8.83660  -5.18913 1.12897e-04</span></span>
<span><span class="co">##   adj.P.Val         B     z.std</span></span>
<span><span class="co">##   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1  0.451791  2.064035  -5.65493</span></span>
<span><span class="co">## 2  0.460027  0.826826  -5.27996</span></span>
<span><span class="co">## 3  0.460027 -3.020427  -5.18913</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test linear, quadratic and cubic effcts</span></span>
<span><span class="va">coefs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TimePoint.L"</span>, <span class="st">"TimePoint.Q"</span>, <span class="st">"TimePoint.C"</span><span class="op">)</span></span>
<span><span class="fu">topTable</span><span class="op">(</span><span class="va">res.dl</span>, coef<span class="op">=</span><span class="va">coefs</span>, number<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 3 rows and 9 columns</span></span>
<span><span class="co">##            assay          ID TimePoint.L TimePoint.Q TimePoint.C   AveExpr</span></span>
<span><span class="co">##      &lt;character&gt; &lt;character&gt;   &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1 Megakaryocytes       TIMP1   -0.191992     1.97379          NA  10.94510</span></span>
<span><span class="co">## 2    CD8 T cells      GPR183   -0.834378     1.12444    1.155308   8.47413</span></span>
<span><span class="co">## 3    CD8 T cells        CD52    0.879689    -1.50610   -0.960712   8.69458</span></span>
<span><span class="co">##           F     P.Value adj.P.Val</span></span>
<span><span class="co">##   &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1   19.9684 1.36276e-06 0.0199699</span></span>
<span><span class="co">## 2   16.5751 1.55974e-05 0.0826416</span></span>
<span><span class="co">## 3   15.9934 1.98237e-05 0.0826416</span></span></code></pre>
<div class="section level3">
<h3 id="sample-filtering">Sample filtering<a class="anchor" aria-label="anchor" href="#sample-filtering"></a>
</h3>
<p>Due to variation in cell and read count for each sample,
<code><a href="../reference/processAssays.html">processAssays()</a></code> filters out some sample. This filtering is
summarized here:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/details-methods.html">details</a></span><span class="op">(</span><span class="va">res.dl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##               assay n_retain    formula formDropsTerms n_errors error_initial</span></span>
<span><span class="co">## 1           B cells       16 ~TimePoint          FALSE        0         FALSE</span></span>
<span><span class="co">## 2   CD14+ Monocytes       16 ~TimePoint          FALSE        0         FALSE</span></span>
<span><span class="co">## 3       CD4 T cells       16 ~TimePoint          FALSE        0         FALSE</span></span>
<span><span class="co">## 4       CD8 T cells       16 ~TimePoint          FALSE        0         FALSE</span></span>
<span><span class="co">## 5   Dendritic cells        5 ~TimePoint          FALSE        0         FALSE</span></span>
<span><span class="co">## 6 FCGR3A+ Monocytes       16 ~TimePoint          FALSE        0         FALSE</span></span>
<span><span class="co">## 7    Megakaryocytes        9 ~TimePoint          FALSE        0         FALSE</span></span>
<span><span class="co">## 8          NK cells       16 ~TimePoint          FALSE        0         FALSE</span></span></code></pre>
<p>Whle all 16 samples are detained in B cells, only 9 are retained for
megakaryocytes. This can result in a time point being dropped, and so
the polynomial expansion for some cell types can have a lower degree.
The combined results will then have <code>NA</code> values for these
coefficients. For example, for <code>TIMP1</code> in
<code>Megakaryocytes</code> above there is not enought data to fit the
cubic term, so <code>TimePoint.C</code> is <code>NA</code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><pre><code><span><span class="co">## R version 4.2.0 (2022-04-22)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin19.6.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Big Sur/Monterey 10.16</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] muscData_1.10.0             scater_1.24.0              </span></span>
<span><span class="co">##  [3] scuttle_1.6.3               SingleCellExperiment_1.18.1</span></span>
<span><span class="co">##  [5] SummarizedExperiment_1.26.1 Biobase_2.58.0             </span></span>
<span><span class="co">##  [7] GenomicRanges_1.48.0        GenomeInfoDb_1.32.4        </span></span>
<span><span class="co">##  [9] IRanges_2.30.1              S4Vectors_0.34.0           </span></span>
<span><span class="co">## [11] MatrixGenerics_1.8.1        matrixStats_0.63.0         </span></span>
<span><span class="co">## [13] ExperimentHub_2.4.0         AnnotationHub_3.4.0        </span></span>
<span><span class="co">## [15] BiocFileCache_2.4.0         dbplyr_2.3.1               </span></span>
<span><span class="co">## [17] BiocGenerics_0.44.0         muscat_1.11.2              </span></span>
<span><span class="co">## [19] dreamlet_0.99.10            variancePartition_2.0.0    </span></span>
<span><span class="co">## [21] BiocParallel_1.32.0         limma_3.54.0               </span></span>
<span><span class="co">## [23] ggplot2_3.4.1               BiocStyle_2.24.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] rappdirs_0.3.3                estimability_1.4.1           </span></span>
<span><span class="co">##   [3] scattermore_0.8               mashr_0.2.57                 </span></span>
<span><span class="co">##   [5] coda_0.19-4                   ragg_1.2.5                   </span></span>
<span><span class="co">##   [7] tidyr_1.3.0                   clusterGeneration_1.3.7      </span></span>
<span><span class="co">##   [9] bit64_4.0.5                   knitr_1.42                   </span></span>
<span><span class="co">##  [11] irlba_2.3.5.1                 multcomp_1.4-23              </span></span>
<span><span class="co">##  [13] DelayedArray_0.22.0           data.table_1.14.8            </span></span>
<span><span class="co">##  [15] KEGGREST_1.36.3               RCurl_1.98-1.10              </span></span>
<span><span class="co">##  [17] doParallel_1.0.17             generics_0.1.3               </span></span>
<span><span class="co">##  [19] ScaledMatrix_1.4.1            RhpcBLASctl_0.23-42          </span></span>
<span><span class="co">##  [21] TH.data_1.1-1                 RSQLite_2.3.0                </span></span>
<span><span class="co">##  [23] future_1.32.0                 bit_4.0.5                    </span></span>
<span><span class="co">##  [25] httpuv_1.6.9                  assertthat_0.2.1             </span></span>
<span><span class="co">##  [27] viridis_0.6.2                 xfun_0.37                    </span></span>
<span><span class="co">##  [29] hms_1.1.2                     jquerylib_0.1.4              </span></span>
<span><span class="co">##  [31] promises_1.2.0.1              babelgene_22.9               </span></span>
<span><span class="co">##  [33] evaluate_0.20                 fansi_1.0.4                  </span></span>
<span><span class="co">##  [35] progress_1.2.2                caTools_1.18.2               </span></span>
<span><span class="co">##  [37] Rgraphviz_2.40.0              DBI_1.1.3                    </span></span>
<span><span class="co">##  [39] geneplotter_1.74.0            purrr_1.0.1                  </span></span>
<span><span class="co">##  [41] ellipsis_0.3.2                dplyr_1.1.0                  </span></span>
<span><span class="co">##  [43] backports_1.4.1               bookdown_0.33                </span></span>
<span><span class="co">##  [45] annotate_1.74.0               aod_1.3.2                    </span></span>
<span><span class="co">##  [47] sparseMatrixStats_1.8.0       vctrs_0.6.1                  </span></span>
<span><span class="co">##  [49] abind_1.4-5                   cachem_1.0.7                 </span></span>
<span><span class="co">##  [51] withr_2.5.0                   emmeans_1.8.5                </span></span>
<span><span class="co">##  [53] sctransform_0.3.5             prettyunits_1.1.1            </span></span>
<span><span class="co">##  [55] softImpute_1.4-1              cluster_2.1.4                </span></span>
<span><span class="co">##  [57] crayon_1.5.2                  genefilter_1.78.0            </span></span>
<span><span class="co">##  [59] edgeR_3.38.4                  pkgconfig_2.0.3              </span></span>
<span><span class="co">##  [61] nlme_3.1-160                  vipor_0.4.5                  </span></span>
<span><span class="co">##  [63] blme_1.0-5                    rlang_1.1.0                  </span></span>
<span><span class="co">##  [65] globals_0.16.2                lifecycle_1.0.3              </span></span>
<span><span class="co">##  [67] sandwich_3.0-2                filelock_1.0.2               </span></span>
<span><span class="co">##  [69] rsvd_1.0.5                    invgamma_1.1                 </span></span>
<span><span class="co">##  [71] rprojroot_2.0.3               graph_1.74.0                 </span></span>
<span><span class="co">##  [73] Matrix_1.5-3                  ashr_2.2-54                  </span></span>
<span><span class="co">##  [75] boot_1.3-28.1                 zoo_1.8-11                   </span></span>
<span><span class="co">##  [77] beeswarm_0.4.0                GlobalOptions_0.1.2          </span></span>
<span><span class="co">##  [79] png_0.1-8                     viridisLite_0.4.1            </span></span>
<span><span class="co">##  [81] rjson_0.2.21                  bitops_1.0-7                 </span></span>
<span><span class="co">##  [83] KernSmooth_2.23-20            EnrichmentBrowser_2.26.0     </span></span>
<span><span class="co">##  [85] Biostrings_2.64.1             blob_1.2.3                   </span></span>
<span><span class="co">##  [87] DelayedMatrixStats_1.18.2     shape_1.4.6                  </span></span>
<span><span class="co">##  [89] mixsqp_0.3-48                 stringr_1.5.0                </span></span>
<span><span class="co">##  [91] SQUAREM_2021.1                parallelly_1.34.0            </span></span>
<span><span class="co">##  [93] remaCor_0.0.11                rmeta_3.0                    </span></span>
<span><span class="co">##  [95] beachmat_2.12.0               scales_1.2.1                 </span></span>
<span><span class="co">##  [97] memoise_2.0.1                 GSEABase_1.58.0              </span></span>
<span><span class="co">##  [99] magrittr_2.0.3                plyr_1.8.8                   </span></span>
<span><span class="co">## [101] gplots_3.1.3                  zlibbioc_1.42.0              </span></span>
<span><span class="co">## [103] compiler_4.2.0                RColorBrewer_1.1-3           </span></span>
<span><span class="co">## [105] clue_0.3-64                   KEGGgraph_1.56.0             </span></span>
<span><span class="co">## [107] lme4_1.1-33                   DESeq2_1.36.0                </span></span>
<span><span class="co">## [109] cli_3.6.1                     XVector_0.36.0               </span></span>
<span><span class="co">## [111] lmerTest_3.1-3                listenv_0.9.0                </span></span>
<span><span class="co">## [113] TMB_1.9.1                     MASS_7.3-58.3                </span></span>
<span><span class="co">## [115] tidyselect_1.2.0              stringi_1.7.12               </span></span>
<span><span class="co">## [117] textshaping_0.3.6             zenith_1.1.2                 </span></span>
<span><span class="co">## [119] yaml_2.3.7                    BiocSingular_1.12.0          </span></span>
<span><span class="co">## [121] locfit_1.5-9.7                ggrepel_0.9.3                </span></span>
<span><span class="co">## [123] grid_4.2.0                    sass_0.4.5                   </span></span>
<span><span class="co">## [125] tools_4.2.0                   future.apply_1.10.0          </span></span>
<span><span class="co">## [127] parallel_4.2.0                circlize_0.4.15              </span></span>
<span><span class="co">## [129] foreach_1.5.2                 gridExtra_2.3                </span></span>
<span><span class="co">## [131] RcppZiggurat_0.1.6            digest_0.6.31                </span></span>
<span><span class="co">## [133] BiocManager_1.30.20           shiny_1.7.4                  </span></span>
<span><span class="co">## [135] Rcpp_1.0.10                   broom_1.0.4                  </span></span>
<span><span class="co">## [137] later_1.3.0                   BiocVersion_3.15.2           </span></span>
<span><span class="co">## [139] httr_1.4.5                    AnnotationDbi_1.58.0         </span></span>
<span><span class="co">## [141] ComplexHeatmap_2.12.1         Rdpack_2.4                   </span></span>
<span><span class="co">## [143] colorspace_2.1-0              XML_3.99-0.13                </span></span>
<span><span class="co">## [145] fs_1.6.1                      truncnorm_1.0-8              </span></span>
<span><span class="co">## [147] splines_4.2.0                 pkgdown_2.0.7                </span></span>
<span><span class="co">## [149] systemfonts_1.0.4             xtable_1.8-4                 </span></span>
<span><span class="co">## [151] jsonlite_1.8.4                nloptr_2.0.3                 </span></span>
<span><span class="co">## [153] Rfast_2.0.6                   R6_2.5.1                     </span></span>
<span><span class="co">## [155] RUnit_0.4.32                  mime_0.12                    </span></span>
<span><span class="co">## [157] pillar_1.9.0                  htmltools_0.5.4              </span></span>
<span><span class="co">## [159] glue_1.6.2                    fastmap_1.1.1                </span></span>
<span><span class="co">## [161] minqa_1.2.5                   BiocNeighbors_1.14.0         </span></span>
<span><span class="co">## [163] interactiveDisplayBase_1.34.0 codetools_0.2-19             </span></span>
<span><span class="co">## [165] mvtnorm_1.1-3                 utf8_1.2.3                   </span></span>
<span><span class="co">## [167] lattice_0.20-45               bslib_0.4.2                  </span></span>
<span><span class="co">## [169] tibble_3.2.1                  numDeriv_2016.8-1.1          </span></span>
<span><span class="co">## [171] pbkrtest_0.5.2                curl_5.0.0                   </span></span>
<span><span class="co">## [173] ggbeeswarm_0.7.1              gtools_3.9.4                 </span></span>
<span><span class="co">## [175] survival_3.5-5                glmmTMB_1.1.5                </span></span>
<span><span class="co">## [177] rmarkdown_2.20                desc_1.4.2                   </span></span>
<span><span class="co">## [179] munsell_0.5.0                 GetoptLong_1.0.5             </span></span>
<span><span class="co">## [181] GenomeInfoDbData_1.2.8        iterators_1.0.14             </span></span>
<span><span class="co">## [183] reshape2_1.4.4                gtable_0.3.1                 </span></span>
<span><span class="co">## [185] msigdbr_7.5.1                 rbibutils_2.2.13</span></span></code></pre>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
