<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Aggregation of single-cell to pseudobulk data — aggregateToPseudoBulk • dreamlet</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Aggregation of single-cell to pseudobulk data — aggregateToPseudoBulk"><meta name="description" content="Aggregation of single-cell to pseudobulk data.  Adapted from muscat::aggregateData and has same syntax and results.  But can be much faster for SingleCellExperiment backed by H5AD files using on-disk storage."><meta property="og:description" content="Aggregation of single-cell to pseudobulk data.  Adapted from muscat::aggregateData and has same syntax and results.  But can be much faster for SingleCellExperiment backed by H5AD files using on-disk storage."><meta property="og:image" content="http://DiseaseNeurogenomics.github.io/dreamlet/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dreamlet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.4.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/dreamlet.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/mashr.html">mashr analysis after dreamlet</a></li>
    <li><a class="dropdown-item" href="../articles/cell_covs.html">Modeling continuous cell-level covariates</a></li>
    <li><a class="dropdown-item" href="../articles/non_lin_eff.html">Testing non-linear effects</a></li>
    <li><a class="dropdown-item" href="../articles/h5ad_on_disk.html">Handling large H5AD datasets</a></li>
    <li><a class="dropdown-item" href="../articles/errors.html">Error handling</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseNeurogenomics/dreamlet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Aggregation of single-cell to pseudobulk data</h1>
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeurogenomics/dreamlet/blob/HEAD/R/aggregateToPseudoBulk.R" class="external-link"><code>R/aggregateToPseudoBulk.R</code></a></small>
      <div class="d-none name"><code>aggregateToPseudoBulk.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Aggregation of single-cell to pseudobulk data.  Adapted from <code><a href="https://rdrr.io/pkg/muscat/man/aggregateData.html" class="external-link">muscat::aggregateData</a></code> and has same syntax and results.  But can be much faster for <code>SingleCellExperiment</code> backed by H5AD files using on-disk storage.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">aggregateToPseudoBulk</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  assay <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cluster_id <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fun <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sum"</span>, <span class="st">"mean"</span>, <span class="st">"median"</span>, <span class="st">"prop.detected"</span>, <span class="st">"num.detected"</span>, <span class="st">"sem"</span>, <span class="st">"number"</span><span class="op">)</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span>progressbar <span class="op">=</span> <span class="va">verbose</span><span class="op">)</span>,</span>
<span>  checkValues <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  h5adBlockSizes <span class="op">=</span> <span class="fl">1e+09</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>a <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></code>.</p></dd>


<dt id="arg-assay">assay<a class="anchor" aria-label="anchor" href="#arg-assay"></a></dt>
<dd><p>character string specifying the assay slot to use as input data. Defaults to the 1st available (<code>assayNames(x)[1]</code>).</p></dd>


<dt id="arg-sample-id">sample_id<a class="anchor" aria-label="anchor" href="#arg-sample-id"></a></dt>
<dd><p>character string specifying which variable to use as sample id</p></dd>


<dt id="arg-cluster-id">cluster_id<a class="anchor" aria-label="anchor" href="#arg-cluster-id"></a></dt>
<dd><p>character string specifying which variable to use as cluster id</p></dd>


<dt id="arg-fun">fun<a class="anchor" aria-label="anchor" href="#arg-fun"></a></dt>
<dd><p>a character string.
Specifies the function to use as summary statistic.
Passed to <code>summarizeAssayByGroup2</code>.</p></dd>


<dt id="arg-scale">scale<a class="anchor" aria-label="anchor" href="#arg-scale"></a></dt>
<dd><p>logical. Should pseudo-bulks be scaled
with the effective library size &amp; multiplied by 1M?</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>logical. Should information on progress be reported?</p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>a <code><a href="https://rdrr.io/pkg/BiocParallel/man/BiocParallelParam-class.html" class="external-link">BiocParallelParam</a></code>
object specifying how aggregation should be parallelized.</p></dd>


<dt id="arg-checkvalues">checkValues<a class="anchor" aria-label="anchor" href="#arg-checkvalues"></a></dt>
<dd><p>logical. Should we check that signal values are positive integers?</p></dd>


<dt id="arg-h-adblocksizes">h5adBlockSizes<a class="anchor" aria-label="anchor" href="#arg-h-adblocksizes"></a></dt>
<dd><p>set the automatic block size block size (in bytes) for DelayedArray to read an H5AD file.  Larger values use more memory but are faster.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></code>.</p>
<p>Aggregation parameters (<code>assay, by, fun, scaled</code>) are stored in  <code>metadata()$agg_pars</code>, where <code>by = c(cluster_id, sample_id)</code>.  The number of cells that were aggregated are accessible in <code>int_colData()$n_cells</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Adapted from <code><a href="https://rdrr.io/pkg/muscat/man/aggregateData.html" class="external-link">muscat::aggregateData</a></code> and has similar syntax and same results.  This is much faster for <code>SingleCellExperiment</code> backed by H5AD files using <code>DelayedMatrix</code> because this summarizes counts using <code><a href="https://rdrr.io/pkg/DelayedMatrixStats/man/DelayedMatrixStats.html" class="external-link">DelayedMatrixStats</a></code>.  But this function also includes optmizations for <code>sparseMatrix</code> used by <code>Seurat</code> by using <code>sparseMatrixStats</code>.</p>
<p>Keeps variables from <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code> that are constant within <code>sample_id</code>.  For example, sex will be constant for all cells from the same <code>sample_id</code>, so it is retained as a variable in the pseudobulk result.  But number of expressed genes varies across cells within each <code>sample_id</code>, so it is dropped from <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>.  Instead the mean value per cell type is stored in <code>metadata(pb)$aggr_means</code>, and these can be included in regression formulas downstream.  In that case, the value of the covariates used per sample will depend on the cell type analyzed.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Crowell, HL, Soneson, C, Germain, P-L, Calini, D,
Collin, L, Raposo, C, Malhotra, D &amp; Robinson, MD:
Muscat detects subpopulation-specific state transitions from multi-sample multi-condition single-cell transcriptomics data.
<em>Nature Communications</em> <strong>11(1):6077</strong> (2020).
doi: <a href="https://doi.org/10.1038/s41467-020-19894-4" class="external-link">https://doi.org/10.1038/s41467-020-19894-4</a></p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Gabriel Hoffman, Helena L Crowell &amp; Mark D Robinson</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HelenaLC/muscat" class="external-link">muscat</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create pseudobulk for each sample and cell cluster</span></span></span>
<span class="r-in"><span><span class="va">pb</span> <span class="op">&lt;-</span> <span class="fu">aggregateToPseudoBulk</span><span class="op">(</span><span class="va">example_sce</span>,</span></span>
<span class="r-in"><span>  assay <span class="op">=</span> <span class="st">"counts"</span>,</span></span>
<span class="r-in"><span>  cluster_id <span class="op">=</span> <span class="st">"cluster_id"</span>,</span></span>
<span class="r-in"><span>  sample_id <span class="op">=</span> <span class="st">"sample_id"</span>,</span></span>
<span class="r-in"><span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># pseudobulk data from each cell type</span></span></span>
<span class="r-in"><span><span class="co"># is stored as its own assay</span></span></span>
<span class="r-in"><span><span class="va">pb</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SingleCellExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 1267 4 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(3): experiment_info agg_pars aggr_means</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(5): B cells CD14+ Monocytes CD4 T cells CD8 T cells FCGR3A+</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Monocytes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(1267): HES4 ISG15 ... CSTB PRMT2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(2): ENSEMBL SYMBOL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(4): ctrl101 ctrl107 stim101 stim107</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(1): group_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> reducedDimNames(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> mainExpName: NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> altExpNames(0):</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># aggregate by cluster only,</span></span></span>
<span class="r-in"><span><span class="co"># collapsing all samples into the same pseudobulk</span></span></span>
<span class="r-in"><span><span class="va">pb2</span> <span class="op">&lt;-</span> <span class="fu">aggregateToPseudoBulk</span><span class="op">(</span><span class="va">example_sce</span>, </span></span>
<span class="r-in"><span> cluster_id <span class="op">=</span> <span class="st">"cluster_id"</span>, </span></span>
<span class="r-in"><span> verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">pb2</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SingleCellExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 1267 5 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(3): experiment_info agg_pars aggr_means</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(1): ''</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(1267): HES4 ISG15 ... CSTB PRMT2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(2): ENSEMBL SYMBOL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(5): B cells CD14+ Monocytes CD4 T cells CD8 T cells FCGR3A+</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Monocytes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> reducedDimNames(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> mainExpName: NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> altExpNames(0):</span>
<span class="r-in"><span><span class="co">#</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

