<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Test differential variance — diffVar,dreamletResult-method • dreamlet</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Test differential variance — diffVar,dreamletResult-method"><meta name="description" content="Test the association between a covariate of interest and the response's deviation from expectation."><meta property="og:description" content="Test the association between a covariate of interest and the response's deviation from expectation."><meta property="og:image" content="http://DiseaseNeurogenomics.github.io/dreamlet/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dreamlet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.4.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/dreamlet.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/mashr.html">mashr analysis after dreamlet</a></li>
    <li><a class="dropdown-item" href="../articles/cell_covs.html">Modeling continuous cell-level covariates</a></li>
    <li><a class="dropdown-item" href="../articles/non_lin_eff.html">Testing non-linear effects</a></li>
    <li><a class="dropdown-item" href="../articles/h5ad_on_disk.html">Handling large H5AD datasets</a></li>
    <li><a class="dropdown-item" href="../articles/errors.html">Error handling</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseNeurogenomics/dreamlet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Test differential variance</h1>
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeurogenomics/dreamlet/blob/HEAD/R/diffVar.R" class="external-link"><code>R/diffVar.R</code></a></small>
      <div class="d-none name"><code>diffVar-methods.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Test the association between a covariate of interest and the response's deviation from expectation.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for class 'dreamletResult'</span></span>
<span><span class="fu">diffVar</span><span class="op">(</span></span>
<span>  <span class="va">fit</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AD"</span>, <span class="st">"SQ"</span><span class="op">)</span>,</span>
<span>  scale <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"leverage"</span>, <span class="st">"none"</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-fit">fit<a class="anchor" aria-label="anchor" href="#arg-fit"></a></dt>
<dd><p>model fit from <code>dream()</code></p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>transform the residuals using absolute deviation ("AD") or squared deviation ("SQ").</p></dd>


<dt id="arg-scale">scale<a class="anchor" aria-label="anchor" href="#arg-scale"></a></dt>
<dd><p>scale each observation by "leverage", or no scaling ("none")</p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>parameters for parallel evaluation</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>other parameters passed to <code>dream()</code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This method performs a test of differential variance between two subsets of the data, in a way that generalizes to multiple categories, continuous variables and metrics of spread beyond variance.  For the two category test, this method is simular to Levene's test.  This model was adapted from Phipson, et al (2014), extended to linear mixed models, and adapted to be compatible with <code><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/dream-method.html" class="external-link">variancePartition::dream()</a></code> and <code><a href="dreamlet.html">dreamlet::dreamlet()</a></code>.</p>
<p>This method is composed of multiple steps where 1) a typical linear (mixed) model is fit with <code><a href="dreamlet.html">dreamlet()</a></code>, 2) residuals are computed and transformed based on an absolute value or squaring transform, 3) a second regression is performed with <code><a href="dreamlet.html">dreamlet()</a></code> to test if a variable is associated with increased deviation from expectation.  Both regression take advantage of the <code><a href="dreamlet.html">dreamlet()</a></code> linear (mixed) modelling framework followed by empirical Bayes shrinkage that extends the <code><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">limma::voom()</a></code> framework.</p>
<p>Note that <code>diffVar()</code> takes the results of the first regression as a parameter to use as a starting point.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>phipson2014diffvarvariancePartition</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/diffVar-method.html" class="external-link">variancePartition::diffVar()</a></code></p>
<p><code><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/diffVar-method.html" class="external-link">variancePartition::diffVar()</a></code>, <code>missMethyl::diffVar()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HelenaLC/muscat" class="external-link">muscat</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create pseudobulk for each sample and cell cluster</span></span></span>
<span class="r-in"><span><span class="va">pb</span> <span class="op">&lt;-</span> <span class="fu"><a href="aggregateToPseudoBulk.html">aggregateToPseudoBulk</a></span><span class="op">(</span><span class="va">example_sce</span>,</span></span>
<span class="r-in"><span>  assay <span class="op">=</span> <span class="st">"counts"</span>,</span></span>
<span class="r-in"><span>  cluster_id <span class="op">=</span> <span class="st">"cluster_id"</span>,</span></span>
<span class="r-in"><span>  sample_id <span class="op">=</span> <span class="st">"sample_id"</span>,</span></span>
<span class="r-in"><span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># voom-style normalization</span></span></span>
<span class="r-in"><span><span class="va">res.proc</span> <span class="op">&lt;-</span> <span class="fu"><a href="processAssays.html">processAssays</a></span><span class="op">(</span><span class="va">pb</span>, <span class="op">~</span><span class="va">group_id</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   B cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.22 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD14+ Monocytes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.33 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD4 T cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.24 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD8 T cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.14 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   FCGR3A+ Monocytes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.29 secs</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Differential expression analysis within each assay,</span></span></span>
<span class="r-in"><span><span class="co"># evaluated on the voom normalized data</span></span></span>
<span class="r-in"><span><span class="va">res.dl</span> <span class="op">&lt;-</span> <span class="fu"><a href="dreamlet.html">dreamlet</a></span><span class="op">(</span><span class="va">res.proc</span>, <span class="op">~</span><span class="va">group_id</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   B cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.23 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD14+ Monocytes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.27 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD4 T cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.24 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD8 T cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.14 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   FCGR3A+ Monocytes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.25 secs</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Differential variance analysis</span></span></span>
<span class="r-in"><span><span class="co"># result is a dreamlet fit</span></span></span>
<span class="r-in"><span><span class="va">res.dvar</span> <span class="op">&lt;-</span> <span class="fu">diffVar</span><span class="op">(</span><span class="va">res.dl</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Examine results</span></span></span>
<span class="r-in"><span><span class="va">res.dvar</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: dreamletResult </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(5): B cells CD14+ Monocytes CD4 T cells CD8 T cells FCGR3A+</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Monocytes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genes:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  min: 531 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  max: 1130 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> details(7): assay n_retain ... n_errors error_initial</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> coefNames(2): (Intercept) group_idstim</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Examine details for each assay</span></span></span>
<span class="r-in"><span><span class="fu"><a href="details-methods.html">details</a></span><span class="op">(</span><span class="va">res.dvar</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>               assay n_retain   formula formDropsTerms n_genes n_errors</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1           B cells        4 ~group_id          FALSE     847        0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   CD14+ Monocytes        4 ~group_id          FALSE    1130        0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       CD4 T cells        4 ~group_id          FALSE     897        0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       CD8 T cells        4 ~group_id          FALSE     531        0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 FCGR3A+ Monocytes        4 ~group_id          FALSE    1086        0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   error_initial</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1         FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2         FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3         FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5         FALSE</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># show coefficients estimated for each cell type</span></span></span>
<span class="r-in"><span><span class="fu"><a href="coefNames-methods.html">coefNames</a></span><span class="op">(</span><span class="va">res.dvar</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "(Intercept)"  "group_idstim"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># extract results using limma-style syntax</span></span></span>
<span class="r-in"><span><span class="co"># combines all cell types together</span></span></span>
<span class="r-in"><span><span class="co"># adj.P.Val gives study-wide FDR</span></span></span>
<span class="r-in"><span><span class="fu">topTable</span><span class="op">(</span><span class="va">res.dvar</span>, coef <span class="op">=</span> <span class="st">"group_idstim"</span>, number <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 3 rows and 9 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         assay          ID     logFC   AveExpr         t     P.Value   adj.P.Val</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   &lt;character&gt; &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 CD4 T cells     DYNLRB1  0.978587  0.525150   1179.42 1.28303e-11 1.91186e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 CD4 T cells       RNPS1  1.103936  0.686623   1002.75 2.36305e-11 1.91186e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 CD4 T cells       SLFN5  0.818186  0.444949    964.08 2.74004e-11 1.91186e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           B     z.std</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   &lt;numeric&gt; &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   18.1794   1179.42</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   17.6567   1002.75</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3   17.5220    964.08</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Plot top hit to see differential variance</span></span></span>
<span class="r-in"><span><span class="co"># Note that this is a toy example with only 4 samples</span></span></span>
<span class="r-in"><span><span class="va">cellType</span> <span class="op">&lt;-</span> <span class="st">"CD4 T cells"</span></span></span>
<span class="r-in"><span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="st">"DYNLRB1"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">res.proc</span><span class="op">[[</span><span class="va">cellType</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">E</span><span class="op">[</span><span class="va">gene</span>, <span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">res.proc</span><span class="op">)</span><span class="op">$</span><span class="va">group_id</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="va">y</span> <span class="op">~</span> <span class="va">x</span>,</span></span>
<span class="r-in"><span>  xlab <span class="op">=</span> <span class="st">"Stimulation status"</span>,</span></span>
<span class="r-in"><span>  ylab <span class="op">=</span> <span class="st">"Gene expression"</span>,</span></span>
<span class="r-in"><span>  main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">cellType</span>, <span class="va">gene</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="diffVar-methods-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co">#</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

