<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Perform differential expression between a pair of assays using linear (mixed) models"><title>Differential expression between pair of assays — dreamletCompareClusters • dreamlet</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Differential expression between pair of assays — dreamletCompareClusters"><meta property="og:description" content="Perform differential expression between a pair of assays using linear (mixed) models"><meta property="og:image" content="http://DiseaseNeurogenomics.github.io/dreamlet/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dreamlet</a>

<<<<<<< HEAD
    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.1</small>
=======
    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.32</small>
>>>>>>> a2b5f071aa39d4fcc963cf98c450727ccecdb45e

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/dreamlet.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/cell_covs.html">Modeling continuous cell-level covariates</a>
<<<<<<< HEAD
=======
    <a class="dropdown-item" href="../articles/errors.html">Error handling</a>
>>>>>>> a2b5f071aa39d4fcc963cf98c450727ccecdb45e
    <a class="dropdown-item" href="../articles/h5ad_on_disk.html">Handling large H5AD datasets</a>
    <a class="dropdown-item" href="../articles/mashr.html">mashr analysis after dreamlet</a>
    <a class="dropdown-item" href="../articles/non_lin_eff.html">Testing non-linear effects</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/DiseaseNeurogenomics/dreamlet/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Differential expression between pair of assays</h1>
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeurogenomics/dreamlet/blob/HEAD/R/dreamletCompareClusters.R" class="external-link"><code>R/dreamletCompareClusters.R</code></a></small>
      <div class="d-none name"><code>dreamletCompareClusters.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Perform differential expression between a pair of assays using linear (mixed) models</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">dreamletCompareClusters</span><span class="op">(</span></span>
<span>  <span class="va">pb</span>,</span>
<span>  <span class="va">assays</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"fixed"</span>, <span class="st">"random"</span>, <span class="st">"none"</span><span class="op">)</span>,</span>
<span>  formula <span class="op">=</span> <span class="op">~</span><span class="fl">0</span>,</span>
<span>  collapse <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  min.cells <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  min.count <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  min.samples <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  isCounts <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  normalize.method <span class="op">=</span> <span class="st">"TMM"</span>,</span>
<<<<<<< HEAD
=======
<span>  useCountsWeights <span class="op">=</span> <span class="cn">TRUE</span>,</span>
>>>>>>> a2b5f071aa39d4fcc963cf98c450727ccecdb45e
<span>  robust <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>compare <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"cellClustertest - cellClusterbaseline"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  errorsAsWarnings <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>pb</dt>
<dd><p>pseudobulk data as SingleCellExperiment object</p></dd>


<dt>assays</dt>
<dd><p>array of two entries specifying assays (i.e. cell clusters) to compare, or a list of two sets of assays.</p></dd>


<dt>method</dt>
<dd><p>account for repeated measures from donors using a "random" effect, a "fixed" effect, or "none"</p></dd>


<dt>formula</dt>
<dd><p>covariates to include in the analysis.</p></dd>


<dt>collapse</dt>
<dd><p>if TRUE (default), combine all cell clusters within the test set, and separately the baseline set. If FALSE, estimate coefficient for each cell cluster and then identify differential expression using linear contrasts with <code><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/makeContrastsDream.html" class="external-link">variancePartition::makeContrastsDream()</a></code></p></dd>


<dt>min.cells</dt>
<dd><p>minimum number of observed cells for a sample to be included in the analysis</p></dd>


<dt>min.count</dt>
<dd><p>minimum number of reads for a gene to be consider expressed in a sample.  Passed to <code><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">edgeR::filterByExpr</a></code></p></dd>


<dt>min.samples</dt>
<dd><p>minimum number of samples passing cutoffs for cell cluster to be retained</p></dd>


<dt>isCounts</dt>
<dd><p>logical, indicating if data is raw counts</p></dd>


<dt>normalize.method</dt>
<dd><p>normalization method to be used by <code>calcNormFactors</code></p></dd>


<<<<<<< HEAD
=======
<dt>useCountsWeights</dt>
<dd><p>use cell count weights</p></dd>


>>>>>>> a2b5f071aa39d4fcc963cf98c450727ccecdb45e
<dt>robust</dt>
<dd><p>logical, use eBayes method that is robust to outlier genes</p></dd>


<dt>quiet</dt>
<dd><p>show messages</p></dd>


<dt>contrasts</dt>
<dd><p>cell type is encoded in variable <code>cellCluster</code> with levels <code>test</code> and <code>baseline</code>. <code>contrasts</code> specifies contrasts passed to <code><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/makeContrastsDream.html" class="external-link">variancePartition::makeContrastsDream()</a></code>.  Note, advanced users only.</p></dd>


<dt>BPPARAM</dt>
<dd><p>parameters for parallel evaluation</p></dd>


<dt>errorsAsWarnings</dt>
<dd><p>if <code>TRUE</code>, convert error to a warning and return <code>NULL</code></p></dd>


<dt>...</dt>
<dd><p>other arguments passed to <code>dream</code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Object of class <code>dreamletResult</code> storing results for each comparison</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Analyze pseudobulk data to identify differential gene expression between two cell clusters or sets of clusters while modeling the cross-donor expression variation and other aspects of the study design.</p>
<p><code>dreamletCompareClusters()</code> is useful for finding genes that are differentially expressed betweeen cell clusters and estimating their fold change.  However, the p-values and number of differentially expressed genes are problematic for two reasons, so users must be careful not to overinterpret them:</p>
<ol><li><p>Cell clusters are typically identified with the same gene expression data used for this differential expression analysis between clusters. The same data is used both for discovery and testing, and this means that the p-values from the differential expression analysis will not be uniform under the null.  This will produce a lot of findings with small p-values even in the absence of true biological differences.</p></li>
<li><p>The <code>dreamlet</code> package is designed for large datasets with many subjects.  The sample sizes from cohort studies are an order of magnitude larger than typical single cell studies.  This means that these analyses have huge power to detect even subtle difference in expression between cell clusters.  While cluster-specific marker genes are often discovered from an handful of samples, the <code>dreamlet</code> package is applicable to 100s or 1000s of subjects.</p></li>
</ol><p><code>method</code> indicates the regression method used to test differential expression between sets of cell clusters.  Since the same biosample will usually be represented in both sets of cell clusters, <code>method</code> determines how the paired design is modeled.   For <code>method = "mixed"</code>, the sample is modeled as a random effect: <code>~ (1|Sample) + ...</code>. For <code>method = "fixed"</code>, the sample is modeled as a fixed effect: <code>~ Sample + ...</code>. For <code>method = "none"</code>, the pairing is ignored.</p>
<p>When <code>collapse=TRUE</code> (default) combine all cell clusters within the test set, and separately the baseline set, and estimate a coefficient indicating the differential expression between sets for a given gene.  If <code>collapse=FALSE</code>, estimate a coefficient for each cell type and then identify differential expression using linear contrasts with <code><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/makeContrastsDream.html" class="external-link">variancePartition::makeContrastsDream()</a></code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HelenaLC/muscat" class="external-link">muscat</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create pseudobulk for each sample and cell cluster</span></span></span>
<span class="r-in"><span><span class="va">pb</span> <span class="op">&lt;-</span> <span class="fu"><a href="aggregateToPseudoBulk.html">aggregateToPseudoBulk</a></span><span class="op">(</span><span class="va">example_sce</span>,</span></span>
<span class="r-in"><span>  assay <span class="op">=</span> <span class="st">"counts"</span>,</span></span>
<span class="r-in"><span>  cluster_id <span class="op">=</span> <span class="st">"cluster_id"</span>,</span></span>
<span class="r-in"><span>  sample_id <span class="op">=</span> <span class="st">"sample_id"</span>,</span></span>
<span class="r-in"><span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Evaluate the specificity of each gene for each cluster</span></span></span>
<span class="r-in"><span><span class="va">df_cts</span> <span class="op">&lt;-</span> <span class="fu"><a href="cellTypeSpecificity.html">cellTypeSpecificity</a></span><span class="op">(</span><span class="va">pb</span><span class="op">)</span></span></span>
<<<<<<< HEAD
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = TRUE.</span>
=======
>>>>>>> a2b5f071aa39d4fcc963cf98c450727ccecdb45e
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># compare first two assays (i.e. cell types)</span></span></span>
<span class="r-in"><span><span class="va">ct.pairs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B cells"</span>, <span class="st">"CD14+ Monocytes"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># run comparison</span></span></span>
<span class="r-in"><span><span class="co"># use method = 'fixed' here since it is faster</span></span></span>
<span class="r-in"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">dreamletCompareClusters</span><span class="op">(</span><span class="va">pb</span>, <span class="va">ct.pairs</span>, method <span class="op">=</span> <span class="st">"fixed"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Initial filtering...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering for paired samples...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Extract top 10 differentially expressed genes</span></span></span>
<span class="r-in"><span><span class="co"># The coefficient 'compare' is the value logFC between test and baseline:</span></span></span>
<span class="r-in"><span><span class="co"># compare = cellClustertest - cellClusterbaseline</span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">topTable</span><span class="op">(</span><span class="va">fit</span>, coef <span class="op">=</span> <span class="st">"compare"</span>, number <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># genes with highest logFC are most highly expressed in</span></span></span>
<span class="r-in"><span><span class="co"># B cells compared to CD14+ Monocytes</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             logFC   AveExpr         t      P.Value    adj.P.Val        B</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FTL     -4.559570 14.265849 -24.93241 4.100814e-11 1.394120e-08 16.14202</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> LGALS3  -4.815545  9.879829 -24.76997 4.404803e-11 1.394120e-08 15.89842</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> S100A10 -4.551858  9.950588 -21.69418 1.871963e-10 3.822981e-08 14.56393</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> S100A11 -4.687559 10.362616 -21.19151 2.415786e-10 3.822981e-08 14.33464</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CD63    -3.892389 10.337664 -20.05814 4.388171e-10 5.555424e-08 13.77198</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ANXA5   -3.674191 10.097965 -19.29905 6.665193e-10 7.031778e-08 13.35939</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">dreamlet</span><span class="fu">::</span><span class="fu"><a href="plotHeatmap-methods.html">plotHeatmap</a></span><span class="op">(</span><span class="va">df_cts</span>, genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></span>
<<<<<<< HEAD
<span class="r-msg co"><span class="r-pr">#&gt;</span> Scale for <span style="color: #00BB00;">fill</span> is already present.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span>
=======
>>>>>>> a2b5f071aa39d4fcc963cf98c450727ccecdb45e
<span class="r-plt img"><img src="dreamletCompareClusters-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># compare B cells versus the rest of the cell types</span></span></span>
<span class="r-in"><span><span class="co"># 'rest' is a keyword indicating all other assays</span></span></span>
<span class="r-in"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">dreamletCompareClusters</span><span class="op">(</span><span class="va">pb</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B cells"</span>, <span class="st">"rest"</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"fixed"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Initial filtering...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering for paired samples...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">topTable</span><span class="op">(</span><span class="va">fit</span>, coef <span class="op">=</span> <span class="st">"compare"</span>, number <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># genes with highest logFC are most highly expressed in</span></span></span>
<span class="r-in"><span><span class="co"># B cells compared to all others</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             logFC   AveExpr         t      P.Value    adj.P.Val        B</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FTL     -3.600925 13.872811 -33.28366 1.901955e-15 1.690838e-12 25.84574</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FTH1    -2.466503 15.497594 -28.32886 2.047580e-14 8.003456e-12 23.27441</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ANXA1   -5.766836  8.777334 -27.80028 2.700829e-14 8.003456e-12 22.77952</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> LGALS3  -3.906317  9.492690 -21.89782 8.854358e-13 1.820012e-10 19.63566</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> S100A11 -4.195517 10.182616 -21.68039 1.023629e-12 1.820012e-10 19.53932</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> S100A10 -3.753134  9.613494 -20.50358 2.299888e-12 3.407668e-10 18.72582</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get genes upregulated in B cells</span></span></span>
<span class="r-in"><span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">res</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu">dreamlet</span><span class="fu">::</span><span class="fu"><a href="plotHeatmap-methods.html">plotHeatmap</a></span><span class="op">(</span><span class="va">df_cts</span>, genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span><span class="op">)</span></span></span>
<<<<<<< HEAD
<span class="r-msg co"><span class="r-pr">#&gt;</span> Scale for <span style="color: #00BB00;">fill</span> is already present.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span>
=======
>>>>>>> a2b5f071aa39d4fcc963cf98c450727ccecdb45e
<span class="r-plt img"><img src="dreamletCompareClusters-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">lst</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD14+ Monocytes"</span>, <span class="st">"FCGR3A+ Monocytes"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  baseline <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4 T cells"</span>, <span class="st">"CD8 T cells"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># compare 2 monocyte clusters to two T cell clusters</span></span></span>
<span class="r-in"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">dreamletCompareClusters</span><span class="op">(</span><span class="va">pb</span>, <span class="va">lst</span>, method <span class="op">=</span> <span class="st">"fixed"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Initial filtering...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering for paired samples...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">topTable</span><span class="op">(</span><span class="va">fit</span>, coef <span class="op">=</span> <span class="st">"compare"</span>, number <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># genes with highest logFC are most highly expressed in</span></span></span>
<span class="r-in"><span><span class="co"># monocytes compared to T cells</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                            logFC   AveExpr        t      P.Value    adj.P.Val</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FTL                     5.195892 13.603905 53.09308 1.337149e-12 1.159308e-09</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TIMP1                   6.214741 11.076844 41.00811 1.371256e-11 5.944395e-09</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HLA-DRA_ENSG00000204287 5.674527  9.655697 32.33465 1.161417e-10 3.356496e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> S100A4                  4.514156  9.249845 28.04576 4.159764e-10 7.937568e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> PSAP                    3.948238  9.342828 27.30755 5.281552e-10 7.937568e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> LGALS1                  4.544995  9.290447 27.18795 5.493127e-10 7.937568e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                                B</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FTL                     19.61451</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> TIMP1                   17.20190</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> HLA-DRA_ENSG00000204287 15.06115</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> S100A4                  13.95058</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> PSAP                    13.73411</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> LGALS1                  13.66478</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Get genes upregulated in monocytes</span></span></span>
<span class="r-in"><span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">res</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu">dreamlet</span><span class="fu">::</span><span class="fu"><a href="plotHeatmap-methods.html">plotHeatmap</a></span><span class="op">(</span><span class="va">df_cts</span>, genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">[</span><span class="va">idx</span><span class="op">]</span><span class="op">)</span></span></span>
<<<<<<< HEAD
<span class="r-msg co"><span class="r-pr">#&gt;</span> Scale for <span style="color: #00BB00;">fill</span> is already present.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span>
=======
>>>>>>> a2b5f071aa39d4fcc963cf98c450727ccecdb45e
<span class="r-plt img"><img src="dreamletCompareClusters-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

