<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Compute precision weights for pseudobulk using the delta method to approximate the variance of the log2 counts per million considering variation in the number of cells and gene expression variance across cells within each sample. By default, used number of cells; if specified use delta method.  Note that processAssays() uses number of cells as weights when no weights are specificed"><title>Compute precision weights for pseudobulk — pbWeights • dreamlet</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Compute precision weights for pseudobulk — pbWeights"><meta property="og:description" content="Compute precision weights for pseudobulk using the delta method to approximate the variance of the log2 counts per million considering variation in the number of cells and gene expression variance across cells within each sample. By default, used number of cells; if specified use delta method.  Note that processAssays() uses number of cells as weights when no weights are specificed"><meta property="og:image" content="http://DiseaseNeurogenomics.github.io/dreamlet/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dreamlet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.14</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/dreamlet.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/mashr.html">mashr analysis after dreamlet</a>
    <a class="dropdown-item" href="../articles/cell_covs.html">Modeling continuous cell-level covariates</a>
    <a class="dropdown-item" href="../articles/non_lin_eff.html">Testing non-linear effects</a>
    <a class="dropdown-item" href="../articles/h5ad_on_disk.html">Handling large H5AD datasets</a>
    <a class="dropdown-item" href="../articles/errors.html">Error handling</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/DiseaseNeurogenomics/dreamlet/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Compute precision weights for pseudobulk</h1>
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeurogenomics/dreamlet/blob/HEAD/R/initialPrecisionWeights.R" class="external-link"><code>R/initialPrecisionWeights.R</code></a></small>
      <div class="d-none name"><code>pbWeights.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Compute precision weights for pseudobulk using the delta method to approximate the variance of the log2 counts per million considering variation in the number of cells and gene expression variance across cells within each sample. By default, used number of cells; if specified use delta method.  Note that <code><a href="processAssays.html">processAssays()</a></code> uses number of cells as weights when no weights are specificed</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">pbWeights</span><span class="op">(</span></span>
<span>  <span class="va">sce</span>,</span>
<span>  <span class="va">sample_id</span>,</span>
<span>  <span class="va">cluster_id</span>,</span>
<span>  <span class="va">geneList</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"delta"</span>, <span class="st">"ncells"</span><span class="op">)</span>,</span>
<span>  shrink <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  prior.count <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  maxRatio <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  h5adBlockSizes <span class="op">=</span> <span class="fl">1e+09</span>,</span>
<span>  details <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>sce</dt>
<dd><p><code>SingleCellExperiment</code> of where <code>counts(sce)</code> stores the raw count data at the single cell level</p></dd>


<dt>sample_id</dt>
<dd><p>character string specifying which variable to use as sample id</p></dd>


<dt>cluster_id</dt>
<dd><p>character string specifying which variable to use as cluster id</p></dd>


<dt>geneList</dt>
<dd><p>list of genes to be included for each cell type</p></dd>


<dt>method</dt>
<dd><p>select method to compute precision weights.  <code>'delta'</code> use the delta method based on normal approximation to a negative binomial model, slower but can increase power. <code>'ncells'</code> use the number of cells, this is faster; Subsequent arguments are ignored. Included for testing</p></dd>


<dt>shrink</dt>
<dd><p>Defaults to <code>TRUE</code>. Use empirical Bayes variance shrinkage from <code>limma</code> to shrink estimates of expression variance across cells within each sample</p></dd>


<dt>prior.count</dt>
<dd><p>Defaults to <code>0.5</code>. Count added to each observation at the pseudobulk level.  This is scaled but the number of cells before added to the cell level</p></dd>


<dt>maxRatio</dt>
<dd><p>When computing precision as the reciprocal of variance <code>1/(x+tau)</code> select tau to have a maximum ratio between the largest and smallest precision</p></dd>


<dt>h5adBlockSizes</dt>
<dd><p>set the automatic block size block size (in bytes) for DelayedArray to read an H5AD file.  Larger values use more memory but are faster.</p></dd>


<dt>details</dt>
<dd><p>include <code>data.frame</code> of cell-level statistics as <code>attr(., "details")</code></p></dd>


<dt>verbose</dt>
<dd><p>Show messages, defaults to TRUE</p></dd>

</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HelenaLC/muscat" class="external-link">muscat</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create pseudobulk for each sample and cell cluster</span></span></span>
<span class="r-in"><span><span class="va">pb</span> <span class="op">&lt;-</span> <span class="fu"><a href="aggregateToPseudoBulk.html">aggregateToPseudoBulk</a></span><span class="op">(</span><span class="va">example_sce</span>,</span></span>
<span class="r-in"><span>  assay <span class="op">=</span> <span class="st">"counts"</span>,</span></span>
<span class="r-in"><span>  sample_id <span class="op">=</span> <span class="st">"sample_id"</span>,</span></span>
<span class="r-in"><span>  cluster_id <span class="op">=</span> <span class="st">"cluster_id"</span>,</span></span>
<span class="r-in"><span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Gene expressed genes for each cell type</span></span></span>
<span class="r-in"><span><span class="va">geneList</span> <span class="op">=</span> <span class="fu"><a href="getExprGeneNames.html">getExprGeneNames</a></span><span class="op">(</span><span class="va">pb</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create precision weights for pseudobulk</span></span></span>
<span class="r-in"><span><span class="co"># By default, weights are set to cell count,</span></span></span>
<span class="r-in"><span><span class="co"># which is the default in processAssays()</span></span></span>
<span class="r-in"><span><span class="co"># even when no weights are specified</span></span></span>
<span class="r-in"><span><span class="va">weightsList</span> <span class="op">&lt;-</span> <span class="fu">pbWeights</span><span class="op">(</span><span class="va">example_sce</span>,</span></span>
<span class="r-in"><span>  sample_id <span class="op">=</span> <span class="st">"sample_id"</span>,</span></span>
<span class="r-in"><span>  cluster_id <span class="op">=</span> <span class="st">"cluster_id"</span>,</span></span>
<span class="r-in"><span>  geneList <span class="op">=</span> <span class="va">geneList</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Processing: B cells</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Computing library sizes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Processing samples...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Processing: CD14+ Monocytes</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Computing library sizes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Processing samples...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Processing: CD4 T cells</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Computing library sizes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Processing samples...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Processing: CD8 T cells</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Computing library sizes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Processing samples...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Processing: FCGR3A+ Monocytes</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Computing library sizes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Processing samples...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># voom-style normalization using initial weights</span></span></span>
<span class="r-in"><span><span class="va">res.proc</span> <span class="op">&lt;-</span> <span class="fu"><a href="processAssays.html">processAssays</a></span><span class="op">(</span><span class="va">pb</span>, <span class="op">~</span><span class="va">group_id</span>, weightsList <span class="op">=</span> <span class="va">weightsList</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   B cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.2 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD14+ Monocytes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.31 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD4 T cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.23 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD8 T cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.11 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   FCGR3A+ Monocytes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.32 secs</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

