<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Processing SingleCellExperiment to dreamletProcessedData — processAssays • dreamlet</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Processing SingleCellExperiment to dreamletProcessedData — processAssays"><meta name="description" content="For raw counts, estimate precision weights using linear mixed model weighting by number of cells observed for each sample.  For normalized data, only weight by number of cells."><meta property="og:description" content="For raw counts, estimate precision weights using linear mixed model weighting by number of cells observed for each sample.  For normalized data, only weight by number of cells."><meta property="og:image" content="http://DiseaseNeurogenomics.github.io/dreamlet/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dreamlet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/dreamlet.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/mashr.html">mashr analysis after dreamlet</a></li>
    <li><a class="dropdown-item" href="../articles/cell_covs.html">Modeling continuous cell-level covariates</a></li>
    <li><a class="dropdown-item" href="../articles/non_lin_eff.html">Testing non-linear effects</a></li>
    <li><a class="dropdown-item" href="../articles/h5ad_on_disk.html">Handling large H5AD datasets</a></li>
    <li><a class="dropdown-item" href="../articles/errors.html">Error handling</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseNeurogenomics/dreamlet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Processing SingleCellExperiment to dreamletProcessedData</h1>
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeurogenomics/dreamlet/blob/HEAD/R/processAssays.R" class="external-link"><code>R/processAssays.R</code></a></small>
      <div class="d-none name"><code>processAssays.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>For raw counts, estimate precision weights using linear mixed model weighting by number of cells observed for each sample.  For normalized data, only weight by number of cells.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">processAssays</span><span class="op">(</span></span>
<span>  <span class="va">sceObj</span>,</span>
<span>  <span class="va">formula</span>,</span>
<span>  assays <span class="op">=</span> <span class="fu">assayNames</span><span class="op">(</span><span class="va">sceObj</span><span class="op">)</span>,</span>
<span>  min.cells <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  min.count <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  min.samples <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  min.prop <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>  isCounts <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  normalize.method <span class="op">=</span> <span class="st">"TMM"</span>,</span>
<span>  span <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  weightsList <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-sceobj">sceObj<a class="anchor" aria-label="anchor" href="#arg-sceobj"></a></dt>
<dd><p>SingleCellExperiment object</p></dd>


<dt id="arg-formula">formula<a class="anchor" aria-label="anchor" href="#arg-formula"></a></dt>
<dd><p>regression formula for differential expression analysis</p></dd>


<dt id="arg-assays">assays<a class="anchor" aria-label="anchor" href="#arg-assays"></a></dt>
<dd><p>array of assay names to include in analysis. Defaults to <code>assayNames(sceObj)</code></p></dd>


<dt id="arg-min-cells">min.cells<a class="anchor" aria-label="anchor" href="#arg-min-cells"></a></dt>
<dd><p>minimum number of observed cells for a sample to be included in the analysis</p></dd>


<dt id="arg-min-count">min.count<a class="anchor" aria-label="anchor" href="#arg-min-count"></a></dt>
<dd><p>used to compute a CPM threshold of <code>CPM.cutoff = min.count/median(lib.size)*1e6</code>.  Passed to <code><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">edgeR::filterByExpr()</a></code></p></dd>


<dt id="arg-min-samples">min.samples<a class="anchor" aria-label="anchor" href="#arg-min-samples"></a></dt>
<dd><p>minimum number of samples passing cutoffs for cell cluster to be retained</p></dd>


<dt id="arg-min-prop">min.prop<a class="anchor" aria-label="anchor" href="#arg-min-prop"></a></dt>
<dd><p>minimum proportion of retained samples with <code>CPM &gt; CPM.cutoff</code></p></dd>


<dt id="arg-iscounts">isCounts<a class="anchor" aria-label="anchor" href="#arg-iscounts"></a></dt>
<dd><p>logical, indicating if data is raw counts</p></dd>


<dt id="arg-normalize-method">normalize.method<a class="anchor" aria-label="anchor" href="#arg-normalize-method"></a></dt>
<dd><p>normalization method to be used by <code>calcNormFactors</code></p></dd>


<dt id="arg-span">span<a class="anchor" aria-label="anchor" href="#arg-span"></a></dt>
<dd><p>Lowess smoothing parameter using by <code><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/voomWithDreamWeights.html" class="external-link">variancePartition::voomWithDreamWeights()</a></code></p></dd>


<dt id="arg-quiet">quiet<a class="anchor" aria-label="anchor" href="#arg-quiet"></a></dt>
<dd><p>show messages</p></dd>


<dt id="arg-weightslist">weightsList<a class="anchor" aria-label="anchor" href="#arg-weightslist"></a></dt>
<dd><p>list storing matrix of precision weights for each cell type. If <code>NULL</code> precision weights are set to 1</p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>parameters for parallel evaluation</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>other arguments passed to <code>dream</code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Object of class <code>dreamletProcessedData</code> storing voom-style normalized expression data</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>For each cell cluster, samples with at least <code>min.cells</code> are retained. Only clusters with at least <code>min.samples</code> retained samples are kept. Genes are retained if they have at least <code>min.count</code> reads in at least <code>min.prop</code> fraction of the samples.  Current values are reasonable defaults, since genes that don't pass these cutoffs are very underpowered for differential expression analysis and only increase the multiple testing burden.  But values of <code>min.cells = 2</code> and <code>min.count = 2</code> are also reasonable to include more genes in the analysis.</p>
<p>The precision weights are estimated using the residuals fit from the specified formula.  These weights are robust to changes in the formula as long as the major variables explaining the highest fraction of the variance are included.</p>
<p>If <code>weightsList</code> is <code>NULL</code>, precision weights are set to 1 internally.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code>voomWithDreamWeights()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HelenaLC/muscat" class="external-link">muscat</a></span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Package version inconsistency detected.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> glmmTMB was built with TMB version 1.9.14</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> Current TMB version is 1.9.15</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> Please re-install glmmTMB from source or restore original ‘TMB’ package (see '?reinstalling' for more information)</span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create pseudobulk for each sample and cell cluster</span></span></span>
<span class="r-in"><span><span class="va">pb</span> <span class="op">&lt;-</span> <span class="fu"><a href="aggregateToPseudoBulk.html">aggregateToPseudoBulk</a></span><span class="op">(</span><span class="va">example_sce</span>,</span></span>
<span class="r-in"><span>  assay <span class="op">=</span> <span class="st">"counts"</span>,</span></span>
<span class="r-in"><span>  cluster_id <span class="op">=</span> <span class="st">"cluster_id"</span>,</span></span>
<span class="r-in"><span>  sample_id <span class="op">=</span> <span class="st">"sample_id"</span>,</span></span>
<span class="r-in"><span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># voom-style normalization</span></span></span>
<span class="r-in"><span><span class="va">res.proc</span> <span class="op">&lt;-</span> <span class="fu">processAssays</span><span class="op">(</span><span class="va">pb</span>, <span class="op">~</span><span class="va">group_id</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   B cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.11 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD14+ Monocytes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.13 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD4 T cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.087 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD8 T cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.034 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   FCGR3A+ Monocytes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.095 secs</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Differential expression analysis within each assay,</span></span></span>
<span class="r-in"><span><span class="co"># evaluated on the voom normalized data</span></span></span>
<span class="r-in"><span><span class="va">res.dl</span> <span class="op">&lt;-</span> <span class="fu"><a href="dreamlet.html">dreamlet</a></span><span class="op">(</span><span class="va">res.proc</span>, <span class="op">~</span><span class="va">group_id</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   B cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.37 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD14+ Monocytes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.074 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD4 T cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.054 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD8 T cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.036 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   FCGR3A+ Monocytes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.069 secs</span>
<span class="r-in"><span><span class="co">#</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

