<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Processing expression data from assay — processOneAssay • dreamlet</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Processing expression data from assay — processOneAssay"><meta name="description" content="For raw counts, filter genes and samples, then estimate precision weights using linear mixed model weighting by number of cells observed for each sample.  For normalized data, only weight by number of cells"><meta property="og:description" content="For raw counts, filter genes and samples, then estimate precision weights using linear mixed model weighting by number of cells observed for each sample.  For normalized data, only weight by number of cells"><meta property="og:image" content="http://DiseaseNeurogenomics.github.io/dreamlet/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dreamlet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/dreamlet.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/mashr.html">mashr analysis after dreamlet</a></li>
    <li><a class="dropdown-item" href="../articles/cell_covs.html">Modeling continuous cell-level covariates</a></li>
    <li><a class="dropdown-item" href="../articles/non_lin_eff.html">Testing non-linear effects</a></li>
    <li><a class="dropdown-item" href="../articles/h5ad_on_disk.html">Handling large H5AD datasets</a></li>
    <li><a class="dropdown-item" href="../articles/errors.html">Error handling</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DiseaseNeurogenomics/dreamlet/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Processing expression data from assay</h1>
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeurogenomics/dreamlet/blob/HEAD/R/processAssays.R" class="external-link"><code>R/processAssays.R</code></a></small>
      <div class="d-none name"><code>processOneAssay.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>For raw counts, filter genes and samples, then estimate precision weights using linear mixed model weighting by number of cells observed for each sample.  For normalized data, only weight by number of cells</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">processOneAssay</span><span class="op">(</span></span>
<span>  <span class="va">y</span>,</span>
<span>  <span class="va">formula</span>,</span>
<span>  <span class="va">data</span>,</span>
<span>  <span class="va">n.cells</span>,</span>
<span>  min.cells <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  min.count <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  min.samples <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  min.prop <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>  min.total.count <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  isCounts <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  normalize.method <span class="op">=</span> <span class="st">"TMM"</span>,</span>
<span>  span <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  weights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rescaleWeightsAfter <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  BPPARAM <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-y">y<a class="anchor" aria-label="anchor" href="#arg-y"></a></dt>
<dd><p>matrix of counts or log2 CPM</p></dd>


<dt id="arg-formula">formula<a class="anchor" aria-label="anchor" href="#arg-formula"></a></dt>
<dd><p>regression formula for differential expression analysis</p></dd>


<dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>metadata used in regression formula</p></dd>


<dt id="arg-n-cells">n.cells<a class="anchor" aria-label="anchor" href="#arg-n-cells"></a></dt>
<dd><p>array of cell count for each sample</p></dd>


<dt id="arg-min-cells">min.cells<a class="anchor" aria-label="anchor" href="#arg-min-cells"></a></dt>
<dd><p>minimum number of observed cells for a sample to be included in the analysis</p></dd>


<dt id="arg-min-count">min.count<a class="anchor" aria-label="anchor" href="#arg-min-count"></a></dt>
<dd><p>used to compute a CPM threshold of <code>CPM.cutoff = min.count/median(lib.size)*1e6</code>.  Passed to <code><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">edgeR::filterByExpr()</a></code></p></dd>


<dt id="arg-min-samples">min.samples<a class="anchor" aria-label="anchor" href="#arg-min-samples"></a></dt>
<dd><p>minimum number of samples passing cutoffs for cell cluster to be retained</p></dd>


<dt id="arg-min-prop">min.prop<a class="anchor" aria-label="anchor" href="#arg-min-prop"></a></dt>
<dd><p>minimum proportion of retained samples with <code>CPM &gt; CPM.cutoff</code></p></dd>


<dt id="arg-min-total-count">min.total.count<a class="anchor" aria-label="anchor" href="#arg-min-total-count"></a></dt>
<dd><p>minimum total count required per gene for inclusion</p></dd>


<dt id="arg-iscounts">isCounts<a class="anchor" aria-label="anchor" href="#arg-iscounts"></a></dt>
<dd><p>logical, indicating if data is raw counts</p></dd>


<dt id="arg-normalize-method">normalize.method<a class="anchor" aria-label="anchor" href="#arg-normalize-method"></a></dt>
<dd><p>normalization method to be used by <code>calcNormFactors</code></p></dd>


<dt id="arg-span">span<a class="anchor" aria-label="anchor" href="#arg-span"></a></dt>
<dd><p>Lowess smoothing parameter using by <code><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/voomWithDreamWeights.html" class="external-link">variancePartition::voomWithDreamWeights()</a></code></p></dd>


<dt id="arg-quiet">quiet<a class="anchor" aria-label="anchor" href="#arg-quiet"></a></dt>
<dd><p>show messages</p></dd>


<dt id="arg-weights">weights<a class="anchor" aria-label="anchor" href="#arg-weights"></a></dt>
<dd><p>matrix of precision weights</p></dd>


<dt id="arg-rescaleweightsafter">rescaleWeightsAfter<a class="anchor" aria-label="anchor" href="#arg-rescaleweightsafter"></a></dt>
<dd><p>default = FALSE, should the output weights be scaled by the input weights</p></dd>


<dt id="arg-bpparam">BPPARAM<a class="anchor" aria-label="anchor" href="#arg-bpparam"></a></dt>
<dd><p>parameters for parallel evaluation</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>other arguments passed to <code>dream</code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><code>EList</code> object storing log2 CPM and precision weights</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="processAssays.html">processAssays()</a></code></p></div>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

