<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Run mash analysis on dreamlet results — run_mash • dreamlet</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous"><link href="../docsearch.css" rel="stylesheet"><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Run mash analysis on dreamlet results — run_mash"><meta property="og:description" content="Run mash analysis on dreamlet results"><meta property="og:image" content="http://gabrielhoffman.github.io/dreamlet/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dreamlet</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.32</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/dreamlet.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/h5ad_on_disk.html">Loading large-scale H5AD datasets</a>
    </li>
    <li>
      <a href="../articles/mashr.html">mashr analysis following dreamlet</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/GabrielHoffman/dreamlet/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul><form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off"></div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run mash analysis on dreamlet results</h1>
    <small class="dont-index">Source: <a href="https://github.com/GabrielHoffman/dreamlet/blob/HEAD/R/mash.R" class="external-link"><code>R/mash.R</code></a></small>
    <div class="hidden name"><code>run_mash.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Run mash analysis on dreamlet results</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">run_mash</span><span class="op">(</span><span class="va">fit</span>, <span class="va">coefList</span><span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>fit</dt>
<dd><p>result from <code><a href="dreamlet.html">dreamlet()</a></code></p></dd>
<dt>coefList</dt>
<dd><p>coefficient to be analyzed</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>a list storing the <code>mashr</code> model as <code>model</code> and the original coefficients as <code>logFC.original</code></p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Apply <a href="https://cran.r-project.org/web/packages/mashr/index.html" class="external-link">mashr</a> analysis (Urbut et al. 2019)
 on the joint set of coefficients for each gene and cell type.  <code>mashr</code> is a Bayesian statistical method that borrows strength across tests (i.e. genes and cell types) by learning the distribution of non-zero effects based the obesrved logFC and standard errors.  The method then estimates the posterior distributions of each coefficient based on the observed value and the genome-wide emprical distribution.</p>
<p><code>mashr</code> has been previously applied to differential expression in <a href="https://www.gtexportal.org" class="external-link">GTEx</a> data using multiple tissues from the same set of donors (Oliva et al. 2020)
.</p>
<p>In single cell data, a given gene is often not sufficiently expressed in all cell types.  So it is not evaluated in a subsets of cell types, and its coefficient value is <code>NA</code>. Since mashr assumes coefficients and standard errors for every gene and cell type pair, entries with these missing values are set to have <code>coef = 0</code>, and <code>se = 1e6</code>.  The output of mashr is then modified to set the corresponding values to <code>NA</code>, to avoid nonsensical results downstream.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Oliva M, Munoz-Aguirre M, Kim-Hellmuth S, Wucher V, Gewirtz AD, Cotter DJ, Parsana P, Kasela S, Balliu B, Vinuela A, others (2020).
“The impact of sex on gene expression across human tissues.”
<em>Science</em>, <b>369</b>(6509), eaba3066.
<a href="https://doi.org/10.1126/science.aba3066" class="external-link">https://doi.org/10.1126/science.aba3066</a>.<br><br> Urbut SM, Wang G, Carbonetto P, Stephens M (2019).
“Flexible statistical methods for estimating and testing effects in genomic studies with multiple conditions.”
<em>Nature genetics</em>, <b>51</b>(1), 187--195.
<a href="https://doi.org/10.1038/s41588-018-0268-8" class="external-link">https://doi.org/10.1038/s41588-018-0268-8</a>.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/mashr/man/mash_estimate_corr_em.html" class="external-link">mashr::mash_estimate_corr_em()</a></code>, <code><a href="https://rdrr.io/pkg/mashr/man/cov_canonical.html" class="external-link">mashr::cov_canonical</a></code>, <code><a href="https://rdrr.io/pkg/mashr/man/mash_set_data.html" class="external-link">mashr::mash_set_data</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HelenaLC/muscat" class="external-link">muscat</a></span><span class="op">)</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stephenslab/mashr" class="external-link">mashr</a></span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: ashr</span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/pkg/muscat/man/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># create pseudobulk for each sample and cell cluster</span></span>
<span class="r-in"><span class="va">pb</span> <span class="op">&lt;-</span> <span class="fu"><a href="aggregateToPseudoBulk.html">aggregateToPseudoBulk</a></span><span class="op">(</span><span class="va">example_sce</span>, </span>
<span class="r-in">  assay <span class="op">=</span> <span class="st">"counts"</span>,    </span>
<span class="r-in">  cluster_id <span class="op">=</span> <span class="st">'cluster_id'</span>, </span>
<span class="r-in">  sample_id <span class="op">=</span> <span class="st">'sample_id'</span>,</span>
<span class="r-in">  verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># voom-style normalization</span></span>
<span class="r-in"><span class="va">res.proc</span> <span class="op">=</span> <span class="fu"><a href="processAssays.html">processAssays</a></span><span class="op">(</span> <span class="va">pb</span>, <span class="op">~</span> <span class="va">group_id</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   B cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.032 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD14+ Monocytes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.018 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD4 T cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.02 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD8 T cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.019 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   FCGR3A+ Monocytes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.021 secs</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Differential expression analysis within each assay,</span></span>
<span class="r-in"><span class="co"># evaluated on the voom normalized data </span></span>
<span class="r-in"><span class="va">res.dl</span> <span class="op">=</span> <span class="fu"><a href="dreamlet.html">dreamlet</a></span><span class="op">(</span> <span class="va">res.proc</span>, <span class="op">~</span> <span class="va">group_id</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   B cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.039 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD14+ Monocytes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.059 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD4 T cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.043 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD8 T cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.021 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   FCGR3A+ Monocytes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.052 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># run MASH model</span></span>
<span class="r-in"><span class="co"># This can take 10s of minutes on real data</span></span>
<span class="r-in"><span class="co"># This small datasets should take ~30s</span></span>
<span class="r-in"><span class="va">res_mash</span> <span class="op">=</span> <span class="fu">run_mash</span><span class="op">(</span> <span class="va">res.dl</span>, <span class="st">'group_idstim'</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># extract statistics from mashr model</span></span>
<span class="r-in"><span class="co"># NA values indicate genes not sufficiently expressed</span></span>
<span class="r-in"><span class="co"># in a given cell type</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># original logFC</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res_mash</span><span class="op">$</span><span class="va">logFC.original</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          B cells CD14+ Monocytes CD4 T cells CD8 T cells FCGR3A+ Monocytes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ISG20  3.5977107       6.1228165   3.0731003   3.7280605         4.9675542</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ISG15         NA       8.1862294   5.5005509          NA         7.5386815</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> IRF8   2.8610495              NA          NA          NA                NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CXCL10        NA              NA          NA          NA         5.2692770</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> UBE2L6 2.6902060       3.0387733   2.2976869          NA         1.9672296</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> B2M    0.7765687       0.9465299   0.7490134   0.5667585         0.4578611</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># posterior mean for logFC</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ashr/man/get_lfdr.html" class="external-link">get_pm</a></span><span class="op">(</span><span class="va">res_mash</span><span class="op">$</span><span class="va">model</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          B cells CD14+ Monocytes CD4 T cells CD8 T cells FCGR3A+ Monocytes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ISG20  3.6201965       5.9543024   3.0940579   3.7474017         4.8601198</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ISG15         NA       7.9791899   5.5248155          NA         7.4170174</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> IRF8   2.7637566              NA          NA          NA                NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> CXCL10        NA              NA          NA          NA         5.2524025</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> UBE2L6 2.6253831       2.9768342   2.2885609          NA         1.9513343</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> B2M    0.7662939       0.9141552   0.7442989   0.5792546         0.4655056</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># how many gene-by-celltype tests are significant</span></span>
<span class="r-in"><span class="co"># i.e.  if a gene is significant in 2 celltypes, it is counted twice</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ashr/man/get_lfdr.html" class="external-link">get_lfsr</a></span><span class="op">(</span><span class="va">res_mash</span><span class="op">$</span><span class="va">model</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, useNA<span class="op">=</span><span class="st">"ifany"</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FALSE  TRUE  &lt;NA&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  2407  1102  2346 </span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># how many genes are significant in at least one cell type</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ashr/man/get_lfdr.html" class="external-link">get_lfsr</a></span><span class="op">(</span><span class="va">res_mash</span><span class="op">$</span><span class="va">model</span><span class="op">)</span>, <span class="fl">1</span>, <span class="va">min</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FALSE  TRUE </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   593   578 </span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># how many genes are significant in each cell type</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ashr/man/get_lfdr.html" class="external-link">get_lfsr</a></span><span class="op">(</span><span class="va">res_mash</span><span class="op">$</span><span class="va">model</span><span class="op">)</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           B cells   CD14+ Monocytes       CD4 T cells       CD8 T cells </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>               143               488               144                77 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FCGR3A+ Monocytes </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>               250 </span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># examine top set of genes</span></span>
<span class="r-in"><span class="co"># which genes are significant in at least 1 cell type</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/mashr/man/get_significant_results.html" class="external-link">get_significant_results</a></span><span class="op">(</span><span class="va">res_mash</span><span class="op">$</span><span class="va">model</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "ACTB"                  "ACTG1_ENSG00000184009" "ARHGDIB"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4] "ATP5L_ENSG00000167283" "ATP6V1G1"              "B2M"                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7] "BRK1"                  "BTF3"                  "BTG1"                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10] "CALM1"                </span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Lets examine CXCL3</span></span>
<span class="r-in"><span class="co"># There is a lot of variation in the raw logFC</span></span>
<span class="r-in"><span class="va">res_mash</span><span class="op">$</span><span class="va">logFC.original</span><span class="op">[</span><span class="st">"ISG20"</span>,<span class="op">]</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           B cells   CD14+ Monocytes       CD4 T cells       CD8 T cells </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          3.597711          6.122816          3.073100          3.728061 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FCGR3A+ Monocytes </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          4.967554 </span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># posterior mean after borrowing across cell type and genes</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/pkg/ashr/man/get_lfdr.html" class="external-link">get_pm</a></span><span class="op">(</span><span class="va">res_mash</span><span class="op">$</span><span class="va">model</span><span class="op">)</span><span class="op">[</span><span class="st">"ISG20"</span>,<span class="op">]</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           B cells   CD14+ Monocytes       CD4 T cells       CD8 T cells </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          3.620196          5.954302          3.094058          3.747402 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FCGR3A+ Monocytes </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          4.860120 </span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># gene set analysis using mashr results</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">zenith</span><span class="op">)</span></span>
<span class="r-in"><span class="va">go.gs</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/zenith/man/get_GeneOntology.html" class="external-link">get_GeneOntology</a></span><span class="op">(</span><span class="st">"CC"</span>, to<span class="op">=</span><span class="st">"SYMBOL"</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using cached version from 2022-02-18 20:07:31</span>
<span class="r-in"><span class="co"># valid values for statistic: </span></span>
<span class="r-in"><span class="co"># "tstatistic", "abs(tstatistic)", "logFC", "abs(logFC)"</span></span>
<span class="r-in"><span class="va">df_gs</span> <span class="op">=</span> <span class="fu"><a href="zenith_gsa-methods.html">zenith_gsa</a></span><span class="op">(</span><span class="va">res_mash</span>, <span class="va">go.gs</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Heatmap of results</span></span>
<span class="r-in"><span class="fu"><a href="plotZenithResults.html">plotZenithResults</a></span><span class="op">(</span><span class="va">df_gs</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span></span>
<span class="r-plt img"><img src="run_mash-1.png" alt="" width="700" height="433"></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># forest plot based on mashr results</span></span>
<span class="r-in"><span class="fu"><a href="plotForest-methods.html">plotForest</a></span><span class="op">(</span><span class="va">res_mash</span>, <span class="st">"ISG20"</span><span class="op">)</span></span>
<span class="r-plt img"><img src="run_mash-2.png" alt="" width="700" height="433"></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># volcano plot based on mashr results</span></span>
<span class="r-in"><span class="co"># yaxis uses local false sign rate (lfsr)</span></span>
<span class="r-in"><span class="fu"><a href="plotVolcano-methods.html">plotVolcano</a></span><span class="op">(</span><span class="va">res_mash</span><span class="op">)</span></span>
<span class="r-plt img"><img src="run_mash-3.png" alt="" width="700" height="433"></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'dedcc260e9d3743bc634e728855cdbaf',
    indexName: 'dreamlet',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script></body></html>

