<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Run mash analysis on dreamlet results"><title>Run mash analysis on dreamlet results — run_mash • dreamlet</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Run mash analysis on dreamlet results — run_mash"><meta property="og:description" content="Run mash analysis on dreamlet results"><meta property="og:image" content="http://DiseaseNeurogenomics.github.io/dreamlet/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dreamlet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.21</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/dreamlet.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/mashr.html">mashr analysis after dreamlet</a>
    <a class="dropdown-item" href="../articles/cell_covs.html">Modeling continuous cell-level covariates</a>
    <a class="dropdown-item" href="../articles/non_lin_eff.html">Testing non-linear effects</a>
    <a class="dropdown-item" href="../articles/h5ad_on_disk.html">Handling large H5AD datasets</a>
    <a class="dropdown-item" href="../articles/errors.html">Error handling</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/DiseaseNeurogenomics/dreamlet/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Run mash analysis on dreamlet results</h1>
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeurogenomics/dreamlet/blob/HEAD/R/mash.R" class="external-link"><code>R/mash.R</code></a></small>
      <div class="d-none name"><code>run_mash.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Run mash analysis on dreamlet results</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_mash</span><span class="op">(</span><span class="va">fit</span>, <span class="va">coefList</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>fit</dt>
<dd><p>result from <code><a href="dreamlet.html">dreamlet()</a></code></p></dd>


<dt>coefList</dt>
<dd><p>coefficient to be analyzed. Assumes 1) the null distribution of the two coefficients is simular, 2) the effects sizes are on the same scale, and 3) the effect estimates should be shrunk towards each other.  If these are not satisfied, run separately on each coefficient</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>a list storing the <code>mashr</code> model as <code>model</code> and the original coefficients as <code>logFC.original</code></p>


    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Apply <a href="https://cran.r-project.org/web/packages/mashr/index.html" class="external-link">mashr</a> analysis (Urbut et al. 2019)
 on the joint set of coefficients for each gene and cell type.  <code>mashr</code> is a Bayesian statistical method that borrows strength across tests (i.e. genes and cell types) by learning the distribution of non-zero effects based the obesrved logFC and standard errors.  The method then estimates the posterior distributions of each coefficient based on the observed value and the genome-wide emprical distribution.</p>
<p><code>mashr</code> has been previously applied to differential expression in <a href="https://www.gtexportal.org" class="external-link">GTEx</a> data using multiple tissues from the same set of donors (Oliva et al. 2020)
.</p>
<p>In single cell data, a given gene is often not sufficiently expressed in all cell types.  So it is not evaluated in a subsets of cell types, and its coefficient value is <code>NA</code>. Since mashr assumes coefficients and standard errors for every gene and cell type pair, entries with these missing values are set to have <code>coef = 0</code>, and <code>se = 1e6</code>.  The output of mashr is then modified to set the corresponding values to <code>NA</code>, to avoid nonsensical results downstream.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Oliva M, Munoz-Aguirre M, Kim-Hellmuth S, Wucher V, Gewirtz AD, Cotter DJ, Parsana P, Kasela S, Balliu B, Vinuela A, others (2020).
“The impact of sex on gene expression across human tissues.”
<em>Science</em>, <b>369</b>(6509), eaba3066.
<a href="https://doi.org/10.1126/science.aba3066" class="external-link">https://doi.org/10.1126/science.aba3066</a>.<br><br> Urbut SM, Wang G, Carbonetto P, Stephens M (2019).
“Flexible statistical methods for estimating and testing effects in genomic studies with multiple conditions.”
<em>Nature genetics</em>, <b>51</b>(1), 187--195.
<a href="https://doi.org/10.1038/s41588-018-0268-8" class="external-link">https://doi.org/10.1038/s41588-018-0268-8</a>.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/mashr/man/mash_estimate_corr_em.html" class="external-link">mashr::mash_estimate_corr_em()</a></code>, <code><a href="https://rdrr.io/pkg/mashr/man/cov_canonical.html" class="external-link">mashr::cov_canonical</a></code>, <code><a href="https://rdrr.io/pkg/mashr/man/mash_set_data.html" class="external-link">mashr::mash_set_data</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HelenaLC/muscat" class="external-link">muscat</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stephenslab/mashr" class="external-link">mashr</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: ashr</span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create pseudobulk for each sample and cell cluster</span></span></span>
<span class="r-in"><span><span class="va">pb</span> <span class="op">&lt;-</span> <span class="fu"><a href="aggregateToPseudoBulk.html">aggregateToPseudoBulk</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>, <span class="op">]</span>,</span></span>
<span class="r-in"><span>  assay <span class="op">=</span> <span class="st">"counts"</span>,</span></span>
<span class="r-in"><span>  cluster_id <span class="op">=</span> <span class="st">"cluster_id"</span>,</span></span>
<span class="r-in"><span>  sample_id <span class="op">=</span> <span class="st">"sample_id"</span>,</span></span>
<span class="r-in"><span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># voom-style normalization</span></span></span>
<span class="r-in"><span><span class="va">res.proc</span> <span class="op">&lt;-</span> <span class="fu"><a href="processAssays.html">processAssays</a></span><span class="op">(</span><span class="va">pb</span>, <span class="op">~</span><span class="va">group_id</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   B cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.051 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD14+ Monocytes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.056 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD4 T cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.06 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD8 T cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.054 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   FCGR3A+ Monocytes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.057 secs</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Differential expression analysis within each assay,</span></span></span>
<span class="r-in"><span><span class="co"># evaluated on the voom normalized data</span></span></span>
<span class="r-in"><span><span class="va">res.dl</span> <span class="op">&lt;-</span> <span class="fu"><a href="dreamlet.html">dreamlet</a></span><span class="op">(</span><span class="va">res.proc</span>, <span class="op">~</span><span class="va">group_id</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   B cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.031 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD14+ Monocytes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.032 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD4 T cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.03 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   CD8 T cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.024 secs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   FCGR3A+ Monocytes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0.032 secs</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># run MASH model</span></span></span>
<span class="r-in"><span><span class="co"># This can take 10s of minutes on real data</span></span></span>
<span class="r-in"><span><span class="co"># This small datasets should take ~30s</span></span></span>
<span class="r-in"><span><span class="va">res_mash</span> <span class="op">&lt;-</span> <span class="fu">run_mash</span><span class="op">(</span><span class="va">res.dl</span>, <span class="st">"group_idstim"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># extract statistics from mashr model</span></span></span>
<span class="r-in"><span><span class="co"># NA values indicate genes not sufficiently expressed</span></span></span>
<span class="r-in"><span><span class="co"># in a given cell type</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># original logFC</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res_mash</span><span class="op">$</span><span class="va">logFC.original</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                          B cells CD14+ Monocytes CD4 T cells CD8 T cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AGTRAP                        NA       0.1413919          NA          NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AKR1A1                 0.2593698       0.4062277          NA          NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> APH1A_ENSG00000117362         NA       0.8309128 -0.01276698          NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ATP1A1                 0.4155968      -0.8213405 -0.50007017  -0.4932812</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ATP5F1                -0.5963830      -1.4192064 -0.21508103  -0.3187010</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ATP6V0B               -0.5477765      -0.6362835  0.09203165   0.4527028</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                       FCGR3A+ Monocytes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AGTRAP                       0.09674802</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AKR1A1                      -0.39808296</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> APH1A_ENSG00000117362       -0.47531235</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ATP1A1                      -0.28642822</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ATP5F1                      -1.11230092</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ATP6V0B                     -0.59144737</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># posterior mean for logFC</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ashr/man/get_lfdr.html" class="external-link">get_pm</a></span><span class="op">(</span><span class="va">res_mash</span><span class="op">$</span><span class="va">model</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                           B cells CD14+ Monocytes CD4 T cells CD8 T cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AGTRAP                         NA      0.03622517          NA          NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AKR1A1                 0.06170192      0.19704031          NA          NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> APH1A_ENSG00000117362          NA      0.65248459  0.03004709          NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ATP1A1                 0.06322554     -0.74159540 -0.24386689  -0.2217077</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ATP5F1                -0.51539775     -1.15909302 -0.32247805  -0.4103796</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ATP6V0B               -0.37035262     -0.53126075 -0.07334986   0.1531089</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                       FCGR3A+ Monocytes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AGTRAP                       0.01943363</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> AKR1A1                      -0.15086480</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> APH1A_ENSG00000117362       -0.22254771</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ATP1A1                      -0.13633909</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ATP5F1                      -0.75977859</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ATP6V0B                     -0.55588230</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># how many gene-by-celltype tests are significant</span></span></span>
<span class="r-in"><span><span class="co"># i.e.  if a gene is significant in 2 celltypes, it is counted twice</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ashr/man/get_lfdr.html" class="external-link">get_lfsr</a></span><span class="op">(</span><span class="va">res_mash</span><span class="op">$</span><span class="va">model</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, useNA <span class="op">=</span> <span class="st">"ifany"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FALSE  TRUE  &lt;NA&gt; </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   276    99   110 </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># how many genes are significant in at least one cell type</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ashr/man/get_lfdr.html" class="external-link">get_lfsr</a></span><span class="op">(</span><span class="va">res_mash</span><span class="op">$</span><span class="va">model</span><span class="op">)</span>, <span class="fl">1</span>, <span class="va">min</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FALSE  TRUE </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    48    49 </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># how many genes are significant in each cell type</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ashr/man/get_lfdr.html" class="external-link">get_lfsr</a></span><span class="op">(</span><span class="va">res_mash</span><span class="op">$</span><span class="va">model</span><span class="op">)</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           B cells   CD14+ Monocytes       CD4 T cells       CD8 T cells </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                14                43                12                 6 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FCGR3A+ Monocytes </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                24 </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># examine top set of genes</span></span></span>
<span class="r-in"><span><span class="co"># which genes are significant in at least 1 cell type</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/mashr/man/get_significant_results.html" class="external-link">get_significant_results</a></span><span class="op">(</span><span class="va">res_mash</span><span class="op">$</span><span class="va">model</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "ATP1A1"  "ATP5F1"  "ATP6V0B" "CAPZB"   "CD52"    "CD53"    "ENO1"   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [8] "GNG5"    "MCL1"    "NDUFS5" </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Lets examine ENO1</span></span></span>
<span class="r-in"><span><span class="co"># There is a lot of variation in the raw logFC</span></span></span>
<span class="r-in"><span><span class="va">res_mash</span><span class="op">$</span><span class="va">logFC.original</span><span class="op">[</span><span class="st">"ENO1"</span>, <span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           B cells   CD14+ Monocytes       CD4 T cells       CD8 T cells </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        0.04258043       -1.25091407        0.58319641        0.44458114 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FCGR3A+ Monocytes </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       -1.07292439 </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># posterior mean after borrowing across cell type and genes</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/ashr/man/get_lfdr.html" class="external-link">get_pm</a></span><span class="op">(</span><span class="va">res_mash</span><span class="op">$</span><span class="va">model</span><span class="op">)</span><span class="op">[</span><span class="st">"ENO1"</span>, <span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           B cells   CD14+ Monocytes       CD4 T cells       CD8 T cells </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        -0.0149462        -1.1116816         0.3896256         0.2298794 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> FCGR3A+ Monocytes </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        -0.9704445 </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># forest plot based on mashr results</span></span></span>
<span class="r-in"><span><span class="fu"><a href="plotForest-methods.html">plotForest</a></span><span class="op">(</span><span class="va">res_mash</span>, <span class="st">"ENO1"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="run_mash-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># volcano plot based on mashr results</span></span></span>
<span class="r-in"><span><span class="co"># yaxis uses local false sign rate (lfsr)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="plotVolcano-methods.html">plotVolcano</a></span><span class="op">(</span><span class="va">res_mash</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Removed 12 rows containing missing values (`geom_text_repel()`).</span>
<span class="r-plt img"><img src="run_mash-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Comment out to reduce package runtime</span></span></span>
<span class="r-in"><span><span class="co"># gene set analysis using mashr results</span></span></span>
<span class="r-in"><span><span class="co"># library(zenith)</span></span></span>
<span class="r-in"><span><span class="co"># go.gs = get_GeneOntology("CC", to="SYMBOL")</span></span></span>
<span class="r-in"><span><span class="co"># df_gs = zenith_gsa(res_mash, go.gs)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Heatmap of results</span></span></span>
<span class="r-in"><span><span class="co"># plotZenithResults(df_gs, 2, 1)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

