<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Extract a table of the top-ranked genes from a dreamlet fit."><title>Table of Top Genes from dreamlet fit — topTable,dreamletResult-method • dreamlet</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Table of Top Genes from dreamlet fit — topTable,dreamletResult-method"><meta property="og:description" content="Extract a table of the top-ranked genes from a dreamlet fit."><meta property="og:image" content="http://DiseaseNeurogenomics.github.io/dreamlet/logo.svg"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dreamlet</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.62</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/dreamlet.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/cell_level_covariates.html">Modeling continuous cell-level covariates</a>
    <a class="dropdown-item" href="../articles/h5ad_on_disk.html">Handling large H5AD datasets</a>
    <a class="dropdown-item" href="../articles/mashr.html">mashr analysis after dreamlet</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/DiseaseNeurogenomics/dreamlet/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Table of Top Genes from dreamlet fit</h1>
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeurogenomics/dreamlet/blob/HEAD/R/dreamlet.R" class="external-link"><code>R/dreamlet.R</code></a></small>
      <div class="d-none name"><code>topTable-methods.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Extract a table of the top-ranked genes from a dreamlet fit.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for dreamletResult</span></span>
<span><span class="fu">topTable</span><span class="op">(</span></span>
<span>  <span class="va">fit</span>,</span>
<span>  coef <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  number <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  genelist <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  sort.by <span class="op">=</span> <span class="st">"P"</span>,</span>
<span>  resort.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p.value <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  lfc <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  confint <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>fit</dt>
<dd><p>dreamletResult object</p></dd>


<dt>coef</dt>
<dd><p>coef</p></dd>


<dt>number</dt>
<dd><p>number</p></dd>


<dt>genelist</dt>
<dd><p>genelist</p></dd>


<dt>adjust.method</dt>
<dd><p>adjust.method</p></dd>


<dt>sort.by</dt>
<dd><p>sort.by</p></dd>


<dt>resort.by</dt>
<dd><p>resort.by</p></dd>


<dt>p.value</dt>
<dd><p>p.value</p></dd>


<dt>lfc</dt>
<dd><p>lfc</p></dd>


<dt>confint</dt>
<dd><p>confint</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p><code>data.frame</code> storing hypothesis test for each gene and cell type</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">limma::topTable()</a></code>, <code>variancePartition::topTable()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HelenaLC/muscat" class="external-link">muscat</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create pseudobulk for each sample and cell cluster</span></span></span>
<span class="r-in"><span><span class="va">pb</span> <span class="op">&lt;-</span> <span class="fu"><a href="aggregateToPseudoBulk.html">aggregateToPseudoBulk</a></span><span class="op">(</span><span class="va">example_sce</span>, </span></span>
<span class="r-in"><span>   assay <span class="op">=</span> <span class="st">"counts"</span>,    </span></span>
<span class="r-in"><span>   cluster_id <span class="op">=</span> <span class="st">'cluster_id'</span>, </span></span>
<span class="r-in"><span>   sample_id <span class="op">=</span> <span class="st">'sample_id'</span>,</span></span>
<span class="r-in"><span>   verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># voom-style normalization</span></span></span>
<span class="r-in"><span><span class="va">res.proc</span> <span class="op">=</span> <span class="fu"><a href="processAssays.html">processAssays</a></span><span class="op">(</span> <span class="va">pb</span>, <span class="op">~</span> <span class="va">group_id</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   B cells...</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in (function (cond) .Internal(C_tryCatchHelper(addr, 1L, cond)))(structure(list(message = structure("Problem while computing `..1 = cell == k`.", names = ""),     trace = structure(list(call = list(pkgdown::build_reference(),         purrr::map(topics, build_reference_topic, pkg = pkg,             lazy = lazy, examples_env = examples_env, run_dont_run = run_dont_run),         .f(.x[[i]], ...), withCallingHandlers(data_reference_topic(topic,             pkg, examples_env = examples_env, run_dont_run = run_dont_run),             error = function(err) {                msg &lt;- c(paste0("Failed to parse Rd in ", topic$file_in),                   i = err$message)                abort(msg, parent = err)            }), data_reference_topic(topic, pkg, examples_env = examples_env,             run_dont_run = run_dont_run), run_examples(tags$tag_examples[[1]],             env = if (is.null(examples_env)) NULL else new.env(parent = examples_env),             topic = tools::file_path_sans_ext(topic$file_in),             run_dont_run = run_dont_run), highlight_examples(code,             topic, env = env), downlit::evaluate_and_highlight(code,             fig_save = fig_save_topic, env = child_env(env),             output_handler = evaluate::new_output_handler(value = pkgdown_print)),         evaluate::evaluate(code, child_env(env), new_device = TRUE,             output_handler = output_handler), evaluate_call(expr,             parsed$src[[i]], envir = envir, enclos = enclos,             debug = debug, last = i == length(out), use_try = stop_on_error !=                 2L, keep_warning = keep_warning, keep_message = keep_message,             output_handler = output_handler, include_timing = include_timing),         timing_fn(handle(ev &lt;- withCallingHandlers(withVisible(eval_with_user_handlers(expr,             envir, enclos, user_handlers)), warning = wHandler,             error = eHandler, message = mHandler))), handle(ev &lt;- withCallingHandlers(withVisible(eval_with_user_handlers(expr,             envir, enclos, user_handlers)), warning = wHandler,             error = eHandler, message = mHandler)), try(f, silent = TRUE),         tryCatch(expr, error = function(e) {            call &lt;- conditionCall(e)            if (!is.null(call)) {                if (identical(call[[1L]], quote(doTryCatch)))                   call &lt;- sys.call(-4L)                dcall &lt;- deparse(call, nlines = 1L)                prefix &lt;- paste("Error in", dcall, ": ")                LONG &lt;- 75L                sm &lt;- strsplit(conditionMessage(e), "\n")[[1L]]                w &lt;- 14L + nchar(dcall, type = "w") + nchar(sm[1L],                   type = "w")                if (is.na(w))                   w &lt;- 14L + nchar(dcall, type = "b") + nchar(sm[1L],                     type = "b")                if (w &gt; LONG)                   prefix &lt;- paste0(prefix, "\n  ")            }            else prefix &lt;- "Error : "            msg &lt;- paste0(prefix, conditionMessage(e), "\n")            .Internal(seterrmessage(msg[1L]))            if (!silent &amp;&amp; isTRUE(getOption("show.error.messages"))) {                cat(msg, file = outFile)                .Internal(printDeferredWarnings())            }            invisible(structure(msg, class = "try-error", condition = e))        }), tryCatchList(expr, classes, parentenv, handlers),         tryCatchOne(expr, names, parentenv, handlers[[1L]]),         doTryCatch(return(expr), name, parentenv, handler), withCallingHandlers(withVisible(eval_with_user_handlers(expr,             envir, enclos, user_handlers)), warning = wHandler,             error = eHandler, message = mHandler), withVisible(eval_with_user_handlers(expr,             envir, enclos, user_handlers)), eval_with_user_handlers(expr,             envir, enclos, user_handlers), eval(expr, envir,             enclos), eval(expr, envir, enclos), processAssays(pb,             ~group_id), lapply(assays, function(k) {            if (!quiet)                 message("  ", k, "...", appendLF = FALSE)            startTime = Sys.time()            y = assay(sceObj, k)            n.cells = n.cells_full[colnames(y), k, drop = FALSE]            data = merge(data_constant, dplyr::filter(metadata(sceObj)$aggr_means,                 cell == k), by.x = "row.names", by.y = metadata(sceObj)$agg_pars$by[2])            rownames(data) = data$Row.names            data = data[rownames(data_constant), , drop = FALSE]            data = droplevels(data)            res = processOneAssay(y, formula, data, n.cells,                 min.cells = min.cells, min.count = min.count,                 min.samples = min.samples, min.prop = min.prop,                 isCounts = isCounts, normalize.method = normalize.method,                 useCountsWeights = useCountsWeights, BPPARAM = BPPARAM,                 ...)            if (!quiet)                 message(format(Sys.time() - startTime, digits = 2))            res        }), FUN(X[[i]], ...), merge(data_constant, dplyr::filter(metadata(sceObj)$aggr_means,             cell == k), by.x = "row.names", by.y = metadata(sceObj)$agg_pars$by[2]),         dplyr::filter(metadata(sceObj)$aggr_means, cell == k),         filter.data.frame(metadata(sceObj)$aggr_means, cell ==             k), filter_rows(.data, ..., caller_env = caller_env()),         filter_eval(dots, mask = mask, error_call = error_call),         withCallingHandlers({            mask$eval_all_filter(dots, env_filter)        }, error = function(e) {            local_error_context(dots = dots, .index = env_filter$current_expression,                 mask = mask)            bullets &lt;- c(cnd_bullet_header("computing"), filter_bullets(e))            abort(bullets, call = error_call, parent = skip_internal_condition(e))        }), mask$eval_all_filter(dots, env_filter), dplyr_internal_error("dplyr:::filter_incompatible_size",             `&lt;named list&gt;`), abort(class = c(class, "dplyr:::internal_error"),             dplyr_error_data = data), signal_abort(cnd, .file),         signalCondition(cnd), `&lt;fn&gt;`(`&lt;dpl:::__&gt;`), abort(bullets,             call = error_call, parent = skip_internal_condition(e))),         parent = c(0L, 1L, 2L, 3L, 3L, 5L, 6L, 7L, 8L, 9L, 10L,         10L, 12L, 13L, 14L, 15L, 16L, 10L, 10L, 10L, 20L, 21L,         22L, 23L, 24L, 25L, 25L, 25L, 28L, 29L, 30L, 30L, 0L,         33L, 34L, 35L, 0L, 37L), visible = c(TRUE, TRUE, TRUE,         TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE,         TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE,         TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, FALSE, FALSE,         FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE        ), namespace = c("pkgdown", "purrr", "pkgdown", "base",         "pkgdown", "pkgdown", "pkgdown", "downlit", "evaluate",         "evaluate", "evaluate", "evaluate", "base", "base", "base",         "base", "base", "base", "base", "evaluate", "base", "base",         "dreamlet", "base", "dreamlet", "base", "dplyr", "dplyr",         "dplyr", "dplyr", "base", NA, "dplyr", "rlang", "rlang",         "base", "dplyr", "rlang"), scope = c("::", "::", "local",         "::", ":::", ":::", ":::", "::", "::", ":::", "local",         "local", "::", "::", "local", "local", "local", "::",         "::", ":::", "::", "::", "::", "::", "local", "::", "::",         ":::", ":::", ":::", "::", NA, ":::", "::", ":::", "::",         "local", "::"), error_frame = c(FALSE, FALSE, FALSE,         FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE,         FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE,         FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE,         TRUE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE,         FALSE, FALSE, FALSE)), row.names = c(NA, -38L), version = 2L, class = c("rlang_trace",     "rlib_trace", "tbl", "data.frame")), parent = NULL, body = c(x = "Input `..1` must be of size 1, not size 0.",     i = "The error occurred in group 1: metadata(pb)$agg_pars$by[2] = \"sample_id\"."    ), call = dplyr::filter(metadata(sceObj)$aggr_means, cell ==         k), use_cli_format = TRUE), class = c("rlang_error", "error", "condition"))):</span> error in evaluating the argument 'y' in selecting a method for function 'merge': Problem while computing `..1 = cell == k`.</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span style="color: #BB0000;">✖</span> Input `..1` must be of size 1, not size 0.</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> The error occurred in group 1: metadata(pb)$agg_pars$by[2] = "sample_id".</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Differential expression analysis within each assay,</span></span></span>
<span class="r-in"><span><span class="co"># evaluated on the voom normalized data </span></span></span>
<span class="r-in"><span><span class="va">res.dl</span> <span class="op">=</span> <span class="fu"><a href="dreamlet.html">dreamlet</a></span><span class="op">(</span> <span class="va">res.proc</span>, <span class="op">~</span> <span class="va">group_id</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in h(simpleError(msg, call)):</span> error in evaluating the argument 'x' in selecting a method for function 'dreamlet': object 'res.proc' not found</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># show coefficients estimated for each cell type</span></span></span>
<span class="r-in"><span><span class="fu"><a href="coefNames-methods.html">coefNames</a></span><span class="op">(</span><span class="va">res.dl</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in h(simpleError(msg, call)):</span> error in evaluating the argument 'obj' in selecting a method for function 'coefNames': object 'res.dl' not found</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># extract results using limma-style syntax</span></span></span>
<span class="r-in"><span><span class="co"># combines all cell types together</span></span></span>
<span class="r-in"><span><span class="co"># adj.P.Val gives study-wide FDR   </span></span></span>
<span class="r-in"><span><span class="fu">topTable</span><span class="op">(</span><span class="va">res.dl</span>, coef<span class="op">=</span><span class="st">"group_idstim"</span>, number<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in h(simpleError(msg, call)):</span> error in evaluating the argument 'fit' in selecting a method for function 'topTable': object 'res.dl' not found</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

